---
layout: tutorial_hands_on

title: Preparing a DMP
questions:
- What to cover in your DMP?
- How to collect data?
- How to copyright your data?
- How to estimate costs related to data management?
- What are good practice examples of data management?
- Which tools can be used to write a DMP?

objectives:
- Knowing what to cover in your DMP
- Knowing where to find data
- Understanding different types of licence
- Estimate costs involved with managing your data;
- Check the current and expected costs for your research data;
- Knowing how to collect data
- Knowing which tools can be used to write a DMP
time_estimation: 60M
subtopic: dmp
key_points:
- Importance of access, copyright, licence, costs
contributors:
- abotzki

---

## Lessons and Topics in this Section
- L Tools
  - T Introduction
- L DMPOnline
  - T 7 Main Themes
  - T Theme 1 Data description
  - T Theme 2 Ethical and legal issues
  - T Theme 3 Data documentation and metadata
  - T Theme 4 Data storage and backup during the project
  - T Theme 5 Data preservation after the end of the projects
  - T Theme 6 Data sharing and Reuse
  - T Theme 7 Responsibilities
- L VIB Forms
  - T General Information
  - T Datasets Information
  - T Data type 1 - Observational data
  - T Data type 2. Experimental data
  - T Data type 3. Simulation data
  - T Data type 4. Derived and compiled data
  - T Data type 5. Canonical data
  - T Data purpose
  - T Ethical and legal issues
  - T Making Data Findable
  - T Documentation and metadata
  - T Data storage & backup during the project
  - T Who will be able to access the data and under what conditions?
  - T Data preservation after the end of the project
  - T Responsibilities
  - T Ready to submit?
- L Data collection
  - T The Data Flow
  - T Discover Data
  - T Assess the Usefulness of Existing Data
- L Copyright and Intellectual Property Rights (IPR) Issues
  - T About Data Ownership
  - T Confidential or Privacy-Sensitive Data
- L Licences
  - T About Licences
- L Costs Involved with Managing Your Data
  - T introduction
  - T Data Management Cost Guide


## Tools

### Introduction
Before you start working, let us explore the tools we have at our disposal. In this section we will look at [DMPOnline](https://dmponline.be/) and, specifically for VIB staff, [VIB Forms](https://forms.vib.be/). Especially in the latter section, sample answers are provided in italics.

![Top 10 Excuses for not having a DMP](../../images/01_DMPTop10.png "Top 10 Excuses for not having a DMP")
TOP 10 excuses for not having a DMP by [FranziMachtDas](https://twitter.com/FranziMachtDas), CC-6B0Y-SA 4.0

## DMPOnline

### What is DMPOnline
Have a look at the video below for a general introduction on DMPOnline:

<iframe width="560" height="315" src="https://www.youtube.com/embed/kXuBC6HlO1g" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

DMPonline is an international online service that guides you in creating a DMP by answering a series of questions about your research project. It allows you to create, share, store, and revise your data management plans online. You will be asked to complete different sections of your DMP as we go through the other modules. As a result you will have written your own data management plan at the end of this course.

With DMPonline you can:

- Write your plan and keep it up-to-date
  You can easily update your DMP throughout the lifecycle of a project
- Share plans online
  DMPonline allows collaborative access, so you can share your DMP with other researchers, within and outside of your university.
- Create multiple plans
  You can store different DMPs for different projects. And you can make a copy of a previous plan as the basis for writing a new one.
- Download plans
  You can download your DMP in a variety of formats.

We recommend that graduate researchers share their data management plans with their supervisor(s).

### 7 Main Themes
*This part is based on 'Preparing a DMP' by Research Department University Library (Book Tower), UGent, licensed under CC-BY 4.0.*

In the introduction we mentioned funder's requirements. Most of them require you to write a Data Management Plan. A DMP outlines all key aspects of collecting, storing and managing research data during and after a project. For this they provide you with guidelines, forms, templates and examples. For more information you can check out the websites.

- [EC – Horizon 2020: guidelines](https://ec.europa.eu/research/openscience/index.cfm)
- [FWO template](https://www.fwo.be/media/1023898/fwo-dmp-templatedocx.docx)

Templates for a Data Management Plans (DMP) may vary from funder to funder, but they have many common elements: 'Plans typically state **what [meta]data** will be created and **how**, and outline the plans for **sharing** and **preservation**, noting what is appropriate given the nature of the data and any **restrictions** that may need to be applied' [DCC](https://www.dcc.ac.uk/resources/data-management-plans)

![Funders' Templates -80width](../../images/01_FundersTemplates.png "Funders' Templates")
(from 'Preparing a DMP' by Research Department University Library (Book Tower), UGent, licensed under CC-BY 4.0)


### Theme 1 Data description
This section contains the following questions:

- Will you generate/collect new data and/or make use of existing data?
- Describe the origin, type and format of the data (per dataset) and its (estimated) volume.

Here are a few tips:

- Only provide answers relating to research data (not publications, presentations, etc.)
- Remember: 'research data' come in many different forms!
- Break down and list your data types conveniently: e.g. by technique, purpose, phase...
- Provide enough details for outsiders to understand the sort of data involved:
  - content, digital/non-digital, raw/processed…
  - specify file formats (e.g. .csv, .txt, .docx…)
  - specify data source/collection methods (e.g. experimental, observational,
simulation… data)

### Theme 2 Ethical and legal issues
This section contains the following questions:

- Will you use personal data?
  Bear in mind that **personal data** is any information relating to an identified or (in)directly identiable living individual. This means that names, occupations, addresses, dates of birth, phone numbers, voice recordings and the likes all qualify as personal data. As such, these are subject to the **General Data Protection Regulation (GDPR)**.

- Are there any ethical issues concerning the creation and/or use of the data (e.g. experiments on humans or animals, or dual use)?
  Answer these questions with yes if
  - you need **permission from ethical committee** and/or **informed consent** for collecting & handling data (incl. archiving & sharing);
  - you work with **sensitive data** (could cause harm if disclosed – e.g. to endangered species, vulnerable sites/groups, national security, public health…).

- Is there potential for tech transfer and valorization? Will IP restrictions be claimed for the data you created?
  - Restrictions on collecting, sharing, archiving & reusing data may be applicable
  - e.g. your research has commercial potential. Sharing may not be possible (to protect confidential knowhow), or only after embargo period (to seek patent protection first)
  - e.g. you work with third-party data/collaborate with external partners. Check licenses/agreements to know what you can do.
- Do existing 3rd party agreements restrict dissemination or exploitation of the data you (re)use?

### Theme 3 Data documentation and metadata
This section contains the following questions:

- What documentation will be provided to enable understanding and reuse of the data collected/generated in this project?
- Will a metadata standard be used?

Later in this course you will learn more about data documentation. Suffice it here to say that documentation basically is any information needed to understand & reuse research data, and to reproduce your research – ideally following standards in your domain. In other words, this could be

- a codebook/data dictionary explaining variables
- a dataset structure, relation between files
- a study context
- protocols, methods, SOPs
- notes/annotations
- machine configurations
- the software used
- processing & analysis scripts
- consent information
- etc.

Ensure that it is clear what the study was about, what the variables in your data mean and what the codes mean if used. There should be no blind spots.

**Metadata** refers to a highly structured form of data documentation.

- Needed to search & find data (e.g. catalogues)
- Also used to further describe/annotate data
- Fixed set of elements, as defined by metadata 'schema'
- Use existing standards!
  - generic: e.g. [Simple Dublin Core](https://www.dublincore.org/specifications/dublin-core/usageguide/), [Datacite](https://schema.datacite.org/)
  - domain-specific: check via [fairsharing.org](https://fairsharing.org/), [RDA metadata directory](http://rd-alliance.github.io/metadata-directory), [DCC's list of disciplinary metadata](http://www.dcc.ac.uk/resources/metadata-standards)


![Metadata -80width](../../images/01_Metadata.png "Metadata")
'Metadata' by Jørgen Stamp, attribution: digitalbevaring.dk, licensed under [CC-BY 2.5 DK](https://creativecommons.org/licenses/by/2.5/dk/deed.en_GB)

### Theme 4 Data storage and backup during the project
This section contains the following questions:

- Where will data be stored?
- How will the data be backed up?
- Is there sufficient storage & backup capacity during the project?
- What are the expected costs for data storage & backup during the project? How will these costs be covered?
- Data security: how will you ensure that the data are securely stored and not accessed or modified by unauthorised persons?

Make sure you have multiple copies in different places. If possible, have an automatic back-up system!

### Theme 5 Data preservation after the end of the projects
This section contains the following questions:

- Which data will be retained for the expected 5-year period after the end of the project?
- Where will these data be archived (= stored for the long term)?
- What are the expected costs for data preservation during these 5 years? How will the costs be covered?

While during your research your data will probably be stored on a hard drive or a network share, after your research it will be archived, meaning it is likely to be deposited in a digital repository. Deposit your data in a trusted external data repository

- preferably domain-specific (if available)
- general-purpose, e.g. [Zenodo](https://zenodo.org/), [Dryad](https://datadryad.org/), [OSF](https://osf.io/)
- find a suitable home via [re3data.org](https://www.re3data.org/). Have a look at the video below to see how it works

<iframe width="560" height="315" src="https://www.youtube.com/embed/cutUrqJqjp4" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

Finally, bear in mind that archiving your data externally is not always possible, and sometimes data may need to be kept in-house!

### Theme 6 Data sharing and Reuse
This section contains the following questions:

- Are there any factors restricting or preventing the sharing of (some of) the data?
- Which data will be made available after the end of the project?
- Where/how will the data be made available for reuse?
- When will the data be made available?
- Who will be able to access the data and under what conditions?
- What are the expected costs for data sharing? How will these costs be covered?

Maybe this figure helps to explain this a little bit:

![Data Sharing -80width](../../images/01_DataSharing.png "Data Sharing")
From ''[What you need to know to prepare a data management plan](http://helios-eie.ekt.gr/EIE/bitstream/10442/15546/1/20170615-OpenAIRE-Athene-Grootveld-final.pdf)' by Marjan Grootveld, licensed under [CC-BY](https://creativecommons.org/licenses/by/4.0/)

Do share data via a data repository if possible. Do not share "upon request by email". In a repository like [Zenodo](https://zenodo.org/) data are findable, clearly licenced, and citable

![FAIR Data -80width](../../images/01_FAIRData.png "FAIR Data")

### Theme 7 Responsibilities
This section contains the following questions:

- Who will be responsible for data documentation & metadata?
- Who will be responsible for data storage & backup during the project?
- Who will be responsible for ensuring data preservation &
sharing?
- Who bears the end responsibility for updating & implementing this DMP?

By default, the PI bears end responsibility for the DMP & its implementation. Responsibility for day-to-day data management may be shared among project team members, and with institutional services (e.g. IT service), or it may be outsourced to others (e.g. data repository, external partner).

## VIB Forms

### General Information
This section merely asks for practical details concerning your DMP (your funder, project number, name and acronym, PI, etc.).

### Datasets Information
This section contains the following question:

- Will you generate/collect new data and/or make use of existing data?

For each of the following 5 data types, indicate which datasets you are reusing and/or generating. For each dataset:

- State its purpose, explaining the relation to the objectives of the project;
- List the corresponding data formats;
- Provide an estimation of the expected size of the datasets;
- Specify to whom it will be useful.

### Data type 1 - Observational data
This section contains the following questions:

- dataset 1.1 - Patient details
- dataset 1.2 - Tissue samples
- dataset 1.3 - Survey data

### Data type 2 - Experimental data
This section contains the following questions:

- Dataset 2.1 - Digital images
- Dataset 2.2 - Video and audio files
- Dataset 2.3 - Cytometry data
- Dataset 2.4 - Spectroscopy data
- Dataset 2.5 - Electrophysiology data
- Dataset 2.6 - Omics data
- Dataset 2.7 - Vectors
- Dataset 2.8 - Cell lines
- Dataset 2.9 - Genetically modified organisms
- Dataset 2.10 - Antibodies
- Dataset 2.11 - Synthetic compounds
- Dataset 2.12 - Recombinant compounds

### Data type 3 - Simulation data
This section contains the following questions:

- Do you have simulation data?

### Data type 4 - Derived and compiled data
This section contains the following questions:

- Dataset 4.1 – Research documentation
- Dataset 4.2 – Manuscripts
- Dataset 4.3 – Algorithms and scripts
- Dataset 4.4 – Software
- Dataset 4.5 – Molecular models

### Data type 5 - Canonical data
This section contains the following questions:

- Dataset 5.1 – Nucleic acid sequences
- Dataset 5.2 – Protein sequences
- Dataset 5.3 – Chemical structures
- Dataset 5.4 – Protein structures

### Data purpose
This section contains the following questions:

- For who will the data be valuable?
  *These datasets represent an important source of information for the laboratory of the PI (including future staff), for scientists, journalists and higher education teachers working in the field of <>, but also for non-profit organizations and industries active in the field of <>.*

### Ethical and legal issues
This section contains the following questions:

- Will you use personal data?
  Bear in mind that **personal data** is any information relating to an identified or (in)directly identiable living individual. This means that names, occupations, addresses, dates of birth, phone numbers, voice recordings and the likes all qualify as personal data. As such, these are subject to the **General Data Protection Regulation (GDPR)**.

- Are there any ethical issues concerning the creation and/or use of the data (e.g. experiments on humans or animals, or dual use)?
  Answer these questions with yes if
  - you need **permission from ethical committee** and/or **informed consent** for collecting & handling data (incl. archiving & sharing);
  - you work with **sensitive data** (could cause harm if disclosed – e.g. to endangered species, vulnerable sites/groups, national security, public health…).

- Is there potential for tech transfer and valorization? Will IP restrictions be claimed for the data you created?
  - Restrictions on collecting, sharing, archiving & reusing data may be applicable
  - e.g. your research has commercial potential. Sharing may not be possible (to protect confidential knowhow), or only after embargo period (to seek patent protection first)
  - e.g. you work with third-party data/collaborate with external partners. Check licenses/agreements to know what you can do.
- Do existing 3rd party agreements restrict dissemination or exploitation of the data you (re)use?

### Making Data Findable
Describe the file naming convention that will be used, and if applicable the persistent and unique identifiers e.g., DOI. Here's a sample answer:

*Digital files will be named following a standard procedure, so that all the name of all files in a given dataset will be in the same format: All names will start with the date (and time if applicable), followed by the project acronym, a short but specific descriptive name and a version number (containing leading zeros as needed) if applicable. Whenever possible names will be kept under 32 characters. Names will only contain letters, numbers and underscores. Dots will only be used for version control indicators (minor revisions indicated by decimal numbers, and major revisions by whole numbers):  YYYYMMDD_HHmm_Project_Experiment_ version.format*
*All changes in the files will be recorded. Data files will be stored in suitably labelled and organized folders and sub-folders, accompanied by a README.txt file containing all the associated metadata (see section 3). File names and locations will be recorded in the E-notebook to allow electronic records to be linked to the raw data.*

*Specific naming and search procedures will be applied to:*

- *Omics and sequencing data: Digital files, raw sequencing data files mainly (.fastq, .fastq.gz), will be named following an in-house procedure, so that all the name of all files in a given dataset will be in the same format. All names will start with a project 3 letter code (Project Code) and a random 6-alphanumeric character code (unique to each sample given the Project Code), followed by a specific descriptive name of the sample and the technology used (e.g.: 10x). Names will only contain letters, numbers and underscores.*
- *Human pluripotent cell lines: these cell lines will be named according to the nomenclature rules established by the hPSCreg.eu database (https://hpscreg.eu/about/naming-tool). Names consists of a maximum of 15 characters and have the following structure: up to 6 characters for the institution, an i or e indicating the cell line type, 3 characters for the unique donor ID and finally a letter that is incremented depending on how many cell lines are generated by the same donor. Subclones are indicated by a hyphen followed by the subclone number. Cell lines can be searched for in the hPSCreg.eu database based on cell type, country of origin, associated disease and/or date of derivation.*
- *Manuscripts: Metadata information will be submitted alongside the final version of the manuscript, including the names, titles, email addresses, ORCIDs and affiliations of all authors. Upon publication, this metadata information will also be submitted to bibliographic databases such as Medline. All manuscripts will be assigned a unique Digital Object Identifer (DOI) by the publisher. Manuscripts will be given a descriptive title, and will be accompanied by keywords provided by the authors in order to maximize their findability.*

### Documentation and metadata
This section contains the following questions:

- What documentation will be provided to enable understanding and reuse of the data collected/generated in this project?
  *Metadata will be documented by the research and technical staff at the time of data collection and analysis, by taking careful notes in the electronic laboratory notebook (E-notebook) and/or in hard copy lab notebooks that refer to specific datasets.*

  *All datasets will be accompanied by a README.txt file containing all the associated metadata (see more details below). The data will be generated following standardized protocols. Clear and detailed descriptions of these protocols will be stored in our lab protocol database, and published along with the results.*

- Will a metadata standard be used?
- State in detail which metadata will be created to make the data easy/easier to find and reuse.
  *Metadata will include the following elements:*
  - *Title: free text*
  - *Creator: Last name, first name, organization*
  - *Date and time reference*
  - *Subject: Choice of keywords and classifications*
  - *Description: Text explaining the content of the data set and other contextual information needed for the correct interpretation of the data, the software(s) (including version number) used to produce and to read the data, the purpose of the experiment, etc.*
  - *Format: Details of the file format*
  - *Resource Type: data set, image, audio, etc.*
  - *Identifier: DOI (when applicable)*
  - *Access rights: closed access, embargoed access, restricted access, open access.*
  *For specific datasets, additional metadata will be associated with the data file as appropriate. Give details as needed for the project.
  The final dataset will be accompanied by this information under the form of a README.txt document. This file will be located in the top level directory of the dataset and will also list the contents of the other files and outline the file-naming convention used. This will allow the data to be understood by other members of the laboratory and add contextual value to the dataset for future reuse.*

### Data storage & backup during the project
This section contains the following questions:

- Where will the data be stored?
  *Digital files will be stored on KU Leuven servers, except for private data that will be stored on KU Leuven secure server (digital vault). Add information about other data types as required (adjust and add more as needed):*
  - ***Tissue samples*** *Tissues will be stored locally in the laboratory. All human tissue samples will be registered with a Belgian biobank, in compliance with the Belgian law on human body material (dd 19-12-2008).*
  - ***Omics data*** *Omics data generated during the project will either be stored on KU Leuven servers or on The Flemish Supercomputer Centre (VSC), initially in the staging area and later in the archive area.*
  - ***Vectors*** *As a general rule at least two independently obtained clones will be preserved for each vector, both under the form of purified DNA (in -20°C freezer) and as a bacteria glycerol stock (-80°C). All published vectors and the associated sequences will be sent to the non-profit plasmid repository Addgene, which will take care of vector storage and shipping upon request.*
  - ***Cell lines*** *Newly created human cell lines will be stored locally in the laboratory in liquid nitrogen storage and will be deposited in the UZ Leuven-KU Leuven Biobank. Other human cell lines will be stored locally in liquid nitrogen cryostorage of the laboratory when actively used for experiments. Animal cell lines will be stored in liquid nitrogen cryostorage of the laboratory.*
  - ***Genetically modified organisms*** *Mice / Zebrafishes / Xenopus will be maintained in facilities of the Laboratory Animal Center of KU Leuven, which applies Standard Operation Procedures concerning housing, feeding, health monitoring to assure consistent care in accordance with European and national regulations and guidelines. All animals will be registered in the Leuven Animal Information System (LAIS) database, along with corresponding genotyping information, ethical approval documents and animal provider receipts. Drosophila lines will be stored in a dedicated room and managed using a specific database for storage of the corresponding information (including genotype, origin, number of vials and date of transfer, crossing schemes) and vial tracking via unique QR codes. Other biological and chemical samples: storage at 4°C and/or as frozen samples in cryovials as appropriate.*
  - ***Algorithms, scripts and softwares*** *All the relevant algorithms, scripts and software code driving the project will be stored in a private online git repository from the GitHub account of the department (https://github.com/vibcbd).*
  - ***Nucleic acid and protein sequences*** *All nucleic acid and protein sequences generated during the project will be stored on KU Leuven servers. Upon publication, all sequences supporting a manuscript will be made publicly available via repositories  such as the GenBank database or the European Nucleotide Archive (nucleotide sequences from primers / new genes / new genomes), NCBI Gene Expression Omnibus (microarray data / RNA-seq data / CHIPseq data), the Protein Database (for protein sequences), the EBI European Genome-phenome Archive (EGA) for personally identifiable (epi)genome and transcriptome sequences.*

- How will the data be backed up?
  *KU Leuven drives are backed-up according to the following scheme:*
  - *Data stored on the 'L-drive' is backed up daily using snapshot technology, where all incremental changes in respect of the previous version are kept online; the last 14 backups are kept.*
  - *Data stored on the 'J-drive' is backed up hourly, daily (every day at midnight) and weekly (at midnight between Saturday and Sunday); in each case the last 6 backups are kept.*
  - *Data stored on the digital vault is backed up using snapshot technology, where all incremental changes in respect of the previous version are kept online. As standard, 10% of the requested storage is reserved for backups using the following backup regime: an hourly backup (at 8 a.m., 12 p.m., 4 p.m. and 8 p.m.), the last 6 of which are kept; a daily backup (every day) at midnight, the last 6 of which are kept; and a weekly backup (every week) at midnight between Saturday and Sunday, the last 2 of which are kept.*
  - *All omics data stored on the Flemish Supercomputer Centre (VSC) will be transferred on a weekly basis to the archive area which is backed up.
  Incremental backups are done daily from one 20 TB QNAP NAS to a second 20 TB QNAP NAS.*

- Is there currently sufficient storage & backup capacity during the project?
- What are the expected costs for data storage and backup during the project? How will these costs be covered?
  *The total estimated cost of data storage during the project is <euro>. This estimation is based on the following costs:*
  - *The costs of digital data storage are as follows: 173,78€/TB/Year for the 'L-drive' and 519€/TB/Year for the 'J-drive'.*
  - *Maintaining a mouse colony alive costs about 1,200 euro per year (for 6 cages), excluding the costs of genotyping. When no experiment is planned with a particular mouse strain, and in compliance with the 3R's rule (https://www.nc3rs.org.uk), cryopreservation will thus be used to safeguard the strain, prevent genetic drift, loss of transgene and potential infections or breeding problems. Cryopreservation of sperm/embryos costs about 500 to 700 euro per genotype, plus a minimal annual storage fee (25 euro per strain for 250 to 500 embryos). Frozen specimen are kept in two separate liquid nitrogen tanks at two different sites on campus. When necessary, the costs of revitalization from cryopreserved sperm/embryos are about 1,100/600 euro.*
  - *Maintaining a Drosophila line alive costs about 4 euro per year including the fly food, vials and personnel's cost. The PI and the lab manager if applicable are responsible for the preservation of fly lines as well as administrative and experimental data. All published lines will be preserved for the remainder of the PI's research career.  All unpublished lines will be preserved for a minimum of 5 years after the end of the project.*
  - *Maintaining a Zebrafish line alive costs about <euro> per year. When no experiment is planned with a particular line, and in compliance with the 3R's rule (https://www.nc3rs.org.uk), cryopreservation will thus be used to safeguard the line, prevent genetic drift, loss of transgene and potential infections or breeding problems. Cryopreservation of sperm costs about <euro> per genotype, plus a minimal annual storage fee of <euro>. Frozen specimen are kept in two separate liquid nitrogen tanks at two different sites on campus. When necessary, the costs of revitalization from cryopreserved sperm are about <euro>.*
  - *Maintaining a Xenopus line alive costs about <euro> per year. When no experiment is planned with a particular line, and in compliance with the 3R's rule (https://www.nc3rs.org.uk), cryopreservation will thus be used to safeguard the line, prevent genetic drift, loss of transgene and potential infections or breeding problems. Cryopreservation of sperm costs about <euro> per genotype, plus a minimal annual storage fee of <euro>. Frozen specimen are kept in two separate liquid nitrogen tanks at two different sites on campus. When necessary, the costs of revitalization from cryopreserved sperm are about <euro>.*

  *Electricity costs for the -80° freezers present in the labs are included in general lab costs. Data storage and backup costs are included in general lab costs.*

- Data security: how will you ensure that the data are securely stored and not accessed or modified by unauthorized persons?
  *Both the 'L-drive' and 'J-drive' servers are accessible only by laboratory members, and are mirrored in the second ICTS datacenter for business continuity and disaster recovery so that a copy of the data can be recovered within an hour.*

  *Access to the digital vault is possible only through using a KU Leuven user-id and password, and user rights only grant access to the data in their own vault. Sensitive data transfer will be performed according to the best practices for 'Copying data to the secure environment' defined by KU Leuven. The operating system of the vault is maintained on a monthly basis, including the application of upgrades and security patches. The server in the vault is managed by ICTS, and only ICTS personnel (bound by the ICT code of conduct for staff) have administrator/root rights. A security service monitors the technical installations continuously, even outside working hours.*
  *All private data will be rendered anonymous before processing outside the digital vault. Only the PI will be granted access to the server to deposit private data. The PI will be the only responsible for linking patient information, survey data and/or tissue samples, and will strictly respect confidentiality. All de-identified data will be exported from the database by the PI, and stored on KU Leuven servers from where it can be accessed by the research and technical staff from the laboratory.*

### Who will be able to access the data and under what conditions?
This section contains the following questions:

- Are there any factors restricting or preventing the sharing of (some of) the data (e.g. as defined in an agreement with a 3rd party, legal restrictions)?

- Which data will be made available after the end of the project?
  *Participants to the present project are committed to publish research results to communicate them to peers and to a wide audience. All research outputs supporting publications will be made openly accessible. Depending on their nature, some data may be made available prior to publication, either on an individual basis to interested researchers and/or potential new collaborators, or publicly via repositories (e.g. negative data). We aim at communicating our results in top journals that require full disclosure upon publication of all included data, either in the main text, in supplementary material or in a data repository if requested by the journal and following deposit advice given by the journal. Depending on the journal, accessibility restrictions may apply. Biological material will be distributed to other parties if requested*

- Where/how will the data be made available for reuse?
  Options here include
  - In an Open Access repository
  - In a restricted access repository
  - Upon request by mail
  - Other
    *As a general rule, datasets will be made openly accessible via existing platforms that support FAIR data sharing (www.fairsharing.org). Sharing policies for specific research outputs are detailed below:*
    - ***Open-access publications*** *in peer-reviewed journals, including supplemental information.Patient data: Upon publication, all anonymized patient details supporting a manuscript will be made publicly available as supplemental information.*
    - ***Omics datasets*** *will be deposited in open access repositories such as the PRIDE Archive for proteomics data, the Panorama platform for targeted proteomics data, the EMBL-EBI platform for genomics and epigenomics data, the LIPID MAPS Lipidomics Gateway for lipidomics data, the Metabolomics Workbench Data Repository for metabolomics data, or the NCBI Gene Expression Omnibus (GEO) or the EBI ArrayExpress databases for functional genomics data.*
    - ***Vectors*** *Upon publication, all vectors supporting a manuscript will be made publicly available via the non-profit plasmid repository Addgene, along with the corresponding DNA sequences. Addgene in turns performs quality control on the DNA, curates the plasmids online with all relevant information (maps, sequences), and for a minimal cost (typically ? EUR) ships the vectors upon simple request and signature of a material transfer agreement. The MTA will be prepared before depositing the vectors with the help of our organization's Tech Transfer office. For transfer between nonprofit or academic institutions, Addgene typically uses the Uniform Biological Material Transfer Agreement (https://www.addgene.org/terms/1047/). All non-published vectors and the associated documentation will be shared by the PI upon request and after signature of a material transfer agreement, at no cost except the cost of shipment.*
    - ***Cell lines*** *All human pluripotent cell lines supporting publications will be registered in hPSCreg, the European human embryonic stem cell registry supported by the European Commission (https://hpscreg.eu/). Information about the deposited lines (including donor information, derivation method, availability and characterization) will also be made accessible. Registration of cell lines in hPSCreg will provide visibility, confirm ethical procurement and facilitate comparison with other hPSC lines. The PI will remain the distributor of the pluripotent cell lines. All other cell lines supporting publications will be deposited in the American Type Culture Collection (ATCC) database (https://www.atcc.org/), which is a private, non-profit biological resource center. This will provide a secure back-up for this material. Investigators can purchase cell lines from the ATCC database upon signature of a material transfer agreement (https://www.lgcstandards-atcc.org/~/media/PDFs/MTA_2.ashx) and, in some cases, of a Limited Use/Label License (e.g. for CRISPR products or iPSC materials) and/or a Customer Acceptance of Responsibility (for potentially highly pathogenic materials). Information about the cell lines (including organism, cell type, tissue, biosafety level and disease if applicable) will also be made accessible.*
    - ***Genetically modified organisms*** *All genetically modified organisms used in publications will be made available to researchers upon request at the time of publication.*
    - ***Other digital datasets*** *that support publications (including image, video or audio files, electrophysiology data, cytometry data, spectroscopy data and simulation data) will be made publicly available via an open research data platform such as Mendeley Data or Zenodo.*
    - ***Antibodies, synthetic and recombinant compounds*** *Samples will be stored as appropriate in the laboratory. Within availability, they will be shared with interested researchers upon request.*
    - ***Research documentation*** *All protocols used to generate published data will be described in the corresponding manuscript(s), and the related documentation will be included as supplementary information. These data and all other documents (daily logs, raw data) deposited in the E-Notebook are accessible to the PI and the research staff, and will be made available upon request.*
    - ***Manuscripts*** *All scientific publications will be shared openly. Manuscripts submitted for publication will be deposited in a pre-print server such as bioRxiv, arXiv, Nature Precedings or ASAPbio. At the time of publication, research results will be summarized on the PI's website (add website address) and post-print pdf versions of publications will be made available there if allowed by copyright agreements, possibly after an embargo as determined by the publisher. Before the end of the embargo or in cases where sharing the post-print is not allowed due to copyright agreements, a pre-print version of the manuscript will be made available. Publications will also be automatically listed in our institutional repository, Lirias 2.0, based on the authors name and ORCID ID.*
    - ***Algorithms, scripts and softwares*** *As soon as a manuscript is publicly available, the online private git repository containing the corresponding algorithms, scripts and software code will be changed to a public repository.*
    - ***Molecular models*** *All molecular models generated during the project will be stored on KU Leuven servers. Upon publication, all models supporting a manuscript will be made publicly available as supplementary information and via the BioModels Database. The datasets will be cited using the unique identifier ("BIOMD" or "MODEL" followed by 10 digits) generated by the BioModels database, and will also be accessible via the search page on the platform's website.*
    - ***Nucleic acid and protein sequences*** *Upon publication, all sequences supporting a manuscript will be made publicly available via repositories  such as the GenBank database or the European Nucleotide Archive (nucleotide sequences from primers / new genes / new genomes), NCBI Gene Expression Omnibus (microarray data / RNA-seq data / CHIPseq data), the Protein Database (for protein sequences).*
    - ***Chemical structures*** *Upon publication, all chemical structures supporting a manuscript will be made publicly available as supporting information and via the appropriate PubChem database (PubChem Substance, PubChem Compound, and PubChem BioAssay) within the NCBI's Entrez information retrieval system. Datasets will be cited using the unique Substance Identifier (SID) / Single Compound (SIC) / BioAssay Identifier (AID) record generated by PubChem, and will also be accessible via the Entrez search page.*
    - ***Protein structures*** *Upon publication, all structures supporting a manuscript will be made publicly available as supporting information and via the Protein Data Bank in Europe (PDBe) database. Coordinate sets produced by X-ray crystallography / NMR / electron microscopy / neutron diffraction / powder diffraction / fiber diffraction will be deposited to the PDBe together with the corresponding metadata (project acronym, author contact information, context and method of data collection, information about the sequence, chemistry, etc.). Datasets will be cited using the Digital Object Identifier (DOI) link generated by the PDBe, and will also be accessible via the elaborate search function available via PDBe.*
    - *Data that do not support publication will be either deposited in an open access repository or made available upon request by email.*

- When will the data be made available?

- Specify
  *As a general rule all research outputs will be made openly accessible at the latest at the time of publication. No embargo will be foreseen unless imposed e.g. by pending publications, potential IP requirements – note that patent application filing will be planned so that publications need not be delayed - or ongoing projects requiring confidential data. In those cases, datasets will be made publicly available as soon as the embargo date is reached.*

- Who will be able to access the data and under what conditions?
  *Whenever possible, datasets and the appropriate metadata will be made publicly available through repositories that support FAIR data sharing. As detailed above, metadata will contain sufficient information to support data interpretation and reuse, and will be conform to community norms. These repositories clearly describe their conditions of use (typically under a Creative Commons CC0 1.0 Universal (CC0 1.0) Public Domain Dedication, a Creative Commons Attribution (CC-BY)  or an ODC Public Domain Dedication and Licence, with a material transfer agreement when applicable). Interested parties will thereby be allowed to access data directly, and they will give credit to the authors for the data used by citing the corresponding DOI. For data shared directly by the PI, a material transfer agreement (and a non-disclosure agreement if applicable) will be concluded with the beneficiaries in order to clearly describe the types of reuse that are permitted.*

- What are the expected costs for data sharing? How will these costs be covered?
  *It is the intention to minimize data management costs by implementing standard procedures e.g. for metadata collection and file storage and organization from the start of the project, and by using free-to-use data repositories and dissemination facilities whenever possible. Data management costs will be covered by the laboratory budget.*


### Data preservation after the end of the project
This section contains the following questions:

- Which data will be retained for the expected 5 year period after the end of the project?
  *The minimum preservation term of 5 years after the end of the project will be applied to all datasets. If applicable: Datasets collected in the context of clinical research, which fall under the scope of the Belgian Law of 7 May 2004, will be archived for 25 years, in agreement with UZ Leuven policy and the European Regulation 536/2014 on clinical trials of medicinal products for human use.*

- Where will these data be archived (= stored for the long term)?
  *As a general rule, datasets will be made openly accessible, whenever possible via existing platforms that support FAIR data sharing (www.fairsharing.org), at the latest at the time of publication. For all other datasets, long term storage will be ensured as follows:*
  - ***Digital datasets*** *Files will be stored on the 'L-drive'.*
  - ***Tissue samples*** *Tissues will be stored locally in the laboratory.*
  - ***Omics data*** *Datasets will be stored on the 'L-drive' or, for larger datasets, on the Vlaams Supercomputer Centrum.*
  - ***Vectors*** *As a general rule at least two independently obtained clones will be preserved for each vector, both under the form of purified DNA (in -20°C freezer) and as a bacteria glycerol stock (-80°C).*
  - ***Cell lines*** *Human cell lines will be stored in the UZ Leuven Biobank (-80°C). Human pluripotent stem cell lines generated during this project will be deposited in hPSCreg. Animal cell lines will be stored in liquid nitrogen cryostorage of the laboratory.*
  - ***Genetically modified organisms*** *Drosophila lines will be housed locally. All other lines that are no actively used for experiments will be cryopreserved.*
  - *Other biological and chemical samples: storage at 4°C and/or as frozen samples in cryovials as appropriate.*


- What are the expected costs for data preservation during these 5 years? How will the costs be covered?
  *The total estimated cost of data storage during 5 years after the end of the project is <euro>. This estimation is based on the following costs:*

  - *The costs of digital data storage are as follows: 173,78€/TB/Year for the 'L-drive' and 519€/TB/Year for the 'J-drive'.*
  - *Maintaining a mouse colony alive costs about 1,200 euro per year (for 6 cages), excluding the costs of genotyping. When no experiment is planned with a particular mouse strain, and in compliance with the 3R's rule (https://www.nc3rs.org.uk), cryopreservation will thus be used to safeguard the strain, prevent genetic drift, loss of transgene and potential infections or breeding problems. Cryopreservation of sperm/embryos costs about 500 to 700 euro per genotype, plus a minimal annual storage fee (25 euro per strain for 250 to 500 embryos). Frozen specimen are kept in two separate liquid nitrogen tanks at two different sites on campus. When necessary, the costs of revitalization from cryopreserved sperm/embryos are about 1,100/600 euro.*
  - *Maintaining a Drosophila line alive costs about 4 euro per year including the fly food, vials and personnel's cost. The PI and the lab manager if applicable are responsible for the preservation of fly lines as well as administrative and experimental data. All published lines will be preserved for the remainder of the PI's research career.  All unpublished lines will be preserved for a minimum of 5 years after the end of the project.*
  - *Maintaining a Zebrafish line alive costs about <euro> per year. When no experiment is planned with a particular line, and in compliance with the 3R's rule (https://www.nc3rs.org.uk), cryopreservation will thus be used to safeguard the line, prevent genetic drift, loss of transgene and potential infections or breeding problems. Cryopreservation of sperm costs about <euro> per genotype, plus a minimal annual storage fee of <euro>. Frozen specimen are kept in two separate liquid nitrogen tanks at two different sites on campus. When necessary, the costs of revitalization from cryopreserved sperm are about <euro>.*
  - *Maintaining a Xenopus line alive costs about <euro> per year. When no experiment is planned with a particular line, and in compliance with the 3R's rule (https://www.nc3rs.org.uk), cryopreservation will thus be used to safeguard the line, prevent genetic drift, loss of transgene and potential infections or breeding problems. Cryopreservation of sperm costs about <euro> per genotype, plus a minimal annual storage fee of <euro>. Frozen specimen are kept in two separate liquid nitrogen tanks at two different sites on campus. When necessary, the costs of revitalization from cryopreserved sperm are about <euro>.*

  *Electricity costs for the -80° freezers present in the labs are included in general lab costs. Data storage and backup costs are included in general lab costs.*


### Responsibilities
This section contains the following questions:

- Who will be responsible for the data documentation & metadata?
  *Metadata will be documented by the research and technical staff at the time of data collection and analysis, by taking careful notes in the electronic laboratory notebook (E-notebook) that refer to specific datasets.*

- Who will be responsible for data storage & back up during the project?
  *The research and technical staff will ensure data storage and back up, with support from René Custers and Alexander Botzki for the electronic laboratory notebook (ELN) and from Raf De Coster for the KU Leuven drives.*

- Who will be responsible for ensuring data preservation and sharing?
  *The PI is responsible for data preservation and sharing, with support from the research and technical staff involved in the project, from René Custers and Alexander Botzki for the electronic laboratory notebook (ELN) and from Raf De Coster for the KU Leuven drives.*

- Who bears the end responsibility for updating & implementing this DMP?
  *The PI is ultimately responsible for all data management during and after data collection, including implementing and updating the DMP.*


### Ready to submit?
Before submitting this form, please make sure that you have read and edited all the defaults answers as required for your project. In particular, pay attention to the following points:

- Your answers should be consistent throughout the document. For example: if you mention personal data when listing your different datasets, do not overlook the corresponding ethics question on tab 9 , etc...
- All the text fields marked with <> should be filled in. For example: <euro> in the budget-related questions on tabs 12, 13 and 14.
- Make all answers as specific as possible. For example: if you are collecting or generating particular data types that require specific metadata, or will be deposited in a specific repository, specify it.


If you require further assistance, please contact:

- For the VIB-KU Leuven Center for Brain and Disease Research: Elsa.Lauwers@kuleuven.vib.be


## Data Collection

By now you will have obtained some idea of what research data management is all about. Now we will have a more in-depth look into the different phases of your research by starting with data collection.

Data collection involves understanding the different types of data you collect. Depending on the nature of your research, there are different methods of collecting data and thus different types of data.

Your data may be physical (paper records or archival forms) or digital (database contents or Excel data). The source of your data may be external, you collect it yourself or you generate it from a machine.

When you write your data management plan you will need to take into account the type of data you collect, the source of the data, and how you will process and analyse your data.

You can watch the video below, provided by TU Delft, about data collection.

<iframe src="https://www.youtube.com/embed/AqnVrnVdv2Y" allowfullscreen="" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" height="515px" style="display: inline-block;" width="800px" title=""></iframe>

Having a clear view of what data you will generate will enable you to plan its management. You can create an overview of the data you produce or collect by drawing the data in a workflow, or noting down in a table.

### The Data Flow
Describing your data can be done in several ways. Please watch the video below. Tessa Pronk will explain to you how to describe your data.

<iframe src="https://www.youtube.com/embed/KE2UpZY4wYA" allowfullscreen="" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" height="515px" style="display: inline-block;" width="800px" title=""></iframe>

#### Exercise Order Elements in Your Data Flow

<iframe src="https://elearning.bits.vib.be/wp-admin/admin-ajax.php?action=h5p_embed&id=16" width="958" height="920" frameborder="0" allowfullscreen="allowfullscreen"></iframe><script src="https://elearning.bits.vib.be/wp-content/plugins/h5p/h5p-php-library/js/h5p-resizer.js" charset="UTF-8"></script>

### Discover data
Watch the screencast below. In this screencast, you will be guided through different ways to find data.

<iframe src="https://www.youtube.com/embed/AZMUKgM8X-A" allowfullscreen="" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" height="515px" style="display: inline-block;" width="800px" title=""></iframe>

You have just learned that there are different places to find data. By actively searching the different places, you will get an understanding of the differences. Look at the different portals below. Some of them have been showcased in the screencast, some of them are additional.

- [Google](http://www.google.be) - add "database OR registry OR dataset OR archive OR statistics" to your search
- [Registry of Research Data Repositories re3data](https://www.re3data.org/) - find appropriate repositories holding interesting data
- [Elsevier Data Search](https://datasearch.elsevier.com/#/) - try out chip-seq drosophila
- [Google Dataset Search](https://datasetsearch.research.google.com/) - try out chip-seq drosophila. Google Dataset Search indexes [OmicsDI](https://www.omicsdi.org/), an index providing a knowledge discovery framework across heterogeneous omics data (genomics, proteomics, transcriptomics and metabolomics).   

### Assess the Usefulness of Existing Data

How useful is a dataset? Follow this short tutorial.

<iframe src="https://www.youtube.com/embed/t1SZutbCAxI" allowfullscreen="" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" height="515px" style="display: inline-block;" width="800px" title=""></iframe>

In the previous activity, the lecture described four main points to check if you want to reuse existing data:

- Condition for reuse
- Context
- Trustworthiness
- Persistence

In the following quiz, take a closer look at the description or metadata of some datasets and assess the usefulness of these datasets yourself. As the description or metadata of datasets can be lacking in several different areas at the same time, it will be indicated per assignment on which of the four main points your focus should be.

<iframe src="https://elearning.bits.vib.be/wp-admin/admin-ajax.php?action=h5p_embed&id=15" width="958" height="217" frameborder="0" allowfullscreen="allowfullscreen"></iframe><script src="https://elearning.bits.vib.be/wp-content/plugins/h5p/h5p-php-library/js/h5p-resizer.js" charset="UTF-8"></script>

## Copyright and Intellectual Property Rights (IPR) Issues

### About Data Ownership
Copyright is a form of intellectual property right which arises automatically if an original work is created. Copyright may affect the way data may be stored, shared and reused. You should ask yourself who the copyright holder of your datasets is, especially when you use existing data or when you collaborate with external parties.

Here's a [brief guide to determining what consent is necessary to reuse someone else's data](../../docs/01_Brief_Guide_To_Determining_Consent_To_Reuse_Data.pdf).

Officially VIB, as your employer, is considered the rights holder to the research data you create. You, as a researcher, have the primary responsibility for taking care of the data. Questions on data exploitation may be even more important than those of ownership. Who can use the data? Who can publish it? Who can provide it to third parties?  

We strongly recommend that you deal with the issues around data exploitation at an early stage of your research project. Write down agreements between yourself, your supervisor, project members and other interested parties in your Data Management Plan.

**Additional Resources**

- [Legal instruments and agreements](https://www.uu.nl/en/research/research-data-management/guides/legal-instruments-and-agreements)

### Confidential or Privacy-Sensitive Data
When your research project has received data under confidentiality or under legal privacy restrictions, you will have to identify and explain how you will deal with these restrictions in your data management plan (also see '[Learning Unit: Handle - Data security](../handle-security/tutorial.html)').

## Licences

### About Licences
In order to publish your data and make it reusable, you require a licence. A licence creates clarity and certainty for potential users of your data. A licence is not an option for all data; some of it may be too confidential or privacy-sensitive to be published.

Licences such as the [Creative Commons](https://creativecommons.org/share-your-work/licensing-types-examples/) (CC) licences replace 'all rights reserved' copyright with 'some rights reserved'. There are seven standard CC licences. CC-BY is the most commonly used licence, in which attribution is mandatory when using data. You can also choose restrictions like non-commercial, no derivatives, or share alike. Creative Commons offers a [guide](https://creativecommons.org/choose/?lang=en) to help you determine your preferred licence.

![Creative Commons](../../images/CC.png "Creative Commons")

In the last chapter of this course, you will learn more about licensing.

## Costs Involved with Managing Your Data

### Introduction

The costs of data management and sharing activities must be included into your research, in terms of time and resources needed.
The University of Utrecht has a very valuable overview on their [web site](https://www.uu.nl/en/research/research-data-management/guides/costs-of-data-management) about what to consider for estimating the costs of data management following around the data life cycle. 

### Data Management Cost Guide

When you plan your research you may not be able to oversee all costs involved. Nevertheless, it is useful to have an idea of possible costs at an early stage. You can use the Guide [Costs Overview](../../docs/01_UU_CostOverview.docx), which is a practical overview of possible costs per activity within each phase of the research process. Note: The Cost Guide offers cost indications and examples. These are not real prices.

#### Budget Your Data Management Costs

You are advised to budget the data management costs as separate data management costs. These costs are eligible for funding with funders like FWO and the European Commission, as long as the costs are invoiced before the end of the project.

#### Planning Can Save Time and Money

Planning an early start for certain activities within your research project can lower the costs for data management in the run of your project. You can save time by:

- Properly describing your data while collecting it, instead of doing it afterwards
- Choosing the right file format so that file conversion afterwards is not necessary
- Hiring an experienced data manager

Spending time to think about data activities beforehand can help prevent unexpected extra efforts and costs later on in your research project.

#### Check Current and Expected Costs for Your Research Data

You have just learned that in many parts of a research project there are data related costs. These costs depend on the type and volume of data you produce, analyse and store.
