<!doctype html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="google-site-verification" content="dw_1zBYRNxF7-vYNDXh3Ght_JuqgJt3BD1N_Pynqmv0" />
    <title>Introduction to Genome Assembly</title>
    <script async="" src="https://www.google-analytics.com/analytics.js"></script>
    <script>
        (function(i, s, o, g, r, a, m) {
            i['GoogleAnalyticsObject'] = r;
            i[r] = i[r] || function() {
                (i[r].q = i[r].q || []).push(arguments)
            }, i[r].l = 1 * new Date();
            a = s.createElement(o),
                m = s.getElementsByTagName(o)[0];
            a.async = 1;
            a.src = g;
            m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');
        ga('create', 'UA-61194136-18', 'auto');
        ga('send', 'pageview');
    </script>
    <link rel="stylesheet" href="/assets/css/bootstrap.min.css?v=3">
    <link rel="stylesheet" href="/assets/css/bootstrap-toc.min.css">
    <link rel="stylesheet" href="/assets/css/font-awesome.css">
    <link rel="stylesheet" href="/assets/css/main.scss">
    <link rel="stylesheet" href="/assets/css/VIB.css">
    <link rel="stylesheet" href="/assets/css/academicons.css">
    <link rel="stylesheet" href="/assets/css/syntax_highlighting.css">
    <link rel="shortcut icon" href="/favicon.ico " type="image/x-icon" />     
    <meta name="description" content="Collection of tutorials developed and maintained by the V..." />
    <meta property="og:title" content="VIB Bioinformatics Training: Introduction to Genome Assembly" />
    <meta property="og:description" content="Collection of tutorials developed and maintained by the V..." />
    <meta property="og:image" content="/assets/images/bioinformatics_core_rgb_neg.png" />
</head>

<body data-spy="scroll" data-target="#toc">
    <div class="top-bar-vib">
 <a href="http://vib.be/en" target="_blank" style="font-size:14px">VIB Home </a>
</div>

<div class="container main-content">

<div class="container page-top__container">
          <div class="region region-top">
  <div id="logo" class="col-md-3 col-lg-3 text-left" style="padding-left:0px;margin-top:20px">
       <a href="/">
           <img src="/assets/images/logo.svg" height="60" alt="VIB logo" class="margin20">
       </a>
    </div>
    <div class="col-md-9 col-lg-9 text-right" style="margin-top:50px;">
    <nav role="navigation" aria-labelledby="block-mainnavigation-menu" id="block-mainnavigation" class="block block-menu navigation menu--main">
  


        <a href="/" class="topmenu">Home</a>


        <a href="/about.html" class="topmenu" >About</a>


        <a href="/hall-of-fame.html" class="topmenu">Contribute</a>


        <a href="/faq.html" class="topmenu">Faq</a>


        <a href="/search/index.html" class="topmenu">Search</a>


  </nav>

    </div>
  </div>
</div>














    <script type="application/ld+json">
        


{}
    </script>

    <div class="row tutorial">
      <div class="col-md-12" role="main">
        <h1 data-toc-skip>Introduction to Genome Assembly</h1>
        <div class="contributors-line">By: 

<a href="/hall-of-fame#abotzki" class="contributor-badge">Alexander Botzki</a>


</div>
        <blockquote class="overview">

            <strong><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</strong>
            <ul>
            
            <li>How do we perform a very basic genome assembly from short read data?</li>
            
            </ul>

            <strong><i class="fa fa-bullseye" aria-hidden="true"></i><span class="visually-hidden">objectives</span> Objectives</strong>
            <ul>
            
            <li>assemble some paired end reads using Velvet</li>
            
            <li>examine the output of the assembly.</li>
            
            </ul>

            

            
            <p><strong><i class="fa fa-hourglass-end" aria-hidden="true"></i><span class="visually-hidden">time</span> Time estimation:</strong> 30 minutes</p>
            

            


            
            <div>
                <div>
                     <strong>
                        <i class="fa fa-external-link" aria-hidden="true"></i> 
                             Supporting Materials
                     </strong>
                </div>
                <div id="supporting_materials">
                    <ul>
                        <li>
                            <div>
                    
                    <div>
                        <a class="topic-icon" href="/topics//tutorials/general-introduction/slides.html" title="Slides for this tutorial">
                            <i class="fa fa-slideshare" aria-hidden="true"></i><span class="visually-hidden">slides</span> Slides
                        </a>
                    </div>
                    

                    
                    <div>
                        
<a class="topic-icon" href="https://doi.org/10.5281/zenodo.582600">
    <i class="fa fa-files-o" aria-hidden="true"></i><span class="visually-hidden">zenodo_link</span> Datasets
</a>


                    </div>
                    

                    
                    <div>
                        
    <a class="topic-icon" href="https://github.com/vibbits/training-material/tree/master/topics//tutorials/general-introduction/workflows/" title="Workflows" alt="Introduction to Genome Assembly workflows">
        <i class="fa fa-share-alt" aria-hidden="true"></i><span class="visually-hidden">workflow</span> Workflows
    </a>


                    </div>
                    
                    
                    <div>
                        
<a class="topic-icon" href="https://github.com/vibbits/training-material/tree/master/topics//tutorials/general-introduction/tours/">
    <i class="fa fa-magic" aria-hidden="true"></i><span class="visually-hidden">interactive_tour</span> Tours
</a>


                    </div>
                    


                    
                            </div>
                        </li>
                    </ul>
                </div>

            </div>
            

        </blockquote>

        <div class="container" data-field="body">
            <div class="row">
                <div class="col-md-10" role="main">
                    <h1 class="no_toc" id="genome-assembly-with-velvet-background">Genome assembly with Velvet: Background</h1>

<p>Velvet is one of a number of <em>de novo</em> assemblers that use short read sets as input (<em>e.g.</em> Illumina Reads). The assembly method is based on the manipulation of de Bruijn graphs, via the removal of errors and the simplication of repeated regions.</p>

<blockquote class="comment">
  <h3 id="comment-comment"><i class="fa fa-commenting-o" aria-hidden="true"></i><span class="visually-hidden">comment</span> Comment</h3>

  <p>For information about Velvet, you can check its (nice) <a href="https://en.wikipedia.org/wiki/Velvet_assembler">Wikipedia page</a>.</p>
</blockquote>

<p>For this tutorial, we have a set of reads from an imaginary <em>Staphylococcus aureus</em> bacterium with a miniature genome (197,394 bp). Our mutant strain read set was sequenced with the whole genome shotgun method, using an Illumina DNA sequencing instrument. From these reads, we would like to rebuild our imaginary <em>Staphylococcus aureus</em> bacterium via a <em>de novo</em> assembly of a short read set using the Velvet assembler.</p>

<blockquote class="agenda">
  <h3 id="agenda">Agenda</h3>

  <p>In this tutorial, we will deal with:</p>

<ol id="markdown-toc">
  <li><a href="#get-the-data" id="markdown-toc-get-the-data">Get the data</a></li>
  <li><a href="#evaluate-the-input-reads" id="markdown-toc-evaluate-the-input-reads">Evaluate the input reads</a></li>
  <li><a href="#assemble-reads-with-velvet" id="markdown-toc-assemble-reads-with-velvet">Assemble reads with Velvet</a></li>
  <li><a href="#collect-some-statistics-on-the-contigs" id="markdown-toc-collect-some-statistics-on-the-contigs">Collect some statistics on the contigs</a></li>
  <li><a href="#discussion" id="markdown-toc-discussion">Discussion</a></li>
</ol>

</blockquote>

<h1 id="get-the-data">Get the data</h1>

<p>We will now import the data that we will use for the tutorial.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-getting-the-data"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Getting the data</h3>

  <ol>
    <li>Create and name a new history for this tutorial.</li>
    <li>Import from <a href="https://doi.org/10.5281/zenodo.582600">Zenodo</a> or from the data library the files:
      <ul>
        <li><a href="https://zenodo.org/record/582600/files/mutant_R1.fastq"><code class="language-plaintext highlighter-rouge">mutant_R1.fastq</code></a></li>
        <li><a href="https://zenodo.org/record/582600/files/mutant_R2.fastq"><code class="language-plaintext highlighter-rouge">mutant_R2.fastq</code></a></li>
      </ul>

      <blockquote class="tip">
        <h3 id="tip-tip-importing-data-via-links"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Importing data via links</h3>

        <ul>
          <li>Copy the link location (Right-click on the filename then “Copy Link Address”)</li>
          <li>Open the Galaxy Upload Manager</li>
          <li>Select <strong>Paste/Fetch Data</strong></li>
          <li>Paste the link into the text field</li>
          <li>Change the data-type to <strong>fastqsanger</strong></li>
          <li>Press <strong>Start</strong></li>
        </ul>
      </blockquote>
    </li>
    <li>
      <p>Change the name of the files to <code class="language-plaintext highlighter-rouge">mutant_R1</code> and <code class="language-plaintext highlighter-rouge">mutant_R2</code>.</p>

      <p>As a default, Galaxy uses the link as the name of the new dataset. It also does not link the dataset to a database or a reference genome.</p>

      <blockquote class="tip">

        <h3 id="tip-tip-renaming-a-dataset"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Renaming a dataset</h3>
        <ul>
          <li>Click on the <i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">galaxy-pencil</span> <strong>pencil icon</strong> for the dataset to edit its attributes</li>
          <li>In the central panel, change the <strong>Name</strong> field</li>
          <li>Click the <strong>Save</strong> button</li>
        </ul>
      </blockquote>
    </li>
    <li>
      <p>Inspect the content of a file.</p>

      <blockquote class="tip">
        <h3 id="tip-tip-inspecting-the-content-of-a-dataset"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Inspecting the content of a dataset</h3>

        <ul>
          <li>Click on the <i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">galaxy-eye</span> (eye) icon next to the relevant history entry</li>
          <li>View the content of the file in the central panel</li>
        </ul>
      </blockquote>

      <blockquote class="question">
        <h3 id="question-questions"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</h3>

        <ol>
          <li>What are four key features of a FASTQ file?</li>
          <li>What is the main difference between a FASTQ and a FASTA file?</li>
        </ol>

        <blockquote>
          <details>
            <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
            <ol>
              <li>Each sequence in a FASTQ file is represented by 4 lines: 1st line is the id, 2nd line is the sequence, 3rd line is not used, and 4th line is the quality of sequencing per nucleotide</li>
              <li>In a FASTQ file, not only are the sequences present, but information about the quality of sequencing is also included.</li>
            </ol>
          </details>
        </blockquote>

      </blockquote>
    </li>
  </ol>

</blockquote>

<p>The reads have been sequenced from an imaginary <em>Staphylococcus aureus</em> bacterium using an Illumina DNA sequencing instrument. We obtained the 2 files we imported (<code class="language-plaintext highlighter-rouge">mutant_R1</code> and <code class="language-plaintext highlighter-rouge">mutant_R2</code>)</p>

<blockquote class="question">
  <h3 id="question-question"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

  <p>Why do we have 2 files here if we only sequenced the bacteria once?</p>

  <blockquote>
    <details>
      <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
      <ol>
        <li>The bacteria has been sequenced using paired-end sequencing. The first file corresponds to forward reads and the second file to reverse reads.</li>
      </ol>
    </details>
  </blockquote>

</blockquote>

<h1 id="evaluate-the-input-reads">Evaluate the input reads</h1>

<p>Before doing any assembly, the first questions you should ask about your input reads include:</p>

<ul>
  <li>What is the coverage of my genome?</li>
  <li>How good is my read set?</li>
  <li>Do I need to ask for a new sequencing run?</li>
  <li>Is it suitable for the analysis I need to do?</li>
</ul>

<p>We will evaluate the input reads using the FastQC tool. This tool runs a standard series of tests on your read set and returns a relatively easy-to-interpret report. We will use it to evaluate the quality of our FASTQ files and combine the results with MultiQC.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-fastqc-on-a-fastq-file"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: FastQC on a fastq file</h3>

  <ol>
    <li><strong>FastQC</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with the following parameters
      <ul>
        <li>“Short read data from your current history” to (<strong>Multiple datasets</strong>) <code class="language-plaintext highlighter-rouge">mutant_R1.fastq</code> and <code class="language-plaintext highlighter-rouge">mutant_R2.fastq</code></li>
      </ul>
    </li>
    <li><strong>MultiQC</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with the following parameters
      <ul>
        <li>“Software name” to <code class="language-plaintext highlighter-rouge">FastQC</code></li>
        <li>“Result file” to the raw data files generated by FastQC</li>
      </ul>
    </li>
  </ol>

</blockquote>

<p>MultiQC generates a webpage combining reports for FastQC on both datasets. It includes these graphs and tables:</p>

<ul>
  <li>
    <p>General statistics</p>

    <p>This is important in setting maximum k-mer size for an assembly.</p>

    <blockquote class="comment">
      <h3 id="comment-getting-the-length-of-sequences"><i class="fa fa-commenting-o" aria-hidden="true"></i><span class="visually-hidden">comment</span> Getting the length of sequences</h3>

      <ul>
        <li>Click on <strong>Configure Columns</strong></li>
        <li>Check <strong>Length</strong></li>
        <li>Close the window</li>
      </ul>
    </blockquote>

    <blockquote class="question">
      <h3 id="question-questions-1"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</h3>

      <ol>
        <li>How long are the sequences?</li>
        <li>What is the average coverage of the genome, given our imaginary <em>Staphylococcus aureus</em> bacterium has a genome of 197,394 bp?</li>
      </ol>

      <blockquote>
        <details>
          <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
          <ol>
            <li>The sequences are 150 bp long</li>
            <li>We have 2 x 12,480 sequences of 150 bp, so the average genome coverage is: 2 * 12480 * 150 / 197394, or approximately 19 X coverage.</li>
          </ol>
        </details>
      </blockquote>

    </blockquote>
  </li>
  <li>
    <p>Sequence Quality Histograms</p>

    <p>Dips in quality near the beginning, middle or end of the reads may determine the trimming/cleanup methods and parameters to be used, or may indicate technical problems with the sequencing process/machine run.</p>

    <figure id="figure-1"><img src="../../images/fastqc_per_base_sequence_quality_plot.png" alt="Sequence Quality Histograms with the mean quality value across each base position in the read" /><figcaption><span class="figcaption-prefix">Figure 1:</span> The mean quality value across each base position in the read</figcaption></figure>

    <blockquote class="question">
      <h3 id="question-questions-2"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</h3>

      <ol>
        <li>What does the y-axis represent?</li>
        <li>Why is the quality score decreasing across the length of the reads?</li>
      </ol>

      <blockquote>
        <details>
          <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
          <ol>
            <li>The y-axis represents the quality score for each base (an estimate of the error during sequencing).</li>
            <li>The quality score is decreasing accross the length of the reads because the sequencing become less and less reliable at the end of the reads.</li>
          </ol>
        </details>
      </blockquote>

    </blockquote>
  </li>
  <li>
    <p>Per Sequence GC Content</p>

    <p>High GC organisms tend not to assemble well and may have an uneven read coverage distribution.</p>
  </li>
  <li>
    <p>Per Base N Content</p>

    <p>The presence of large numbers of Ns in reads may point to a poor quality sequencing run. You will need to trim these reads to remove Ns.</p>
  </li>
  <li>
    <p>k-mer content</p>

    <p>The presence of highly recurring k-mers may point to contamination of reads with barcodes or adapter sequences.</p>
  </li>
</ul>

<blockquote class="comment">
  <h3 id="comment-comment-1"><i class="fa fa-commenting-o" aria-hidden="true"></i><span class="visually-hidden">comment</span> Comment</h3>

  <p>For a fuller discussion of FastQC outputs and warnings, see the <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC website link</a>, including the section on each of the output <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/">reports</a>, and examples of <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/good_sequence_short_fastqc.html">“good”</a> and <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/bad_sequence_fastqc.html">“bad”</a> Illumina data.</p>
</blockquote>

<p>We won’t be doing anything to these data to clean it up as there isn’t much need. Therefore we will get on with the assembly!</p>

<h1 id="assemble-reads-with-velvet">Assemble reads with Velvet</h1>

<p>Now, we want to assemble our reads to find the sequence of our imaginary <em>Staphylococcus aureus</em> bacterium. We will perform a <em>de novo</em> assembly of the reads into long contiguous sequences using the Velvet short read assembler.</p>

<p>The first step of the assembler is to build a de Bruijn graph. For that, it will break our reads into k-mers, <em>i.e.</em> fragments of length <em>k</em>. Velvet requires the user to input a value of <em>k</em> (k-mer size) for the assembly process. Small k-mers will give greater connectivity, but large k-mers will give better specificity.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-assemble-the-reads"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Assemble the reads</h3>

  <ol>
    <li><strong>FASTQ interlacer</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with the following parameters
      <ul>
        <li>“Type of paired-end datasets” to <code class="language-plaintext highlighter-rouge">2 separate datasets</code></li>
        <li>“Left-hand mates” to <code class="language-plaintext highlighter-rouge">mutant_R1.fastq</code></li>
        <li>“Right-hand mates” to <code class="language-plaintext highlighter-rouge">mutant_R2.fastq</code></li>
      </ul>

      <p>Currently our paired-end reads are in 2 files (one with the forward reads and one with the reverse reads), but Velvet requires only one file, where each read is next to its mate read. In other words, if the reads are indexed from 0, then reads 0 and 1 are paired, 2 and 3, 4 and 5, etc. Before doing the assembly <em>per se</em>, we need to prepare the files by combining them.</p>
    </li>
    <li><strong>velveth</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with the following parameters
      <ul>
        <li>“Hash Length” to <code class="language-plaintext highlighter-rouge">29</code></li>
        <li>“Input Files”: click on <code class="language-plaintext highlighter-rouge">Insert Input Files</code></li>
        <li>“file format” to <code class="language-plaintext highlighter-rouge">fastq</code></li>
        <li>“read type” to <code class="language-plaintext highlighter-rouge">shortPaired reads</code></li>
        <li>“Dataset” to the pairs output of <strong>FASTQ interlacer</strong></li>
      </ul>

      <p>The tool takes our reads and break them into k-mers.</p>
    </li>
    <li><strong>velvetg</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with the following parameters
      <ul>
        <li>“Velvet Dataset” to the output of <strong>velveth</strong></li>
        <li>“Using Paired Reads” to <code class="language-plaintext highlighter-rouge">Yes</code></li>
      </ul>

      <p>This last tool actually does the assembly.</p>
    </li>
  </ol>
</blockquote>

<p>Two files are generated:</p>

<ul>
  <li>
    <p>A “Contigs” file</p>

    <p>This file contains the sequences of the contigs longer than 2k. In the header of each contig, a bit of information is added:</p>
    <ul>
      <li>the k-mer length (called “length”): For the value of k chosen in the assembly, a measure of how many k-mers overlap (by 1 bp each overlap) to give this length</li>
      <li>the k-mer coverage (called “coverage”): For the value of k chosen in the assembly, a measure of how many k-mers overlap each base position (in the assembly).</li>
    </ul>

    <p><img src="../../images/image10.png" alt="Contigs output" /></p>
  </li>
  <li>
    <p>A “Stats” file</p>

    <p>This is a tabular file giving for each contig the k-mer lengths, k-mer coverages and other measures.</p>

    <p><img src="../../images/image11.png" alt="Contigs stats output" /></p>
  </li>
</ul>

<h1 id="collect-some-statistics-on-the-contigs">Collect some statistics on the contigs</h1>

<blockquote class="question">
  <h3 id="question-question-1"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

  <ol>
    <li>How many contigs have been built?</li>
    <li>What is the mean, min and max length of the contigs?</li>
  </ol>

  <blockquote>
    <details>
      <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
      <ol>
        <li>190</li>
        <li>To compute this information, we can use the Datamash tool on the 2nd columns (length). Be careful with the first line, the header. As a result, we obtain: 597.82 as mean, 1 as min and 12904 as max. It would mean that the smallest contig has a length of 1 bp, even smaller than k. The length on the 2nd column corresponds to length of the contig in k-mers. This means that the smallest contig has a length of 1k = 29. So to obtain the real length, we need to add k-1 to the length. We then obtain a mean contig length of 625.82 bp, a min contig of 29 bp and a max contig of 12,932 bp.</li>
      </ol>
    </details>
  </blockquote>

</blockquote>

<p>This table is limitted, but we will now collect more basic statistics on our assembly.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-collect-fasta-statistics-on-our-contigs"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Collect fasta statistics on our contigs</h3>

  <ol>
    <li><strong>Quast</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with
      <ul>
        <li>“Contigs/scaffolds output file” to the output of <strong>velvetg</strong></li>
        <li>“Type of data” to <code class="language-plaintext highlighter-rouge">contig</code></li>
        <li>“Reference File” to <code class="language-plaintext highlighter-rouge">wildtype.fna</code></li>
        <li>“Type of organism” to <code class="language-plaintext highlighter-rouge">Prokaryotes</code></li>
        <li>“Lower Threshold” to <code class="language-plaintext highlighter-rouge">500</code></li>
        <li>“Thresholds” to <code class="language-plaintext highlighter-rouge">0,1000</code></li>
      </ul>
    </li>
  </ol>
</blockquote>

<p>This tool generates 5 output files, but we will focus on the HTML report and the Icarus viewer.</p>

<blockquote class="question">
  <h3 id="question-question-2"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

  <ol>
    <li>What is represented in the Icarus viewer?</li>
  </ol>

  <blockquote>
    <details>
      <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
      <ol>
        <li>Icarus is a novel genome visualizer for accurate assessment and analysis of genomic draft assemblies. It draws contigs ordered from longest to shortest, highlights N50, N75 (NG50, NG75) and long contigs larger than a user-specified threshold</li>
      </ol>
    </details>
  </blockquote>

</blockquote>

<p>The HTML report reports many statistics computed by QUAST to assess the quality of the assembly:</p>

<ul>
  <li>Statistics about the quality of the assembly when compared to the reference (fraction of the genome, duplication ratio, etc)</li>
  <li>
    <p>Misassembly statistics, including the number of misassemblies</p>

    <p>A misassembly is a position in the contigs (breakpoints) that satisfy one of the following criteria:</p>
    <ul>
      <li>the left flanking sequence aligns over 1 kbp away from the right flanking sequence on the reference;</li>
      <li>flanking sequences overlap on more than 1 kbp</li>
      <li>flanking sequences align to different strands or different chromosomes</li>
    </ul>
  </li>
  <li>Unaligned regions in the assembly</li>
  <li>Mismatches compared to the reference genomes</li>
  <li>Statistics about the assembly <em>per se</em>, such as the number of contigs and the length of the largest contig</li>
</ul>

<blockquote class="question">
  <h3 id="question-question-3"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

  <ol>
    <li>How many contigs have been constructed?</li>
    <li>Which proportion of the reference genome do they represent?</li>
    <li>How many misassemblies have been found?</li>
    <li>Has the assembly introduced mismatches and indels?</li>
    <li>What are N50 and L50?</li>
    <li>Is there a bias in GC percentage induced by the assembly?</li>
  </ol>

  <blockquote>
    <details>
      <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
      <ol>
        <li>190 contigs have been constructed, but only 47 have a length &gt; 500 bp.</li>
        <li>The contigs represents 87.965% of the reference genome.</li>
        <li>1 misassembly has been found: it corresponds to a relocation, <em>i.e.</em> a misassembly event (breakpoint) where the left flanking sequence aligns over 1 kbp away from the right flanking sequence on the reference genome.</li>
        <li>8.06 mismatches per 100 kbp and 4.03 indels per 100 kbp are found.</li>
        <li>N50 is the length for which the collection of all contigs of that length or longer covers at least half an assembly. In other words, if contigs were ordered from small to large, half of all the nucleotides will be in contigs this size or larger. And L50 is the number of contigs equal to or longer than N50: L50 is the minimal number of contigs that cover half the assembly.</li>
        <li>The GC % in the assembly is 33.64%, really similar to the one of the reference genome (33.43%).</li>
      </ol>
    </details>
  </blockquote>

</blockquote>

<h1 id="discussion">Discussion</h1>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-optional-hands-on-rerun-for-values-k-ranging-from-31-to-101"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> (Optional) Hands-on: Rerun for values <em>k</em> ranging from 31 to 101</h3>

  <ol>
    <li><strong>velveth</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with the same parameters as before except
      <ul>
        <li>“Hash Length” to a value between 31 and 101</li>
      </ul>
    </li>
    <li><strong>velvetg</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with the same parameters as before</li>
    <li><strong>Quast</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with the same parameters as before</li>
  </ol>
</blockquote>

<p>We have completed an assembly on this data set for a number of k values ranging from 29 to 101. A few of the assembly metrics appear below.</p>

<figure id="figure-2"><img src="../../images/number_of_contigs.png" alt="contigs" /><figcaption><span class="figcaption-prefix">Figure 2:</span> Number of contigs in the assembly for various k-mer sizes</figcaption></figure>

<figure id="figure-3"><img src="../../images/largest_contig.png" alt="largest_contig" /><figcaption><span class="figcaption-prefix">Figure 3:</span> Largest contig in each of the assemblies by k-mer size</figcaption></figure>

<figure id="figure-4"><img src="../../images/total_bp.png" alt="total_bp" /><figcaption><span class="figcaption-prefix">Figure 4:</span> Total number of base pairs in all the contigs for each assembly by k-mer size</figcaption></figure>

<figure id="figure-5"><img src="../../images/n50.png" alt="n50" /><figcaption><span class="figcaption-prefix">Figure 5:</span> N50 metric for each of the assemblies by k-mer size</figcaption></figure>

<blockquote class="question">
  <h3 id="question-questions-3"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</h3>

  <ol>
    <li>Are there any distinct features in the charts?</li>
    <li>Does it look like one assembly might be better than some of the others?</li>
  </ol>

</blockquote>

<p>The reasons for these patterns will be discussed in detail in the De Bruijn graph assembly slides and tutorial.</p>



                    
                    <blockquote class="key_points">
                        <h3><i class="fa fa-key" aria-hidden="true"></i><span class="visually-hidden">keypoints</span> Key points</h3>

                        <ul>
                            
                            <li>We assembled some Illumina fastq reads into contigs using a short read assembler called Velvet</li>
                            
                            <li>We showed what effect one of the key assembly parameters, the k-mer size, has on the assembly</li>
                            
                            <li>It looks as though there are some exploitable patterns in the metric data vs the k-mer size.</li>
                            
                        </ul>
                    </blockquote>
                    

                    

                    

                </div>
                <!-- sidebar, which will be hidden on a small screen -->
                <nav class="col-md-2 sticky-toc bs-docs-sidebar hidden-print hidden-xs hidden-sm" role="complementary" id="toc" data-toggle="toc"></nav>
            </div>
        </div>

        <h3><i class="fa fa-thumbs-up" aria-hidden="true"></i><span class="visually-hidden">congratulations</span> Congratulations on successfully completing this tutorial!</h3>

        

        

</div>
</div>
</div>
<footer>
   <div class="well row">
       <div class="col-md-2 text-center" style="padding: 25px 20px 25px 20px;">				
         <a href="https://www.bits.vib.be" target="_blank"><img src="/assets/images/bioinformatics_core_rgb_neg.png" width="200"></a>
       </div>
       <div class="col-md-8 text-center" style="padding-top: 20px;">	
        <a href="https://github.com/vibbits/training-material" title="View on GitHub" class="margin20" target="_blank">
                            <i class="fa fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> View on GitHub
                        </a>
       </div>
       <div class="col-md-2 text-center" style="padding-top: 30px;">
         <span class="slogan"><br>SCIENCE MEETS LIFE</span>
       </div>
   </div>
</footer>


</body>
<script type="text/javascript" src="/assets/js/jquery.slim.min.js"></script>
<script type="text/javascript" src="/assets/js/popper.min.js"></script>
<script type="text/javascript" src="/assets/js/bootstrap.min.js?v=3"></script>
<script type="text/javascript" src="/assets/js/details-element-polyfill.js"></script>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script type="text/javascript" src="/assets/js/bootstrap-toc.min.js"></script>
<script type="text/javascript" src="/assets/js/main.js"></script>
<script type="text/javascript" src="/assets/js/clipboard.min.js"></script>
<script type="text/javascript">
    var snippets = document.querySelectorAll('div.highlight');
    [].forEach.call(snippets, function(snippet) {
        snippet.firstChild.insertAdjacentHTML('beforebegin', '<button class="btn btn-light" data-clipboard-snippet><i class="fa fa-copy"></i>&nbsp;Copy</button>');
    });

    var clipboardSnippets = new ClipboardJS('[data-clipboard-snippet]', {
        target: function(trigger) {
            return trigger.nextElementSibling;
        }
    });
</script>

</html>
