<!doctype html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="x-ua-compatible" content="ie=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
        <meta name="google-site-verification" content="hA2THcUT50TVfGM3VQppZL3e4jBG3a2lo2lqN5k2pys" />
        <title>05 Predicting a protein structure - homology modelling</title>
        
        <link rel="stylesheet" href="/assets/css/bootstrap.min.css?v=3">
        <link rel="stylesheet" href="/assets/css/bootstrap-toc.min.css">
        <link rel="stylesheet" href="/assets/css/font-awesome.css">
        <link rel="stylesheet" href="/assets/css/VIB.css">
        <link rel="stylesheet" href="/assets/css/academicons.css">
        <link rel="stylesheet" href="/assets/css/syntax_highlighting.css">
        <link rel="shortcut icon" href="/favicon.ico" type="image/x-icon" />

        
        
        
        
        
        <meta name="description" content="Analyzing the protein structure of your protein-of-intere..." />
        <meta property="og:title" content="VIB Bioinformatics Training: 05 Predicting a protein structure - homology modelling" />
        <meta property="og:description" content="Analyzing the protein structure of your protein-of-intere..." />
        <meta property="og:image" content="/assets/images/bioinformatics_core_rgb_neg.png" />
    </head>
    <body data-spy="scroll" data-target="#toc">
        <div class="top-bar-vib">
 <a href="http://vib.be/en" target="_blank" style="font-size:14px">VIB Home </a>
</div>

<div class="container main-content">

<div class="container page-top__container">
          <div class="region region-top">
  <div id="logo" class="col-md-3 col-lg-3 text-left" style="padding-left:0px;margin-top:20px">
       <a href="/">
           <img src="/assets/images/logo.svg" height="60" alt="VIB logo" class="margin20">
       </a>
    </div>
    <div class="col-md-9 col-lg-9 text-right" style="margin-top:50px;">
    <nav role="navigation" aria-labelledby="block-mainnavigation-menu" id="block-mainnavigation" class="block block-menu navigation menu--main">
  


        <a href="/" id="topmenu" class="is-active">Home</a>


        <a href="/about.html" id="topmenu" >About</a>


        <a href="/hall-of-fame.html" id="topmenu">Contribute</a>


        <a href="/faq.html" id="topmenu">Faq</a>


        <a href="/search.html" id="topmenu">Search</a>


  </nav>

    </div>
  </div>
</div>














    <script type="application/ld+json">
        


{
  "@context": "http://schema.org",
  "@type": "Course",
  "accessMode": [
    "textual",
    "visual"
  ],
  "accessModeSufficient": [
    "textual",
    "visual"
  ],
  "accessibilityControl": [
    "fullKeyboardControl",
    "fullMouseControl"
  ],
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "accessibilitySummary": "Short descriptions are present but long descriptions will be needed for non-visual users",
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "copyrightHolder": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "headline": "05 Predicting a protein structure - homology modelling",
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "license": "/blob//LICENSE.md",
  "producer": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "provider": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "sourceOrganization": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "courseCode": "protein-structure-analysis / homology-modeling / hands-on",
  "learningResourceType": "hands-on tutorial",
  "name": "Hands-on for '05 Predicting a protein structure - homology modelling' tutorial",
  "isPartOf": {
    "@type": "Course",
    "name": "05 Predicting a protein structure - homology modelling",
    "description": "05 Predicting a protein structure - homology modelling",
    "learningResourceType": "tutorial",
    "interactivityType": "expositive",
    "provider": {
      "@type": "Organization",
      "email": "bits@vib.be",
      "name": "VIB Bioinformatics Core",
      "url": "https://www.bits.vib.be"
    }
  },
  "url": "https://material.bits.vib.be//topics/protein-structure-analysis/tutorials/homology-modeling/tutorial.html",
  "timeRequired": "PT30M",
  "description": "05 Predicting a protein structure - homology modelling\\nThe questions this  addresses are:\n - Which biological questions are addressed by the tutorial?\n - Which bioinformatics techniques are important to know for this type of data?\n\n\\nThe objectives are:\n - The learning objectives are the goals of the tutorial\n - They will be informed by your audience and will communicate to them and to yourself what you should focus on during the course\n - They are single sentences describing what a learner should be able to do once they have completed the tutorial\n - You can use Bloom's Taxonomy to write effective learning objectives\n\n",
  "hasPart": [

  ],
  "mentions": [
    {
      "@type": "Thing",
      "url": "",
      "name": "Training data for 05 Predicting a protein structure - homology modelling tutorial"
    }
  ],
  "author": [
    {
      "@type": "Person",
      "name": "Alexander Botzki"
    },
    {
      "@type": "Person",
      "name": "Joost Van Durme"
    }
  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "Alexander Botzki"
    },
    {
      "@type": "Person",
      "name": "Joost Van Durme"
    }
  ],
  "about": [
    {
      "@type": "CreativeWork",
      "name": "Protein Structure Analysis",
      "description": "Analyzing the protein structure of your protein-of-interest can be advantageous in multiple ways. It can help you discover regions which are good candidates to interact with other proteins. It can help you discover new domains. It can help with identifying differences with homologuous proteins and a lot more.",
      "url": "https://material.bits.vib.be//topics/protein-structure-analysis/"
    }
  ]
}
    </script>

    <div class="row tutorial">
      <div class="col-md-12" role="main">
        <h1 data-toc-skip>05 Predicting a protein structure - homology modelling</h1>
        <div class="contributors-line">By: 

<a href="/hall-of-fame#abotzki" class="contributor-badge">Alexander Botzki</a>
, <a href="/hall-of-fame#jvdurme" class="contributor-badge">Joost Van Durme</a>


</div>
        <blockquote class="overview">

            <strong><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</strong>
            <ul>
            
            <li>Which biological questions are addressed by the tutorial?</li>
            
            <li>Which bioinformatics techniques are important to know for this type of data?</li>
            
            </ul>

            <strong><i class="fa fa-bullseye" aria-hidden="true"></i><span class="visually-hidden">objectives</span> Objectives</strong>
            <ul>
            
            <li>The learning objectives are the goals of the tutorial</li>
            
            <li>They will be informed by your audience and will communicate to them and to yourself what you should focus on during the course</li>
            
            <li>They are single sentences describing what a learner should be able to do once they have completed the tutorial</li>
            
            <li>You can use Bloom's Taxonomy to write effective learning objectives</li>
            
            </ul>

            

            
            <p><strong><i class="fa fa-hourglass-end" aria-hidden="true"></i><span class="visually-hidden">time</span> Time estimation:</strong> 30 minutes</p>
            

            



            <div>
                <div><strong><i class="fa fa-external-link" aria-hidden="true"></i> Supporting Materials</strong></div>
                <div id="supporting_materials">
                    <ul>
                        <li>
                            <div>
                    
                    <div>
                        <a class="topic-icon" href="/topics/protein-structure-analysis/slides/introduction.html" title="Introduction slides">
                            <i class="fa fa-slideshare" aria-hidden="true"></i><span class="visually-hidden">slides</span> Introduction Slides
                        </a>
                    </div>
                    

                    

                    
                    


                    
                            </div>
                        </li>
                    </ul>
                </div>

            </div>

        </blockquote>

        <div class="container">
            <div class="row">
                <div class="col-sm-10" role="main">
                    <h1 class="no_toc" id="introduction">Introduction</h1>

<!-- This is a comment. -->

<p>The goal of homology modeling is to predict the 3D structure of a protein that comes close to what would be achieved experimentally with X-Ray experiments.</p>

<p>Main principles of homology modeling</p>

<ul>
  <li>We predict the structure of a protein sequence on the basis of the structure of another protein with a similar sequence (the template)</li>
  <li>If the sequences are similar, the structures will have a similar fold</li>
  <li>Structure is more conserved than sequence</li>
</ul>

<blockquote class="agenda">
  <h3 id="agenda">Agenda</h3>

  <p>In this tutorial, we will cover:</p>

<ol id="markdown-toc">
  <li><a href="#main-ingredients-for-homology-modelling" id="markdown-toc-main-ingredients-for-homology-modelling">Main ingredients for homology modelling</a>    <ol>
      <li><a href="#the-sequence" id="markdown-toc-the-sequence">The sequence</a></li>
      <li><a href="#searching-for-the-template-structure" id="markdown-toc-searching-for-the-template-structure">Searching for the template structure</a></li>
      <li><a href="#aligning-target-and-template-sequence-and-template-selection" id="markdown-toc-aligning-target-and-template-sequence-and-template-selection">Aligning target and template sequence and template selection</a></li>
    </ol>
  </li>
  <li><a href="#building-the-homology-model-with-swiss-model" id="markdown-toc-building-the-homology-model-with-swiss-model">Building the homology model with Swiss Model</a></li>
</ol>

</blockquote>

<h1 id="main-ingredients-for-homology-modelling">Main ingredients for homology modelling</h1>

<h2 id="the-sequence">The sequence</h2>

<p>Last week my colleague sequenced a plant protein. He is not a bioinformatician. Yet, he would like to know what the structure might look like to do some rounds of rational mutagenesis. Let’s try to address the problem for him.</p>

<p>He came up with this sequence:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>SVCCPSLVARTNYNVCRLPGTEAALCATFTGCIIIPGATCGGDYAN
</code></pre></div></div>

<h2 id="searching-for-the-template-structure">Searching for the template structure</h2>

<p>Actually, the first step is to check whether the PDB already contains the structure of this sequence. That would be easy so we don’t have to model anything. We will use Blast again to search with the sequence.</p>

<p><strong>TODO: add hands-on area with Blast search on PDB page</strong></p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-data-upload"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Data upload</h3>

  <ol>
    <li>Create a new history for this tutorial</li>
    <li>
      <p>Import the files from <a href="">Zenodo</a> or from the shared data library</p>

      <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>
</code></pre></div>      </div>
      <p><strong><em>TODO</em></strong>: <em>Add the files by the ones on Zenodo here (if not added)</em></p>

      <p><strong><em>TODO</em></strong>: <em>Remove the useless files (if added)</em></p>

      <blockquote class="tip">

        <h3 id="tip-tip-importing-data-via-links"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Importing data via links</h3>

        <ul>
          <li>Copy the link location</li>
          <li>
            <p>Open the Galaxy Upload Manager (<i class="fa fa fa-upload" aria-hidden="true"></i><span class="visually-hidden">galaxy-upload</span> on the top-right of the tool panel)</p>
          </li>
          <li>Select <strong>Paste/Fetch Data</strong></li>
          <li>
            <p>Paste the link into the text field</p>
          </li>
          <li>
            <p>Press <strong>Start</strong></p>
          </li>
          <li><strong>Close</strong> the window</li>
        </ul>

        <p>By default, Galaxy uses the URL as the name, so rename the files with a more useful name.</p>
      </blockquote>

      <blockquote class="tip">

        <h3 id="tip-tip-importing-data-from-a-data-library"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Importing data from a data library</h3>

        <p>As an alternative to uploading the data from a URL or your computer, the files may also have been made available from a <em>shared data library</em>:</p>

        <ul>
          <li>
            <p>Go into <strong>Shared data</strong> (top panel) then <strong>Data libraries</strong></p>
          </li>
          <li>
            <p>Find the correct folder (ask your instructor)</p>
          </li>
          <li>Select the desired files</li>
          <li>Click on the <strong>To History</strong> button near the top and select <strong>as Datasets</strong> from the dropdown menu</li>
          <li>In the pop-up window, select the history you want to import the files to (or create a new one)</li>
          <li>Click on <strong>Import</strong></li>
        </ul>
      </blockquote>
    </li>
    <li>Rename the datasets</li>
    <li>
      <p>Check that the datatype</p>

      <blockquote class="tip">

        <h3 id="tip-tip-changing-the-datatype"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Changing the datatype</h3>
        <ul>
          <li>Click on the <i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">galaxy-pencil</span> <strong>pencil icon</strong> for the dataset to edit its attributes</li>
          <li>In the central panel, click on the <i class="fa fa fa-database" aria-hidden="true"></i><span class="visually-hidden">galaxy-chart-select-data</span> <strong>Datatypes</strong> tab on the top</li>
          <li>Select <code class="highlighter-rouge">datatypes</code></li>
          <li>Click the <strong>Change datatype</strong> button</li>
        </ul>
      </blockquote>
    </li>
    <li>
      <p>Add to each database a tag corresponding to …</p>

      <blockquote class="tip">

        <h3 id="tip-tip-adding-a-tag"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Adding a tag</h3>
        <ul>
          <li>Click on the dataset</li>
          <li>Click on <i class="fa fa-tags" aria-hidden="true"></i><span class="visually-hidden">galaxy-tags</span> <strong>Edit dataset tags</strong></li>
          <li>
            <p>Add a tag starting with <code class="highlighter-rouge">#</code></p>

            <p>Tags starting with <code class="highlighter-rouge">#</code> will be automatically propagated to the outputs of tools using this dataset.</p>
          </li>
          <li>Check that the tag is appearing below the dataset name</li>
        </ul>

      </blockquote>
    </li>
  </ol>

</blockquote>

<p>A suitable template structure to make a high quality model should have following properties:</p>

<ul>
  <li>The highest possible sequence identity from all structures in the PDB when aligned to the target sequence</li>
  <li>A good resolution (and R-factor): if many identical template structures exist with the same sequence, filter by resolution</li>
  <li>Is solved by X-RAY, not NMR</li>
</ul>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-data-upload-1"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Data upload</h3>

  <ol>
    <li>Create a new history for this tutorial</li>
    <li>
      <p>Import the files from <a href="">Zenodo</a> or from the shared data library</p>

      <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>
</code></pre></div>      </div>
      <p><strong><em>TODO</em></strong>: <em>Add the files by the ones on Zenodo here (if not added)</em></p>

      <p><strong><em>TODO</em></strong>: <em>Remove the useless files (if added)</em></p>

      <blockquote class="tip">

        <h3 id="tip-tip-importing-data-via-links-1"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Importing data via links</h3>

        <ul>
          <li>Copy the link location</li>
          <li>
            <p>Open the Galaxy Upload Manager (<i class="fa fa fa-upload" aria-hidden="true"></i><span class="visually-hidden">galaxy-upload</span> on the top-right of the tool panel)</p>
          </li>
          <li>Select <strong>Paste/Fetch Data</strong></li>
          <li>
            <p>Paste the link into the text field</p>
          </li>
          <li>
            <p>Press <strong>Start</strong></p>
          </li>
          <li><strong>Close</strong> the window</li>
        </ul>

        <p>By default, Galaxy uses the URL as the name, so rename the files with a more useful name.</p>
      </blockquote>

      <blockquote class="tip">

        <h3 id="tip-tip-importing-data-from-a-data-library-1"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Importing data from a data library</h3>

        <p>As an alternative to uploading the data from a URL or your computer, the files may also have been made available from a <em>shared data library</em>:</p>

        <ul>
          <li>
            <p>Go into <strong>Shared data</strong> (top panel) then <strong>Data libraries</strong></p>
          </li>
          <li>
            <p>Find the correct folder (ask your instructor)</p>
          </li>
          <li>Select the desired files</li>
          <li>Click on the <strong>To History</strong> button near the top and select <strong>as Datasets</strong> from the dropdown menu</li>
          <li>In the pop-up window, select the history you want to import the files to (or create a new one)</li>
          <li>Click on <strong>Import</strong></li>
        </ul>
      </blockquote>
    </li>
    <li>Rename the datasets</li>
    <li>
      <p>Check that the datatype</p>

      <blockquote class="tip">

        <h3 id="tip-tip-changing-the-datatype-1"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Changing the datatype</h3>
        <ul>
          <li>Click on the <i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">galaxy-pencil</span> <strong>pencil icon</strong> for the dataset to edit its attributes</li>
          <li>In the central panel, click on the <i class="fa fa fa-database" aria-hidden="true"></i><span class="visually-hidden">galaxy-chart-select-data</span> <strong>Datatypes</strong> tab on the top</li>
          <li>Select <code class="highlighter-rouge">datatypes</code></li>
          <li>Click the <strong>Change datatype</strong> button</li>
        </ul>
      </blockquote>
    </li>
    <li>
      <p>Add to each database a tag corresponding to …</p>

      <blockquote class="tip">

        <h3 id="tip-tip-adding-a-tag-1"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Adding a tag</h3>
        <ul>
          <li>Click on the dataset</li>
          <li>Click on <i class="fa fa-tags" aria-hidden="true"></i><span class="visually-hidden">galaxy-tags</span> <strong>Edit dataset tags</strong></li>
          <li>
            <p>Add a tag starting with <code class="highlighter-rouge">#</code></p>

            <p>Tags starting with <code class="highlighter-rouge">#</code> will be automatically propagated to the outputs of tools using this dataset.</p>
          </li>
          <li>Check that the tag is appearing below the dataset name</li>
        </ul>

      </blockquote>
    </li>
  </ol>

</blockquote>

<blockquote class="question">
  <h3 id="question-questions"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</h3>

  <ol>
    <li>Is there a difference in the number of identities, positives and gaps between the two remaining x-ray structures?</li>
    <li>What is the PDB ID with the highest resolution, does not have insertions or deletions and should thus be the better template structure?</li>
  </ol>

  <blockquote class="solution">
    <h3 id="solution-solution"><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h3>

    <ol>
      <li><strong>TODO</strong></li>
      <li><strong>TODO</strong></li>
    </ol>

  </blockquote>

</blockquote>

<h2 id="aligning-target-and-template-sequence-and-template-selection">Aligning target and template sequence and template selection</h2>

<p>The alignment is the most crucial part of homology modeling. We will not explain what an alignment is and how you make it, this should be known. In an alignment, we put homologous sequences on top of each other in a text file. The point is that amino acids that are on top of each other in the same column are assumed to have the equivalent position in the 3D structure. So if the template sequence has an Ala at position 3, where the target sequence has a Phe, homology modelling tools will use the backbone of the template structure and replace the sidechain at position 3 from Ala to Phe.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-data-upload-2"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Data upload</h3>

  <ol>
    <li>Create a new history for this tutorial</li>
    <li>
      <p>Import the files from <a href="">Zenodo</a> or from the shared data library</p>

      <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>
</code></pre></div>      </div>
      <p><strong><em>TODO</em></strong>: <em>Add the files by the ones on Zenodo here (if not added)</em></p>

      <p><strong><em>TODO</em></strong>: <em>Remove the useless files (if added)</em></p>

      <blockquote class="tip">

        <h3 id="tip-tip-importing-data-via-links-2"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Importing data via links</h3>

        <ul>
          <li>Copy the link location</li>
          <li>
            <p>Open the Galaxy Upload Manager (<i class="fa fa fa-upload" aria-hidden="true"></i><span class="visually-hidden">galaxy-upload</span> on the top-right of the tool panel)</p>
          </li>
          <li>Select <strong>Paste/Fetch Data</strong></li>
          <li>
            <p>Paste the link into the text field</p>
          </li>
          <li>
            <p>Press <strong>Start</strong></p>
          </li>
          <li><strong>Close</strong> the window</li>
        </ul>

        <p>By default, Galaxy uses the URL as the name, so rename the files with a more useful name.</p>
      </blockquote>

      <blockquote class="tip">

        <h3 id="tip-tip-importing-data-from-a-data-library-2"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Importing data from a data library</h3>

        <p>As an alternative to uploading the data from a URL or your computer, the files may also have been made available from a <em>shared data library</em>:</p>

        <ul>
          <li>
            <p>Go into <strong>Shared data</strong> (top panel) then <strong>Data libraries</strong></p>
          </li>
          <li>
            <p>Find the correct folder (ask your instructor)</p>
          </li>
          <li>Select the desired files</li>
          <li>Click on the <strong>To History</strong> button near the top and select <strong>as Datasets</strong> from the dropdown menu</li>
          <li>In the pop-up window, select the history you want to import the files to (or create a new one)</li>
          <li>Click on <strong>Import</strong></li>
        </ul>
      </blockquote>
    </li>
    <li>Rename the datasets</li>
    <li>
      <p>Check that the datatype</p>

      <blockquote class="tip">

        <h3 id="tip-tip-changing-the-datatype-2"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Changing the datatype</h3>
        <ul>
          <li>Click on the <i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">galaxy-pencil</span> <strong>pencil icon</strong> for the dataset to edit its attributes</li>
          <li>In the central panel, click on the <i class="fa fa fa-database" aria-hidden="true"></i><span class="visually-hidden">galaxy-chart-select-data</span> <strong>Datatypes</strong> tab on the top</li>
          <li>Select <code class="highlighter-rouge">datatypes</code></li>
          <li>Click the <strong>Change datatype</strong> button</li>
        </ul>
      </blockquote>
    </li>
    <li>
      <p>Add to each database a tag corresponding to …</p>

      <blockquote class="tip">

        <h3 id="tip-tip-adding-a-tag-2"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Adding a tag</h3>
        <ul>
          <li>Click on the dataset</li>
          <li>Click on <i class="fa fa-tags" aria-hidden="true"></i><span class="visually-hidden">galaxy-tags</span> <strong>Edit dataset tags</strong></li>
          <li>
            <p>Add a tag starting with <code class="highlighter-rouge">#</code></p>

            <p>Tags starting with <code class="highlighter-rouge">#</code> will be automatically propagated to the outputs of tools using this dataset.</p>
          </li>
          <li>Check that the tag is appearing below the dataset name</li>
        </ul>

      </blockquote>
    </li>
  </ol>

</blockquote>

<p>To extract the sequence from the template structure, display the FASTA sequence of the template structure and paste it into the text editor on the first line.</p>

<figure id="figure-1"><img src="../../images/image_name" alt="Alternative text" /><figcaption><span class="figcaption-prefix">Figure 1:</span> Legend of the image</figcaption></figure>

<p>The idea is to keep the theory description before quite simple to focus more on the practical part.</p>

<figure id="figure-2"><img src="../../images/image_name" alt="Alternative text" /><figcaption><span class="figcaption-prefix">Figure 2:</span> Legend of the image</figcaption></figure>

<h1 id="building-the-homology-model-with-swiss-model">Building the homology model with Swiss Model</h1>

<p>Our current request for homology modelling is a rather safe one, so we can use an automatic server for homology modelling. There are many automatic tools available and many of them compete in regular competitions like lastly, the 12th Community Wide Experiment on the Critical Assessment of Techniques for Protein Structure Prediction (CASP12) - [1].</p>

<p>In our example, we take the <a href="https://swissmodel.expasy.org/interactive">Swiss Model server</a>. SWISS-MODEL is a fully automated protein structure homology-modelling server, accessible via the ExPASy web server, or from the program DeepView (Swiss Pdb-Viewer). The purpose of this server is to make Protein Modelling accessible to all biochemists and molecular biologists worldwide.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-template-selection-step-with-swiss-model"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Template selection step with Swiss Model</h3>

  <ol>
    <li>Browse to the <a href="https://swissmodel.expasy.org/interactive">Swiss Model server</a></li>
    <li>On the first page, paste the sequence of our unknown protein in the field ‘Target Sequence’ and give the project a name.</li>
  </ol>
  <figure id="figure-3"><img src="../../images/Modelling_sequence_template_step1.png" alt="Swiss Model Start page" /><figcaption><span class="figcaption-prefix">Figure 3:</span> Start page of Swiss Model</figcaption></figure>
  <ol>
    <li>Click ‘Search templates’ to initiate the first step. 
Thereafter, the server identifies structural template(s) and gives an overview list of hits 
which you can select the templates from.</li>
  </ol>

</blockquote>

<blockquote class="question">
  <h3 id="question-question"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

  <p>Which of the 10 (at the time of writing) possible template structures would you select as template for the model building process?</p>

  <blockquote class="solution">
    <h3 id="solution-solution-1"><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h3>

    <p><strong>TODO: add link with 10 template</strong> 
We suggest as template <strong>1jxx.1.A</strong> given that it is an X-ray structure with high resolution and a very high 
sequence identity (X-ray, 0.9 Å, 78.26 %).</p>

  </blockquote>

</blockquote>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-model-building-step-and-visualisation"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Model Building Step and Visualisation</h3>

  <ol>
    <li>
      <p>Once you have selected the template, hit ‘Build Model’ to start the homology modelling procedure. 
The server will alignment of target sequence and template structure(s), build a model and evaluate it. 
These steps require specialized software and integrate up-to-date protein sequence and structure databases. 
Each of the above steps can be repeated interactively until a satisfying modelling result is achieved.</p>
    </li>
    <li>Once the model has been built, you can download it.</li>
    <li>If the Swiss Model server is too busy at the moment you execute the request, you can download the model from
<a href="https://zenodo.org/record/3551850#.Xdqs4ehKiUk">here</a>.</li>
    <li>Load the created model into YASARA. 
Perform a structural superposition with your reference e.g. 1CRN 
try to detect the differences through manipulating the visualisations.
      <figure id="figure-4"><img src=" " alt="superposition" /><figcaption><span class="figcaption-prefix">Figure 4:</span> Superposition of 1CRN with the obtained model</figcaption></figure>
    </li>
  </ol>

</blockquote>

<h1 class="no_toc" id="conclusion">Conclusion</h1>

<p>Homology modelling evolved over the years and many online tools for homology modelling are available. You have used the Swiss Model service with a reasonable simple modelling request. Often, in research projects, homology modelling can be rather difficult and needs expert knowledge depending on the actual situation (sequence conservation, available templates, etc.).</p>



                    
                    <blockquote class="key_points">
                        <h3><i class="fa fa-key" aria-hidden="true"></i><span class="visually-hidden">keypoints</span> Key points</h3>

                        <ul>
                            
                            <li>Main principles of homology modelling</li>
                            
                            <li>How to build a homology model with Swiss Model</li>
                            
                        </ul>
                    </blockquote>
                    

                    
                    <h1 data-toc-skip>Useful literature</h1>
                    <p>Further information, including links to documentation and original publications, regarding the tools, analysis techniques and the interpretation of results described in this tutorial can be found <a href="/topics/protein-structure-analysis#references">here</a>.</p>
                    

                    

                </div>
                <!-- sidebar, which will move to the top on a small screen -->
                <div class="col-sm-2" role="supplementary">
                    <nav id="toc" data-toggle="toc" class="sticky-top bs-docs-sidebar hidden-print hidden-xs hidden-sm"></nav>
                </div>
            </div>
        </div>

        <h3><i class="fa fa-thumbs-up" aria-hidden="true"></i><span class="visually-hidden">congratulations</span> Congratulations on successfully completing this tutorial!</h3>

        

        

</div>
</div>
</div>
<footer>
   <div class="well row">
       <div class="col-md-2 text-center" style="padding: 25px 20px 25px 20px;">				
         <a href="https://www.bits.vib.be" target="_blank"><img src="/assets/images/bioinformatics_core_rgb_neg.png" width="200"></a>
       </div>
       <div class="col-md-8 text-center" style="padding-top: 20px;">	
        <a href="https://github.com/vibbits/training-material" title="View on GitHub" class="margin20" target="_blank">
                            <i class="fa fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> View on GitHub
                        </a>
       </div>
       <div class="col-md-2 text-center" style="padding-top: 30px;">
         <span class="slogan"><br>SCIENCE MEETS LIFE</span>
       </div>
   </div>
</footer>


    </body>
    <script type="text/javascript" src="/assets/js/jquery.slim.min.js"></script>
    <script type="text/javascript" src="/assets/js/popper.min.js"></script>
    <script type="text/javascript" src="/assets/js/bootstrap.min.js?v=3"></script>
    <script type="text/javascript" src="/assets/js/details-element-polyfill.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="/assets/js/bootstrap-toc.min.js"></script>
    <script type="text/javascript" src="/assets/js/main.js"></script>
    <script type="text/javascript" src="/assets/js/clipboard.min.js"></script>
    <script type="text/javascript" src="https://app.cloudsh.com/js/cloudsh.js"></script>
    <script type="text/javascript">
    var snippets=document.querySelectorAll('div.highlight');
    [].forEach.call(snippets,function(snippet){
        snippet.firstChild.insertAdjacentHTML('beforebegin','<button class="btn btn-light" data-clipboard-snippet><i class="fa fa-copy"></i>&nbsp;Copy</button>');
    });

    var clipboardSnippets=new ClipboardJS('[data-clipboard-snippet]',{
        target:function(trigger){return trigger.nextElementSibling;
    }});
    </script>
</html>
