<!doctype html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="google-site-verification" content="dw_1zBYRNxF7-vYNDXh3Ght_JuqgJt3BD1N_Pynqmv0" />
    <title>05 Predicting a protein structure - homology modelling</title>
    <script async="" src="https://www.google-analytics.com/analytics.js"></script>
    <script>
        (function(i, s, o, g, r, a, m) {
            i['GoogleAnalyticsObject'] = r;
            i[r] = i[r] || function() {
                (i[r].q = i[r].q || []).push(arguments)
            }, i[r].l = 1 * new Date();
            a = s.createElement(o),
                m = s.getElementsByTagName(o)[0];
            a.async = 1;
            a.src = g;
            m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');
        ga('create', 'UA-61194136-18', 'auto');
        ga('send', 'pageview');
    </script>
    <link rel="stylesheet" href="/assets/css/bootstrap.min.css?v=3">
    <link rel="stylesheet" href="/assets/css/bootstrap-toc.min.css">
    <link rel="stylesheet" href="/assets/css/font-awesome.css">
    <link rel="stylesheet" href="/assets/css/main.scss">
    <link rel="stylesheet" href="/assets/css/VIB.css">
    <link rel="stylesheet" href="/assets/css/academicons.css">
    <link rel="stylesheet" href="/assets/css/syntax_highlighting.css">
    <link rel="shortcut icon" href="/favicon.ico " type="image/x-icon" />     
    <meta name="description" content="Analyzing the protein structure of your protein-of-intere..." />
    <meta property="og:title" content="VIB Bioinformatics Training: 05 Predicting a protein structure - homology modelling" />
    <meta property="og:description" content="Analyzing the protein structure of your protein-of-intere..." />
    <meta property="og:image" content="/assets/images/bioinformatics_core_rgb_neg.png" />
</head>

<body data-spy="scroll" data-target="#toc">
    <div class="top-bar-vib">
 <a href="http://vib.be/en" target="_blank" style="font-size:14px">VIB Home </a>
</div>

<div class="container main-content">

<div class="container page-top__container">
          <div class="region region-top">
  <div id="logo" class="col-md-3 col-lg-3 text-left" style="padding-left:0px;margin-top:20px">
       <a href="/">
           <img src="/assets/images/logo.svg" height="60" alt="VIB logo" class="margin20">
       </a>
    </div>
    <div class="col-md-9 col-lg-9 text-right" style="margin-top:50px;">
    <nav role="navigation" aria-labelledby="block-mainnavigation-menu" id="block-mainnavigation" class="block block-menu navigation menu--main">
  


        <a href="/" class="topmenu">Home</a>


        <a href="/about.html" class="topmenu" >About</a>


        <a href="/hall-of-fame.html" class="topmenu">Contribute</a>


        <a href="/faq.html" class="topmenu">Faq</a>


        <a href="/search/index.html" class="topmenu">Search</a>


  </nav>

    </div>
  </div>
</div>














    <script type="application/ld+json">
        


{
  "@context": "http://schema.org",
  "@type": "Course",
  "accessMode": [
    "textual",
    "visual"
  ],
  "accessModeSufficient": [
    "textual",
    "visual"
  ],
  "accessibilityControl": [
    "fullKeyboardControl",
    "fullMouseControl"
  ],
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "accessibilitySummary": "Short descriptions are present but long descriptions will be needed for non-visual users",
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "copyrightHolder": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "headline": "05 Predicting a protein structure - homology modelling",
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "license": "/blob//LICENSE.md",
  "producer": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "provider": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "sourceOrganization": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "courseCode": "protein-structure-analysis / homology-modeling / hands-on",
  "learningResourceType": "hands-on tutorial",
  "name": "Hands-on for '05 Predicting a protein structure - homology modelling' tutorial",
  "isPartOf": {
    "@type": "Course",
    "name": "05 Predicting a protein structure - homology modelling",
    "description": "05 Predicting a protein structure - homology modelling",
    "learningResourceType": "tutorial",
    "interactivityType": "expositive",
    "provider": {
      "@type": "Organization",
      "email": "bits@vib.be",
      "name": "VIB Bioinformatics Core",
      "url": "https://www.bits.vib.be"
    }
  },
  "url": "https://material.bits.vib.be//topics/protein-structure-analysis/tutorials/homology-modeling/tutorial.html",
  "timeRequired": "PT30M",
  "description": "05 Predicting a protein structure - homology modelling\\nThe questions this  addresses are:\n - Which are the main principles of homology modelling?\n - Which highly automated tool can be used to do 3-click homology modelling?\n\n\\nThe objectives are:\n - Understand the importance of sequence alignment and careful template selection as key ingredients for homology modelling.\n - Homology modelling can be a tedious task but Swiss Model allows for high automation of homology modelling tasks.\n\n",
  "hasPart": [

  ],
  "mentions": [
    {
      "@type": "Thing",
      "url": "",
      "name": "Training data for 05 Predicting a protein structure - homology modelling tutorial"
    }
  ],
  "author": [
    {
      "@type": "Person",
      "name": "Alexander Botzki"
    },
    {
      "@type": "Person",
      "name": "Joost Van Durme"
    }
  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "Alexander Botzki"
    },
    {
      "@type": "Person",
      "name": "Joost Van Durme"
    }
  ],
  "about": [
    {
      "@type": "CreativeWork",
      "name": "Protein Structure Analysis",
      "description": "Analyzing the protein structure of your protein-of-interest can be advantageous in multiple ways. It can help you discover regions which are good candidates to interact with other proteins. It can help you discover new domains. It can help with identifying differences with homologuous proteins and a lot more.",
      "url": "https://material.bits.vib.be//topics/protein-structure-analysis/"
    }
  ]
}
    </script>

    <div class="row tutorial">
      <div class="col-md-12" role="main">
        <h1 data-toc-skip>05 Predicting a protein structure - homology modelling</h1>
        <div class="contributors-line">By: 

<a href="/hall-of-fame#abotzki" class="contributor-badge">Alexander Botzki</a>
, <a href="/hall-of-fame#jvdurme" class="contributor-badge">Joost Van Durme</a>


</div>
        <blockquote class="overview">

            <strong><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</strong>
            <ul>
            
            <li>Which are the main principles of homology modelling?</li>
            
            <li>Which highly automated tool can be used to do 3-click homology modelling?</li>
            
            </ul>

            <strong><i class="fa fa-bullseye" aria-hidden="true"></i><span class="visually-hidden">objectives</span> Objectives</strong>
            <ul>
            
            <li>Understand the importance of sequence alignment and careful template selection as key ingredients for homology modelling.</li>
            
            <li>Homology modelling can be a tedious task but Swiss Model allows for high automation of homology modelling tasks.</li>
            
            </ul>

            

            
            <p><strong><i class="fa fa-hourglass-end" aria-hidden="true"></i><span class="visually-hidden">time</span> Time estimation:</strong> 30 minutes</p>
            

            


            

        </blockquote>

        <div class="container" data-field="body">
            <div class="row">
                <div class="col-md-10" role="main">
                    <h2 class="no_toc" id="introduction">Introduction</h2>

<!-- This is a comment. -->

<p>The goal of homology modeling is to predict the 3D structure of a protein that comes close to what would be achieved experimentally with X-Ray experiments.</p>

<p>Main principles of homology modeling</p>

<ul>
  <li>We predict the structure of a protein sequence on the basis of the structure of another protein with a similar sequence (the template)</li>
  <li>If the sequences are similar, the structures will have a similar fold</li>
  <li>Structure is more conserved than sequence</li>
</ul>

<h2 id="main-ingredients-for-homology-modelling">Main ingredients for homology modelling</h2>

<h3 id="the-sequence">The sequence</h3>

<p>Last week my colleague sequenced a plant protein. He is not a bioinformatician. Yet, he would like to know what the structure might look like to do some rounds of rational mutagenesis. Letâ€™s try to address the problem for him.</p>

<p>He came up with this sequence:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>SVCCPSLVARTNYNVCRLPGTEAALCATFTGCIIIPGATCGGDYAN
</code></pre></div></div>

<h3 id="searching-for-the-template-structure">Searching for the template structure</h3>

<p>Actually, the first step is to check whether the PDB already contains the structure of this sequence. That would be easy so we donâ€™t have to model anything. We will use Blast again to search with the sequence.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-blast-search-for-pdb-structure"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: BLAST search for PDB structure</h3>

  <ol>
    <li>Go to the Advanced Search section</li>
    <li>Please select â€˜Sequence BLAST/PSI-BLASTâ€™ in the Query type drop down.
This method allows you to change some parameters for the search.</li>
    <li>Copy and paste the sequence in the â€˜â€˜Sequenceâ€™â€™ field</li>
    <li>Press â€˜â€˜Submit queryâ€™â€™.</li>
    <li>You should see the same structures popping up as you saw in the UniProt page of DnaK.</li>
  </ol>
</blockquote>

<p>A suitable template structure to make a high quality model should have following properties:</p>

<ul>
  <li>The highest possible sequence identity from all structures in the PDB when aligned to the target sequence</li>
  <li>A good resolution (and R-factor): if many identical template structures exist with the same sequence, filter by resolution</li>
  <li>Is solved by X-RAY, not NMR.</li>
</ul>

<blockquote class="question">
  <h3 id="question-questions"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</h3>

  <ol>
    <li>Is there a difference in the number of identities, positives and gaps between the two remaining x-ray structures?</li>
    <li>What is the PDB ID with the highest resolution, does not have insertions or deletions and should thus be the better template structure?</li>
  </ol>

  <details>
    <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>

    <ol>
      <li><strong>TODO</strong></li>
      <li><strong>TODO</strong></li>
    </ol>

  </details>

</blockquote>

<h2 id="aligning-target-and-template-sequence-and-template-selection">Aligning target and template sequence and template selection</h2>

<p>The alignment is the most crucial part of homology modeling. We will not explain what an alignment is and how you make it, this should be known. In an alignment, we put homologous sequences on top of each other in a text file. The point is that amino acids that are on top of each other in the same column are assumed to have the equivalent position in the 3D structure. So if the template sequence has an Ala at position 3, where the target sequence has a Phe, homology modelling tools will use the backbone of the template structure and replace the sidechain at position 3 from Ala to Phe.</p>

<p>Homology modelling evolved over the years and many online tools for homology modelling are available. In my experience, homology modelling can be rather difficult and needs expert knowledge depending on the actual situation (sequence conservation, available templates, etc.).</p>

<p>Can you imagine what could be the reasons?</p>

<h1 id="building-the-homology-model-with-swiss-model">Building the homology model with Swiss Model</h1>

<p>Our current request for homology modelling is a rather safe one, so we can use an automatic server for homology modelling. There are many automatic tools available and many of them compete in regular competitions like lastly, the 12th Community Wide Experiment on the Critical Assessment of Techniques for Protein Structure Prediction (CASP12) - [1].</p>

<p>In our example, we take the <a href="https://swissmodel.expasy.org/interactive">Swiss Model server</a>. SWISS-MODEL is a fully automated protein structure homology-modelling server, accessible via the ExPASy web server, or from the program DeepView (Swiss Pdb-Viewer). The purpose of this server is to make Protein Modelling accessible to all biochemists and molecular biologists worldwide.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-template-selection-step-with-swiss-model"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Template selection step with Swiss Model</h3>

  <ol>
    <li>Browse to the <a href="https://swissmodel.expasy.org/interactive">Swiss Model server</a></li>
    <li>On the first page, paste the sequence of our unknown protein in the field â€˜Target Sequenceâ€™ and give the project a name.
      <figure id="figure-1"><img src="../../images/Modelling_sequence_template_step1.png" alt="Swiss Model Start page -80width" /><figcaption><span class="figcaption-prefix">Figure 1:</span> Start page of Swiss Model</figcaption></figure>
    </li>
    <li>Click â€˜Search templatesâ€™ to initiate the first step. 
Thereafter, the server identifies structural template(s) and gives an overview list of hits 
which you can select the templates from.</li>
  </ol>

</blockquote>

<blockquote class="question">
  <h3 id="question-question"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

  <p>Which of the 10 (at the time of writing) possible template structures would you select as template for the model building process?</p>

  <details>
    <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>

    <p>We suggest as template <strong>1jxx.1.A</strong> given that it is an X-ray structure with high resolution and a very high 
sequence identity (X-ray, 0.9 Ã…, 78.26 %).</p>
  </details>
</blockquote>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-model-building-step-and-visualisation"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Model Building Step and Visualisation</h3>

  <ol>
    <li>Once you have selected the template, hit â€˜Build Modelâ€™ to start the homology modelling procedure. 
The server will alignment of target sequence and template structure(s), build a model and evaluate it. 
These steps require specialized software and integrate up-to-date protein sequence and structure databases. 
Each of the above steps can be repeated interactively until a satisfying modelling result is achieved. 
<img src="../../images/Modelling_template_selection_step2.png" alt="download model -80width" /></li>
    <li>Once the model has been built, you can download it.
<img src="../../images/Modelling_results_step3.png" alt="download model -80width" /></li>
    <li>If the Swiss Model server is too busy at the moment you execute the request, you can download the model from
<a href="https://zenodo.org/record/3551850#.Xdqs4ehKiUk">here</a>.</li>
    <li>Load the created model into YASARA. 
Perform a structural alignment with your reference e.g. 1CRN and try to detect the differences through manipulating the visualisations.
<img src="../../images/1214.png" alt="structural alignment" /></li>
  </ol>
</blockquote>

<h1 class="no_toc" id="conclusion">Conclusion</h1>

<p>Homology modelling evolved over the years and many online tools for homology modelling are available. You have used the Swiss Model service with a reasonable simple modelling request. Often, in research projects, homology modelling can be rather difficult and needs expert knowledge depending on the actual situation (sequence conservation, available templates, etc.).</p>



                    
                    <blockquote class="key_points">
                        <h3><i class="fa fa-key" aria-hidden="true"></i><span class="visually-hidden">keypoints</span> Key points</h3>

                        <ul>
                            
                            <li>Main principles of homology modelling</li>
                            
                            <li>How to build a homology model with Swiss Model</li>
                            
                        </ul>
                    </blockquote>
                    

                    
                    <h1 data-toc-skip>Useful literature</h1>
                    <p>Further information, including links to documentation and original publications, regarding the tools, analysis techniques and the interpretation of results described in this tutorial can be found <a href="/topics/protein-structure-analysis#references">here</a>.</p>
                    

                    

                </div>
                <!-- sidebar, which will be hidden on a small screen -->
                <nav class="col-md-2 sticky-toc bs-docs-sidebar hidden-print hidden-xs hidden-sm" role="complementary" id="toc" data-toggle="toc"></nav>
            </div>
        </div>

        <h3><i class="fa fa-thumbs-up" aria-hidden="true"></i><span class="visually-hidden">congratulations</span> Congratulations on successfully completing this tutorial!</h3>

        

        

</div>
</div>
</div>
<footer>
   <div class="well row">
       <div class="col-md-2 text-center" style="padding: 25px 20px 25px 20px;">				
         <a href="https://www.bits.vib.be" target="_blank"><img src="/assets/images/bioinformatics_core_rgb_neg.png" width="200"></a>
       </div>
       <div class="col-md-8 text-center" style="padding-top: 20px;">	
        <a href="https://github.com/vibbits/training-material" title="View on GitHub" class="margin20" target="_blank">
                            <i class="fa fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> View on GitHub
                        </a>
       </div>
       <div class="col-md-2 text-center" style="padding-top: 30px;">
         <span class="slogan"><br>SCIENCE MEETS LIFE</span>
       </div>
   </div>
</footer>


</body>
<script type="text/javascript" src="/assets/js/jquery.slim.min.js"></script>
<script type="text/javascript" src="/assets/js/popper.min.js"></script>
<script type="text/javascript" src="/assets/js/bootstrap.min.js?v=3"></script>
<script type="text/javascript" src="/assets/js/details-element-polyfill.js"></script>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script type="text/javascript" src="/assets/js/bootstrap-toc.min.js"></script>
<script type="text/javascript" src="/assets/js/main.js"></script>
<script type="text/javascript" src="/assets/js/clipboard.min.js"></script>
<script type="text/javascript">
    var snippets = document.querySelectorAll('div.highlight');
    [].forEach.call(snippets, function(snippet) {
        snippet.firstChild.insertAdjacentHTML('beforebegin', '<button class="btn btn-light" data-clipboard-snippet><i class="fa fa-copy"></i>&nbsp;Copy</button>');
    });

    var clipboardSnippets = new ClipboardJS('[data-clipboard-snippet]', {
        target: function(trigger) {
            return trigger.nextElementSibling;
        }
    });
</script>

</html>
