<!doctype html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="x-ua-compatible" content="ie=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
        <meta name="google-site-verification" content="dw_1zBYRNxF7-vYNDXh3Ght_JuqgJt3BD1N_Pynqmv0" />
        <title>02 Experiment design exercises</title>
        
        <link rel="stylesheet" href="/training-material/assets/css/bootstrap.min.css?v=3">
        <link rel="stylesheet" href="/training-material/assets/css/bootstrap-toc.min.css">
        <link rel="stylesheet" href="/training-material/assets/css/main.css?v=2">
        <link rel="stylesheet" href="/training-material/assets/css/font-awesome.css">
        <link rel="stylesheet" href="/training-material/assets/css/VIB.css">
        <link rel="stylesheet" href="/training-material/assets/css/academicons.css">
        <link rel="stylesheet" href="/training-material/assets/css/syntax_highlighting.css">
        <link rel="shortcut icon" href="/training-material/favicon.ico" type="image/x-icon" />

        
        
        
        
        
        <meta name="description" content="QPCR analysis using qbase+" />
        <meta property="og:title" content="VIB Bioinformatics Training: 02 Experiment design exercises" />
        <meta property="og:description" content="QPCR analysis using qbase+" />
        <meta property="og:image" content="/training-material/assets/images/bioinformatics_core_rgb_neg.png" />
    </head>
    <body data-spy="scroll" data-target="#toc">
        <div class="container main-content">













<header class="well row">
           <div id="logo" class="col-md-4 col-lg-2 text-center">
              <a href="/training-material/">
                <img src="/training-material/assets/images/vib_notag_neg_rgb.jpg" height="50" alt="VIB logo" class="margin20">
              </a>
           </div> 
           <div class="col-md-4 col-lg-8" id="top-navbar">
                <ul class="navbar-nav">
                    <li class="nav-item">
                        <a class="nav-link" href="/training-material/topics/qbase-plus" title="Go back to list of tutorials">
                            <i class="fa fa-folder-o" aria-hidden="true"></i><span class="visually-hidden">topic</span> QPCR analysis using qbase+
                        </a>
                    </li>
            </div>
            <div class="col-md-4 col-lg-2 text-right form-group label-floating" id="help">
                       
    <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-expanded="false" title="Help">
        <i class="fa fa-life-ring" aria-hidden="true"></i><span class="visually-hidden">help</span> Help
    </a>
    <div class="dropdown-menu dropdown-menu-right">
        <form method="get" action="https://tess.elixir-europe.org/materials">
            <input type="text" id="search" name="q" value="" style="margin-left: 0.5em;/*! border-radius: 0px; */">
            <input type="hidden" value="VIB Bioinformatics Core" name="content_provider">
            <input type="submit" value="Search on TeSS" style="width: 92%;border-radius: 0px;margin: 0.5em;background: #f47d20;border: 0px;padding: 0.25em;" class="">
        </form>

        <div class="dropdown-divider"></div>
        <a class="dropdown-item" href="https://www.bits.vib.be" title="Contact us">
            Contact us 
        </a>
    </div>


            </div>
</header>

    <script type="application/ld+json">
        


{
  "@context": "http://schema.org",
  "@type": "Course",
  "accessMode": [
    "textual",
    "visual"
  ],
  "accessModeSufficient": [
    "textual",
    "visual"
  ],
  "accessibilityControl": [
    "fullKeyboardControl",
    "fullMouseControl"
  ],
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "accessibilitySummary": "Short descriptions are present but long descriptions will be needed for non-visual users",
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "copyrightHolder": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "headline": "02 Experiment design exercises",
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "license": "/blob//LICENSE.md",
  "producer": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "provider": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "sourceOrganization": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "courseCode": "qbase-plus / experiment-design / hands-on",
  "learningResourceType": "hands-on tutorial",
  "name": "Hands-on for '02 Experiment design exercises' tutorial",
  "isPartOf": {
    "@type": "Course",
    "name": "02 Experiment design exercises",
    "description": "02 Experiment design exercises",
    "learningResourceType": "tutorial",
    "interactivityType": "expositive",
    "provider": {
      "@type": "Organization",
      "email": "bits@vib.be",
      "name": "VIB Bioinformatics Core",
      "url": "https://www.bits.vib.be"
    }
  },
  "url": "https://vibbits.github.io//training-material/topics/qbase-plus/tutorials/experiment-design/tutorial.html",
  "timeRequired": "PT5M",
  "description": "02 Experiment design exercises\\nThe questions this  addresses are:\n - How do you perform a qPCR analysis using the qbase+ software?\n\n\\nThe objectives are:\n - Design qPCR experiments\n - Use the qbase+ software to analyze the qPCR experiment results.\n\n",
  "hasPart": [

  ],
  "mentions": [

  ],
  "author": [
    {
      "@type": "Person",
      "name": "Christof De Bo"
    }
  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "Christof De Bo"
    }
  ],
  "about": [
    {
      "@type": "CreativeWork",
      "name": "QPCR analysis using qbase+",
      "description": "QPCR analysis using qbase+",
      "url": "https://vibbits.github.io//training-material/topics/qbase-plus/"
    }
  ]
}
    </script>

    <div class="row tutorial">
      <div class="col-md-12" role="main">
        <h1 data-toc-skip>02 Experiment design exercises</h1>
        <div class="contributors-line">By: 

<a href="/training-material/hall-of-fame#chdeb" class="contributor-badge"><img src="https://avatars.githubusercontent.com/chdeb" alt="Christof De Bo">Christof De Bo</a>

</div>
        <blockquote class="overview">
            <h3>Overview</h3>
            <strong><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</strong>
            <ul>
            
            <li>How do you perform a qPCR analysis using the qbase+ software?</li>
            
            </ul>

            <strong><i class="fa fa-bullseye" aria-hidden="true"></i><span class="visually-hidden">objectives</span> Objectives</strong>
            <ul>
            
            <li>Design qPCR experiments</li>
            
            <li>Use the qbase+ software to analyze the qPCR experiment results.</li>
            
            </ul>

            

            
            <p><strong><i class="fa fa-hourglass-end" aria-hidden="true"></i><span class="visually-hidden">time</span> Time estimation:</strong> 5 minutes</p>
            

            



            <div>
                <div><strong><i class="fa fa-external-link" aria-hidden="true"></i> Supporting Materials</strong></div>
                <div id="supporting_materials">
                    <ul>
                        <li>
                            <div>
                    

                    

                    
                    


                    
                            </div>
                        </li>
                    </ul>
                </div>

            </div>

        </blockquote>

        <div class="container">
            <div class="row">
                <div class="col-sm-10" role="main">
                    <h3 id="exercise-1-simple-gene-expression-study">Exercise 1: simple gene expression study</h3>

<p>In my qPCR experiment I want to study the expression of 12 genes of interest in 8 samples of interest. I want to use 2 PCR replicates for each reaction.</p>

<blockquote>
  <p>How many 96 well plates do I need for this experiment ?</p>
  <blockquote>
    <p>I have 12 genes in 8 samples which gives a total of 96 reactions (one plate). I want to perform each reaction twice (2 PCR replicates) so I need two plates. However, I need to include reference genes in my experiment, preferably more than one. I can put these reference genes on a separate plate, I do not have to include them on each plate.
Ideally, you need to include 3 reference genes so having 8 samples and 2 replicates this gives an additional 48 reactions. Thus, I need three 96 well plates to perform this experiment.</p>
  </blockquote>
</blockquote>

<table>
  <thead>
    <tr>
      <th style="text-align: left">Do I need to include IRCs (inter-run calibrators) ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">No, I can easily fit all samples of the same gene on the same plate so I don’t need to include IRCs.</td>
    </tr>
  </tbody>
</table>

<h3 id="exercise-2-a-large-study">Exercise 2: a large study</h3>

<p>In my qPCR experiment I want to study the pattern of expression of 96 genes (genes of interest and reference genes) in 96 samples of interest, divided into a few groups. I want to use 2 PCR replicates for each reaction.</p>

<table>
  <thead>
    <tr>
      <th style="text-align: left">Do I need to include IRCs (inter-run calibrators) ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">No, I can fit all samples of the same gene on the same plate so I don’t need to include IRCs.</td>
    </tr>
  </tbody>
</table>

<p>I want to include PCR replicates.</p>

<table>
  <thead>
    <tr>
      <th style="text-align: left">Do I need to include IRCs when I work on a 96 well plate ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">Yes, I have 192 reactions per gene so I cannot place them on the same plate. Remember that replicates have to be located on the same plate !</td>
    </tr>
  </tbody>
</table>

<table>
  <thead>
    <tr>
      <th style="text-align: left">Do I need to include IRCs when I work on a 384 well plate ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">No, I have 192 reactions per gene so I can even place two genes on the same plate.</td>
    </tr>
  </tbody>
</table>

<p>I want to include no template controls but I don’t want to increase the
number of plates.</p>

<table>
  <thead>
    <tr>
      <th style="text-align: left">What is the most elegant strategy to make room for including negative controls ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">This kind of study screen for expression patterns and requires statistical analysis. Since you have many samples divided over a few groups it means you have many biological replicates so you could easily do without the PCR replicates. By doing so you preserve the biological variability which is often far greater than the technical variation.</td>
    </tr>
  </tbody>
</table>

<h3 id="exercise-3-how-to-fill-plates-">Exercise 3: how to fill plates ?</h3>

<p>In my qPCR experiment I want to study the pattern of expression of 5 genes (genes of interest and reference genes) in 38 samples (samples of interest and control samples). I want to use 2 PCR replicates for each reaction.</p>

<p>| What is the minimum number of 96 well plates I need for this experiment ? |
| :—————————— |
| 5 genes * 38 samples * 2 replicates = 380 reactions.
I need a minimum of 4 plates for this experiment.</p>

<table>
  <thead>
    <tr>
      <th style="text-align: left">If I use the minimum number of 96 well plates do I need to include IRCs ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">Yes, 5 genes spread over 4 plates with 72 reactions per gene means that at least one gene will be spread over multiple plates.</td>
    </tr>
  </tbody>
</table>

<table>
  <thead>
    <tr>
      <th style="text-align: left">What can I do to avoid inter-run variability ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">I can use 5 plates and fill them with one gene each. They will not be completely filled (72 reactions) but at least I do not have to use IRCs (which are additional reactions that also cost money) and I have no inter-run variation.</td>
    </tr>
  </tbody>
</table>

<p>Suppose there’s only one 96-well plate left in your lab. You have 10 samples (samples of interest + control samples) and you want to make the most of what you have.</p>

<p>| How many genes of interest would you measure ? |
| :—————————— |
| Since you want to make most of what you have, let’s assume you are omitting PCR replicates.
Theoretically, you could fit 9 genes on your 96-well plate. However, to avoid pipetting mistakes I would measure only 8 genes so I can work with one row / gene. This is very handy for multichannel pipets.</p>

<h3 id="exercise-4-a-growing-study">Exercise 4: a growing study</h3>

<p>In my qPCR experiment I want to study the pattern of expression of 24 genes (genes of interest and reference genes) in 48 samples (samples of interest and control samples). I want to use 2 PCR replicates for each reaction.</p>

<p>| How many genes can I analyze on one 384 well plate ? |
| :—————————— |
| 48 samples * 2 replicates = 96 reactions per gene.
I can analyze 4 genes on each 384 well plate.</p>

<p>Each week I receive 2 additional samples to analyze.</p>

<p>| Do I analyze them immediately after I get them ? |
| :—————————— |
| No. Since the samples are placed on different plates as in the previous experiment, you have to use IRCs. You typically need 3 IRCs and a no template control sample. It means that if you want to analyze these 2 samples you have to include 4 additional samples for each gene. This is a lot of overhead for just 2 samples !
Try to avoid this: it’s better to wait a few weeks until you have 6 or 8 or even more samples.</p>

<h3 id="exercise-5-a-diagnostic-copy-number-screen">Exercise 5: a diagnostic copy number screen</h3>

<p>In diagnostic screens all samples are important: you cannot leave out samples and all measurements need to be of the highest quality possible. In my qPCR experiment I want to study copy number variation of 16 genes
(genes of interest and reference genes) and 2 calibrator samples (samples with known copy number). Since we need high quality data we will use 4 technical replicates.</p>

<table>
  <thead>
    <tr>
      <th style="text-align: left">Are we going to use sample maximization ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">No. In contrast to gene expression studies, where we want to compare expression levels of a gene between different groups of samples, copy number analyses do compare genes. It means that in this case the sample maximization approach (placing all samples of the same gene on the same plate) is not valid. Instead we use a gene maximization approach here (placing same sample for different genes on the same plate).</td>
    </tr>
  </tbody>
</table>

<p>| How many samples can I fit on a 384 well plate ? |
| :—————————— |
| We have 16 (genes) * 4 (replicates) = 64 reactions per sample.
This means that we can fit 6 samples on a 384 well plate: 4 unknowns and 2 calibrators.</p>

<h3 id="exercise-6-fix-experiments-with-bad-or-missing-data">Exercise 6: fix experiments with bad or missing data</h3>

<p>In my qPCR experiment I want to study gene expression of 6 genes (3 genes of interest and 3 reference genes) in 20 samples (samples of interest and control samples). I want to use 2 technical replicates. One of my genes of interest failed completely and I want to repeat the measurements for this gene in a new run.</p>

<table>
  <thead>
    <tr>
      <th style="text-align: left">Do I need to include IRCs ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">No. We can put the 20 samples of the gene that failed on a single plate so we do not have to include IRCs.</td>
    </tr>
  </tbody>
</table>

<table>
  <thead>
    <tr>
      <th style="text-align: left">Do I need to include reference genes ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">No. We just repeat all samples for the gene that failed and replace the old data with the new results.</td>
    </tr>
  </tbody>
</table>

<p>One of the reference genes failed completely.</p>

<table>
  <thead>
    <tr>
      <th style="text-align: left">What should I do ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">Depending on the quality of the two remaining reference genes, you should either do nothing or do the same as in the previous example where one of your genes of interest failed. If the two remaining reference genes are stable you can do the normalization with the two remaining reference genes.</td>
    </tr>
  </tbody>
</table>

<p>Three samples failed completely.</p>

<table>
  <thead>
    <tr>
      <th style="text-align: left">What’s the first thing I need to do ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">Since they failed completely, they are probably of low quality. Therefore, you have to prepare the samples again, check their quality and then use them for qPCR.</td>
    </tr>
  </tbody>
</table>

<table>
  <thead>
    <tr>
      <th style="text-align: left">Do I need to include IRCs ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">Yes. If you want to compare these samples with the samples that didn’t fail, you have to perform inter-run calibration.</td>
    </tr>
  </tbody>
</table>

<p>Three samples failed for one of the genes of interest</p>

<table>
  <thead>
    <tr>
      <th style="text-align: left">What is the first question I need to ask ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">Is the gene expressed in these samples ?</td>
    </tr>
  </tbody>
</table>

<table>
  <thead>
    <tr>
      <th style="text-align: left">Is it possible the RNA of these three samples was of low quality ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">Not likely, the measurements for the other genes in these samples are ok.</td>
    </tr>
  </tbody>
</table>

<p>Three samples failed for one of the reference genes</p>

<table>
  <thead>
    <tr>
      <th style="text-align: left">Can I use the measurements of that reference gene in the non-failing samples for normalization ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">No, qbasePLUS requires that you use the same reference genes for all samples so you have to discard all samples for that reference gene.</td>
    </tr>
  </tbody>
</table>

<h3 id="exercise-7-dilution-series-for-calculating-amplification-efficiencies">Exercise 7: dilution series for calculating amplification efficiencies</h3>

<p>In my qPCR experiment I want to study 8 new genes for which I had to design new primer pairs in 12 samples (samples of interest and control samples). I want to use 2 technical replicates and 96 well plates.</p>

<table>
  <thead>
    <tr>
      <th style="text-align: left">What is the first thing I need to do ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">Perform a pilot experiment to determine the amplification efficiencies of these primer pairs.</td>
    </tr>
  </tbody>
</table>

<p>For this I need a dilution series of representative cDNA template.</p>

<table>
  <thead>
    <tr>
      <th style="text-align: left">How many dilutions would you include ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">A dilution series with 6 dilutions for 8 genes nicely fits into a 96 well plate.</td>
    </tr>
  </tbody>
</table>

<p>A few weeks after my initial qPCR experiment I want to test these 8 genes in a new set of samples.</p>

<table>
  <thead>
    <tr>
      <th style="text-align: left">Do I have to repeat the pilot experiment ?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left">No, dilution series do not need to be repeated.</td>
    </tr>
  </tbody>
</table>



                    

                    

                    

                </div>
                <!-- sidebar, which will move to the top on a small screen -->
                <div class="col-sm-2" role="supplementary">
                    <nav id="toc" data-toggle="toc" class="sticky-top bs-docs-sidebar hidden-print hidden-xs hidden-sm"></nav>
                </div>
            </div>
        </div>

        <h3><i class="fa fa-thumbs-up" aria-hidden="true"></i><span class="visually-hidden">congratulations</span> Congratulations on successfully completing this tutorial!</h3>

        

        

</div>
</div>


<footer>
        <div class="well row">
         <div class="col-md-3 text-center">             
            <span class="slogan">SCIENCE MEETS LIFE</span>
         </div>
         <div class="col-md-6">
            This material is the result of a collaborative work. Thanks to the
            <a href="https://www.bits.vib.be">VIB Bioinformatics Core</a>
            and all the <a href="/training-material/hall-of-fame">contributors</a>.
            Found a typo? Something is wrong in this tutorial? Edit it on
            <a href="https://github.com/vibbits/training-material/tree/master/topics/qbase-plus/tutorials/experiment-design/tutorial.md">GitHub</a>.
          </div>
          <div class="col-md-3 text-center">
                <a href="http://www.bits.vib.be"><img src="/training-material/assets/images/bioinformatics_core_rgb_neg.png"></a>
       </div>
   </div>
   </footer>
</div>

    </body>
    <script type="text/javascript" src="/training-material/assets/js/jquery.slim.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/popper.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/bootstrap.min.js?v=3"></script>
    <script type="text/javascript" src="/training-material/assets/js/details-element-polyfill.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="/training-material/assets/js/bootstrap-toc.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/main.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/clipboard.min.js"></script>
    <script type="text/javascript" src="https://app.cloudsh.com/js/cloudsh.js"></script>
    <script type="text/javascript">
    var snippets=document.querySelectorAll('div.highlight');
    [].forEach.call(snippets,function(snippet){
        snippet.firstChild.insertAdjacentHTML('beforebegin','<button class="btn btn-light" data-clipboard-snippet><i class="fa fa-copy"></i>&nbsp;Copy</button>');
    });

    var clipboardSnippets=new ClipboardJS('[data-clipboard-snippet]',{
        target:function(trigger){return trigger.nextElementSibling;
    }});
    </script>
</html>
