<!doctype html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="google-site-verification" content="dw_1zBYRNxF7-vYNDXh3Ght_JuqgJt3BD1N_Pynqmv0" />
    <title>06 Protein-ligand interaction - from small molecule to protein</title>
    <script async="" src="https://www.google-analytics.com/analytics.js"></script>
    <script>
        (function(i, s, o, g, r, a, m) {
            i['GoogleAnalyticsObject'] = r;
            i[r] = i[r] || function() {
                (i[r].q = i[r].q || []).push(arguments)
            }, i[r].l = 1 * new Date();
            a = s.createElement(o),
                m = s.getElementsByTagName(o)[0];
            a.async = 1;
            a.src = g;
            m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');
        ga('create', 'UA-61194136-18', 'auto');
        ga('send', 'pageview');
    </script>
    <link rel="stylesheet" href="/assets/css/bootstrap.min.css?v=3">
    <link rel="stylesheet" href="/assets/css/bootstrap-toc.min.css">
    <link rel="stylesheet" href="/assets/css/font-awesome.css">
    <link rel="stylesheet" href="/assets/css/main.scss">
    <link rel="stylesheet" href="/assets/css/VIB.css">
    <link rel="stylesheet" href="/assets/css/academicons.css">
    <link rel="stylesheet" href="/assets/css/syntax_highlighting.css">
    <link rel="shortcut icon" href="/favicon.ico " type="image/x-icon" />     
    <meta name="description" content="Analyzing the protein structure of your protein-of-intere..." />
    <meta property="og:title" content="VIB Bioinformatics Training: 06 Protein-ligand interaction - from small molecule to pr..." />
    <meta property="og:description" content="Analyzing the protein structure of your protein-of-intere..." />
    <meta property="og:image" content="/assets/images/bioinformatics_core_rgb_neg.png" />
</head>

<body data-spy="scroll" data-target="#toc">
    <div class="top-bar-vib">
 <a href="http://vib.be/en" target="_blank" style="font-size:14px">VIB Home </a>
</div>

<div class="container main-content">

<div class="container page-top__container">
          <div class="region region-top">
  <div id="logo" class="col-md-3 col-lg-3 text-left" style="padding-left:0px;margin-top:20px">
       <a href="/">
           <img src="/assets/images/logo.svg" height="60" alt="VIB logo" class="margin20">
       </a>
    </div>
    <div class="col-md-9 col-lg-9 text-right" style="margin-top:50px;">
    <nav role="navigation" aria-labelledby="block-mainnavigation-menu" id="block-mainnavigation" class="block block-menu navigation menu--main">
  


        <a href="/" class="topmenu">Home</a>


        <a href="/about.html" class="topmenu" >About</a>


        <a href="/hall-of-fame.html" class="topmenu">Contribute</a>


        <a href="/faq.html" class="topmenu">Faq</a>


        <a href="/search/index.html" class="topmenu">Search</a>


  </nav>

    </div>
  </div>
</div>














    <script type="application/ld+json">
        


{
  "@context": "http://schema.org",
  "@type": "Course",
  "accessMode": [
    "textual",
    "visual"
  ],
  "accessModeSufficient": [
    "textual",
    "visual"
  ],
  "accessibilityControl": [
    "fullKeyboardControl",
    "fullMouseControl"
  ],
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "accessibilitySummary": "Short descriptions are present but long descriptions will be needed for non-visual users",
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "copyrightHolder": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "headline": "06 Protein-ligand interaction - from small molecule to protein",
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "license": "/blob//LICENSE.md",
  "producer": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "provider": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "sourceOrganization": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "courseCode": "protein-structure-analysis / protein-ligand-interaction / hands-on",
  "learningResourceType": "hands-on tutorial",
  "name": "Hands-on for '06 Protein-ligand interaction - from small molecule to protein' tutorial",
  "isPartOf": {
    "@type": "Course",
    "name": "06 Protein-ligand interaction - from small molecule to protein",
    "description": "06 Protein-ligand interaction - from small molecule to protein",
    "learningResourceType": "tutorial",
    "interactivityType": "expositive",
    "provider": {
      "@type": "Organization",
      "email": "bits@vib.be",
      "name": "VIB Bioinformatics Core",
      "url": "https://www.bits.vib.be"
    }
  },
  "url": "https://material.bits.vib.be//topics/protein-structure-analysis/tutorials/protein-ligand-interaction/tutorial.html",
  "timeRequired": "PT30M",
  "description": "06 Protein-ligand interaction - from small molecule to protein\\nThe questions this  addresses are:\n - How can visualisation techniques help to derive biologcial insights?\n - Which tools exists to display interaction of e.g. small molecules with proteins?\n\n\\nThe objectives are:\n - Describe which interaction types are important for e.g. ligand binding to a receptor.\n\n",
  "hasPart": [

  ],
  "mentions": [
    {
      "@type": "Thing",
      "url": "",
      "name": "Training data for 06 Protein-ligand interaction - from small molecule to protein tutorial"
    }
  ],
  "author": [
    {
      "@type": "Person",
      "name": "Alexander Botzki"
    }
  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "Alexander Botzki"
    }
  ],
  "about": [
    {
      "@type": "CreativeWork",
      "name": "Protein Structure Analysis",
      "description": "Analyzing the protein structure of your protein-of-interest can be advantageous in multiple ways. It can help you discover regions which are good candidates to interact with other proteins. It can help you discover new domains. It can help with identifying differences with homologuous proteins and a lot more.",
      "url": "https://material.bits.vib.be//topics/protein-structure-analysis/"
    }
  ]
}
    </script>

    <div class="row tutorial">
      <div class="col-md-12" role="main">
        <h1 data-toc-skip>06 Protein-ligand interaction - from small molecule to protein</h1>
        <div class="contributors-line">By: 

<a href="/hall-of-fame#abotzki" class="contributor-badge">Alexander Botzki</a>


</div>
        <blockquote class="overview">

            <strong><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</strong>
            <ul>
            
            <li>How can visualisation techniques help to derive biologcial insights?</li>
            
            <li>Which tools exists to display interaction of e.g. small molecules with proteins?</li>
            
            </ul>

            <strong><i class="fa fa-bullseye" aria-hidden="true"></i><span class="visually-hidden">objectives</span> Objectives</strong>
            <ul>
            
            <li>Describe which interaction types are important for e.g. ligand binding to a receptor.</li>
            
            </ul>

            

            
            <p><strong><i class="fa fa-hourglass-end" aria-hidden="true"></i><span class="visually-hidden">time</span> Time estimation:</strong> 30 minutes</p>
            

            


            

        </blockquote>

        <div class="container" data-field="body">
            <div class="row">
                <div class="col-md-10" role="main">
                    <h1 class="no_toc" id="introduction">Introduction</h1>

<!-- This is a comment. -->

<p>The goal of this exercise is appreciate how protein interactions can be studied through visual inspection and other software tools. Protein interactions can be classified into different groups regarding the molecular properties and functions of the interacting partners. (These groups are intertwined in several cases.) Some examples include:</p>

<ul>
  <li>the interactions of proteins with other proteins, small molecules, carbohydrates, lipids or nucleic acids;</li>
  <li>Receptor-ligand interactions;</li>
  <li>Antigen-antibody interactions;</li>
  <li>Enzymatic interactions, enzyme-inhibitor interactions.</li>
</ul>

<h2 id="exploring-the-structure-of-a-nanobody-stabilized-active-state-of-the-β2-adrenoceptor---the-ligand">Exploring the structure of a nanobody-stabilized active state of the β2 adrenoceptor - the ligand</h2>

<p>We will start with exploring one crystal structure of the β2 adrenoceptor. Together with the Steyaert lab from VIB, Kobilka published several crystal structures of the β2 adrenoceptor in its various activation states (Rasmussen et al. Nature 2011, 477)</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-get-the-structure"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Get the structure</h3>

  <ol>
    <li>
      <p>Download the crystal structure 3P0G from the PDB into YASARA.</p>

      <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>File - Load - PDB file from internet    
</code></pre></div>      </div>
      <p>As you can immediately appreciate, it is a bigger crystal structure with more than one molecule.</p>
    </li>
  </ol>

</blockquote>

<blockquote class="question">
  <h3 id="question-questions"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</h3>

  <ol>
    <li>How many molecules are present in the crystallized structures?</li>
    <li>And how many chain identifiers are used?</li>
  </ol>

  <details>
    <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>

    <ol>
      <li>There are three molecules, chain A Beta-2 adrenergic receptor; Endolysin, chain B Camelid Antibody Fragment, and a small molecule ligand. 
  Also have a look at PDBe <a href="https://www.ebi.ac.uk/pdbe/entry/pdb/3p0g">3P0G</a> which gives a very nice overview of the structure and its composition.</li>
      <li>Only two chain identifiers A and B. Sometimes, this leads to issues depending on the software you might want to use for downstream processing.</li>
    </ol>

  </details>

</blockquote>

<p>Some software routines need seperate chain identifiers for molecular entities to work correctly, so we suggest to rename the small molecule to chain L.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span></h3>

  <ol>
    <li>Activate the Head-up display</li>
    <li>Select Rename</li>
    <li>Enter ‘L’ to proceed with the renaming.</li>
  </ol>

</blockquote>

<p>We first have a look whether we can find out if there are specific interactions of the small molecule ligand with the adrenoreceptor.</p>

<p>In order to do so, we first have to add Hydrogens to all present molecules.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-1"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span></h3>

  <ol>
    <li>Edit - Add - hydrogens to : All</li>
    <li>Change the display of the ligand to Sticks</li>
    <li>Select the amino acids of the binding pocket i.e. a sphere of 10 Angstrom around the ligand
      <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Select – in sphere around – Residue and drag with the mouse until the display says 10 Å
</code></pre></div>      </div>
    </li>
    <li>
      <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>View – show interactions – hydrogen bonds of - Residues
</code></pre></div>      </div>
      <p>select ‘Selected’ in the panel Belongs to or has
and press OK in the subsequent window</p>
    </li>
  </ol>

</blockquote>

<p>Given that hydrogen bonding is dependent on the definition of a hydrogen bond in the program, it is not a bad idea to use other tools to compare the analysis. There are many options to do this online if you look at published crystal structures. Next to the tools which are directly linked out from the web site of the crystal structure at the PDB database you can use the <a href="http://proteinsplus.zbh.uni-hamburg.de/">ProteinPlus server</a></p>

<p>Go to the web site of ProteinPlus and enter the PDB code 3P0G into the search box. After clicking on Go, you should be presented with on overview of tools the ProteinPlus server provides.</p>

<p>We do not go into great detail on all the tools but only mention PoseView. With this tool, you can prepare an automatic sketch of the small molecule-protein interactions.</p>

<figure id="figure-1"><img src="../../images/ProteinPlusPoseView.png" alt="Protein Plus Server" /><figcaption><span class="figcaption-prefix">Figure 1:</span> Overview of 3P0G</figcaption></figure>
<figure id="figure-2"><img src="../../images/3P0G_A_PoseView_Input.png" alt="Zoom on ligand of 3P0G" /><figcaption><span class="figcaption-prefix">Figure 2:</span> Zoom on ligand co-crystallized with 3P0G</figcaption></figure>

<blockquote class="question">
  <h3 id="question-questions-1"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</h3>

  <ol>
    <li>Between which amino acids and the ligand do you see hydrogen bonds using YASARA?</li>
    <li>According to PoseView, between which amino acids and the ligand do you see hydrogen bonds?</li>
    <li>What other interactions are presented in the sketch?</li>
    <li>Inspect the visualisation in Yasara: Do you see the interactions in Yasara as well?</li>
  </ol>

  <details>
    <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>

    <ol>
      <li>
        <p>In YASARA, you observe hydrogen bonds between Asp113A as well as the carbonyl function of Asn312A and the charged amine function.</p>
      </li>
      <li>
        <p>PoseView indicates hydrogen bonds between Asp113A as well as the carbonyl function of Asn312A and the charged amine function. Furthermore, hydrogen bonds are indicated between the phenolic OH and Ser207A and Ser203A as well as the amine function and Ser203A.</p>
      </li>
      <li>
        <p>Furthermore, hydrophobic interactions are indicated for the methylbenzyl moiety and pi-pi interactions of Phe290A and the phenolic moiety.</p>
      </li>
      <li>
        <p>With YASARA Structure license, those hydrophobic interactions can also be visualised.</p>
      </li>
    </ol>
  </details>
</blockquote>

<h1 id="exploring-the-structure-of-a-nanobody-stabilized-active-state-of-the-β2-adrenoceptor---the-nanobody">Exploring the structure of a nanobody-stabilized active state of the β2 adrenoceptor - the nanobody</h1>

<p>In order to estimate the binding energy between the nanobody and the β2 adrenoceptor, we can use the FoldX tool AnalyseComplex. It is recommended to calculate these binding energies on energy-minimized structures. To illustrate the effect of the energy minimization, we compare the interaction energy of the current crystal structure and its minimized structure.</p>

<h2 id="use-the-tool-foldx-tool-analysecomplex">Use the tool FoldX tool AnalyseComplex</h2>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-2"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span></h3>

  <ol>
    <li>
      <p>Given that energy-minimization takes a while for this rather large complex,
 please download the Yasara scene <a href="http://data.bits.vib.be/pub/trainingen/PSA/3P0G_1.sce">here</a></p>

      <p>Calculate the interaction energies between the chain A and B of the object 3P0G 
and the RepairObj1, respectively.</p>

      <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Analyze - FoldX - Interaction energy of molecules
</code></pre></div>      </div>
    </li>
  </ol>
</blockquote>

<blockquote class="question">
  <h3 id="question-questions-2"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</h3>

  <ol>
    <li>What is the dG in the two cases?</li>
    <li>Any idea why the difference is rather hugh?</li>
  </ol>

  <details>
    <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>

    <ol>
      <li>first case (X-ray structure): Interaction energy between molecule(s) A and B in object 1 = -9.86 (kcal/mol)
second case: 
Interaction energy between molecule(s) A and B in object 2 = -20.19 (kcal/mol)</li>
      <li>Through the energy minimisation of the Repair Object function, the interactions of the amino acids are optimised.</li>
    </ol>
  </details>
</blockquote>

<p>This command also creates a list of residues forming the interface of the two proteins. Hit space to see the list of residues in the interface.</p>

<p>Tip: This list can also be useful if you want to make visualisations of the interaction site.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Plugin&gt;interface residues between A and B
Plugin&gt;TA66 TA68 IA72 IA127 RA131 AA134 IA135 TA136 SA137 PA138 FA139 KA140 QA142 YA219 VA222 EA225 AA226 LA266 KA267 EA268 AA271 LA272 TA274 LA275 IA278 IA325 YA326 RA328 SA329 PA330 SB27 IB28 FB29 SB30 IB31 TB33 AB50 IB51 eB52 SB56 TB57 NB58 YB100 AB102 VB103 LB104 YB105 EB106 YB107
</code></pre></div></div>

<h1 id="comparing-the-active-and-the-inactive-conformation-of-the-β2-adrenoceptor">Comparing the active and the inactive conformation of the β2 adrenoceptor</h1>

<p>In case, there is still time, I would recommend to try to use some of your capabilities you learned today and create a superposition of the inactive and active conformation of the β2 adrenoceptor. We take one of the crystal structures which are available: 3SN6</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>File - Load - PDB file from Internet
</code></pre></div></div>

<p>You will be kind of overwhelmed once the structure is loaded into YASARA. In order to get a first quick overview, click on the ‘Center’ buttom in the menu of YASARA (5th buttom from the left). Then, it is time to look at the PDB entry of 3SN6 in the PDB database to have a first idea on what molecules are in the PDB file.</p>

<p>As you see on the website <a href="http://www.rcsb.org/pdb/explore/explore.do?structureId=3SN6i">3SN6</a>, the chain R consists of 2 molecules, the β2 adrenoceptor and lysozyme. 
In the corresponding article, it is stated that ‘the unstructured amino terminus of the β2AR is replaced with T4 lysozyme (T4L)’.</p>

<p>Since this is an extra molecule in the crystal structure which disturbes our view, we will delete it.</p>

<p>After the manipulation, the overall picture should look roughly like this.</p>

<figure id="figure-3"><img src="../../images/3SN6_withoutLysozyme.png" alt="Superposition" /><figcaption><span class="figcaption-prefix">Figure 3:</span> Overview of 3SN6 without lysozyme</figcaption></figure>

<p>In the following step, we structurally align only the receptors. The rest of the structures will move along.
It is suggested to use the first chain A from 3P0G as target. In order to do a structural alignment, it is suggested to use the first chain A from 3P0G as target.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Analyze - Align - Pairwise, based on structure - Molecules with MUSTANG
</code></pre></div></div>

<p>Investigate the differences in TM helices and the binding of the nanobody compared to the subunit of the G protein.</p>

<p>Tip: Color the secondary structures to better identify the individual chains/units of G protein.</p>

<h1 class="no_toc" id="conclusion">Conclusion</h1>

<p>Sum up the tutorial and the key takeaways here. We encourage adding an overview image of the
pipeline used.</p>



                    
                    <blockquote class="key_points">
                        <h3><i class="fa fa-key" aria-hidden="true"></i><span class="visually-hidden">keypoints</span> Key points</h3>

                        <ul>
                            
                            <li>Visualisation of interaction types depends on settings of the visualising tool.</li>
                            
                            <li>Visualisation techniques together with structural alignment lead to the discovery of biological insights.</li>
                            
                        </ul>
                    </blockquote>
                    

                    
                    <h1 data-toc-skip>Useful literature</h1>
                    <p>Further information, including links to documentation and original publications, regarding the tools, analysis techniques and the interpretation of results described in this tutorial can be found <a href="/topics/protein-structure-analysis#references">here</a>.</p>
                    

                    

                </div>
                <!-- sidebar, which will be hidden on a small screen -->
                <nav class="col-md-2 sticky-toc bs-docs-sidebar hidden-print hidden-xs hidden-sm" role="complementary" id="toc" data-toggle="toc"></nav>
            </div>
        </div>

        <h3><i class="fa fa-thumbs-up" aria-hidden="true"></i><span class="visually-hidden">congratulations</span> Congratulations on successfully completing this tutorial!</h3>

        

        

</div>
</div>
</div>
<footer>
   <div class="well row">
       <div class="col-md-2 text-center" style="padding: 25px 20px 25px 20px;">				
         <a href="https://www.bits.vib.be" target="_blank"><img src="/assets/images/bioinformatics_core_rgb_neg.png" width="200"></a>
       </div>
       <div class="col-md-8 text-center" style="padding-top: 20px;">	
        <a href="https://github.com/vibbits/training-material" title="View on GitHub" class="margin20" target="_blank">
                            <i class="fa fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> View on GitHub
                        </a>
       </div>
       <div class="col-md-2 text-center" style="padding-top: 30px;">
         <span class="slogan"><br>SCIENCE MEETS LIFE</span>
       </div>
   </div>
</footer>


</body>
<script type="text/javascript" src="/assets/js/jquery.slim.min.js"></script>
<script type="text/javascript" src="/assets/js/popper.min.js"></script>
<script type="text/javascript" src="/assets/js/bootstrap.min.js?v=3"></script>
<script type="text/javascript" src="/assets/js/details-element-polyfill.js"></script>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script type="text/javascript" src="/assets/js/bootstrap-toc.min.js"></script>
<script type="text/javascript" src="/assets/js/main.js"></script>
<script type="text/javascript" src="/assets/js/clipboard.min.js"></script>
<script type="text/javascript">
    var snippets = document.querySelectorAll('div.highlight');
    [].forEach.call(snippets, function(snippet) {
        snippet.firstChild.insertAdjacentHTML('beforebegin', '<button class="btn btn-light" data-clipboard-snippet><i class="fa fa-copy"></i>&nbsp;Copy</button>');
    });

    var clipboardSnippets = new ClipboardJS('[data-clipboard-snippet]', {
        target: function(trigger) {
            return trigger.nextElementSibling;
        }
    });
</script>

</html>
