<!doctype html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="google-site-verification" content="dw_1zBYRNxF7-vYNDXh3Ght_JuqgJt3BD1N_Pynqmv0" />
    <title>02 Linux command line</title>
    <script async="" src="https://www.google-analytics.com/analytics.js"></script>
    <script>
        (function(i, s, o, g, r, a, m) {
            i['GoogleAnalyticsObject'] = r;
            i[r] = i[r] || function() {
                (i[r].q = i[r].q || []).push(arguments)
            }, i[r].l = 1 * new Date();
            a = s.createElement(o),
                m = s.getElementsByTagName(o)[0];
            a.async = 1;
            a.src = g;
            m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');
        ga('create', 'UA-61194136-18', 'auto');
        ga('send', 'pageview');
    </script>
    <link rel="stylesheet" href="/assets/css/bootstrap.min.css?v=3">
    <link rel="stylesheet" href="/assets/css/bootstrap-toc.min.css">
    <link rel="stylesheet" href="/assets/css/font-awesome.css">
    <link rel="stylesheet" href="/assets/css/main.scss">
    <link rel="stylesheet" href="/assets/css/VIB.css">
    <link rel="stylesheet" href="/assets/css/academicons.css">
    <link rel="stylesheet" href="/assets/css/syntax_highlighting.css">
    <link rel="shortcut icon" href="/favicon.ico " type="image/x-icon" />     
    <meta name="description" content="This training is a prerequisite for all Linux novices who..." />
    <meta property="og:title" content="VIB Bioinformatics Training: 02 Linux command line" />
    <meta property="og:description" content="This training is a prerequisite for all Linux novices who..." />
    <meta property="og:image" content="/assets/images/bioinformatics_core_rgb_neg.png" />
</head>

<body data-spy="scroll" data-target="#toc">
    <div class="top-bar-vib">
 <a href="http://vib.be/en" target="_blank" style="font-size:14px">VIB Home </a>
</div>

<div class="container main-content">

<div class="container page-top__container">
          <div class="region region-top">
  <div id="logo" class="col-md-3 col-lg-3 text-left" style="padding-left:0px;margin-top:20px">
       <a href="/">
           <img src="/assets/images/logo.svg" height="60" alt="VIB logo" class="margin20">
       </a>
    </div>
    <div class="col-md-9 col-lg-9 text-right" style="margin-top:50px;">
    <nav role="navigation" aria-labelledby="block-mainnavigation-menu" id="block-mainnavigation" class="block block-menu navigation menu--main">
  


        <a href="/" class="topmenu">Home</a>


        <a href="/about.html" class="topmenu" >About</a>


        <a href="/hall-of-fame.html" class="topmenu">Contribute</a>


        <a href="/faq.html" class="topmenu">Faq</a>


        <a href="/search/index.html" class="topmenu">Search</a>


  </nav>

    </div>
  </div>
</div>














    <script type="application/ld+json">
        


{
  "@context": "http://schema.org",
  "@type": "Course",
  "accessMode": [
    "textual",
    "visual"
  ],
  "accessModeSufficient": [
    "textual",
    "visual"
  ],
  "accessibilityControl": [
    "fullKeyboardControl",
    "fullMouseControl"
  ],
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "accessibilitySummary": "Short descriptions are present but long descriptions will be needed for non-visual users",
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "copyrightHolder": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "headline": "02 Linux command line",
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "license": "/blob//LICENSE.md",
  "producer": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "provider": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "sourceOrganization": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "courseCode": "linux / command-line / hands-on",
  "learningResourceType": "hands-on tutorial",
  "name": "Hands-on for '02 Linux command line' tutorial",
  "isPartOf": {
    "@type": "Course",
    "name": "02 Linux command line",
    "description": "02 Linux command line",
    "learningResourceType": "tutorial",
    "interactivityType": "expositive",
    "provider": {
      "@type": "Organization",
      "email": "bits@vib.be",
      "name": "VIB Bioinformatics Core",
      "url": "https://www.bits.vib.be"
    }
  },
  "url": "https://material.bits.vib.be//topics/linux/tutorials/command-line/tutorial.html",
  "timeRequired": "PT30M",
  "description": "02 Linux command line\\nThe questions this  addresses are:\n - How to get started with command line?\n\n\\nThe objectives are:\n - get familiar with Linux command line\n - learn how to execute programs\n - learn what commands to use when navigating the file system\n\n",
  "hasPart": [

  ],
  "mentions": [

  ],
  "author": [
    {
      "@type": "Person",
      "name": "Christof De Bo"
    }
  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "Christof De Bo"
    }
  ],
  "about": [
    {
      "@type": "CreativeWork",
      "name": "Initiation to Linux Command Line",
      "description": "This training is a prerequisite for all Linux novices who want to follow a training that requires the use of Linux command line (e.g. metagenomics or GATK training).",
      "url": "https://material.bits.vib.be//topics/linux/"
    }
  ]
}
    </script>

    <div class="row tutorial">
      <div class="col-md-12" role="main">
        <h1 data-toc-skip>02 Linux command line</h1>
        <div class="contributors-line">By: 

<a href="/hall-of-fame#chdeb" class="contributor-badge">Christof De Bo</a>


</div>
        <blockquote class="overview">

            <strong><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</strong>
            <ul>
            
            <li>How to get started with command line?</li>
            
            </ul>

            <strong><i class="fa fa-bullseye" aria-hidden="true"></i><span class="visually-hidden">objectives</span> Objectives</strong>
            <ul>
            
            <li>get familiar with Linux command line</li>
            
            <li>learn how to execute programs</li>
            
            <li>learn what commands to use when navigating the file system</li>
            
            </ul>

            

            
            <p><strong><i class="fa fa-hourglass-end" aria-hidden="true"></i><span class="visually-hidden">time</span> Time estimation:</strong> 30 minutes</p>
            

            


            

        </blockquote>

        <div class="container" data-field="body">
            <div class="row">
                <div class="col-md-10" role="main">
                    <h2 id="tutorial-on-the-linux-command-line">Tutorial on the linux command line</h2>
<p>We will first hold your hand: type over these commands below step by step, and watch what they do.
Use <strong>cd</strong> to change the current working directory (user bits). To create your own directories use the <strong>mkdir</strong> (make directory) command.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ cd ~
$ mkdir sequences
$ cd sequences
$ mkdir proteins
$ cd proteins
$ pwd
/home/bits/sequences/proteins
$ cd ../..
$ pwd
/home/bits
</code></pre></div></div>

<p>To create a new file, use the <strong>touch</strong> command:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ cd ~/sequences/proteins/
$ touch my_sequence.txt
$ ls -l
-rw-r--r-- 1 bits users 0 Sep 19 15:56 my_sequence.txt
</code></pre></div></div>

<p>In the last command above, the <strong>-l</strong> (a lowercase “L”, not a “1” (one)) option was used with the ls command. The -l indicates that you want the directory contents shown in the “long listing” format.
Most commands accept options. But which options can you use? The command <strong>man</strong> helps you. Type <strong>man</strong> followed by the command name. E.g. <strong>man ls</strong> to see what options are available for the ls command. You get a the list of options. Keep pressing Space until the page stops scrolling, then enter “q” to return to the command prompt.
Luckily, most tools have the <strong>–help</strong> option. (ls –help for example). These 2 methods should help you further. To see what options can be used with ls, enter <strong>man ls</strong>.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ man ls
</code></pre></div></div>

<p>To delete a file, use the <strong>rm</strong> (remove) command:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ cd ~/sequences/proteins/
$ ls
my_sequence.txt
$ rm my_sequence.txt
$ ls
$
</code></pre></div></div>

<p>To remove a directory, use the <strong>rmdir</strong> (remove directory) command. The directory needs to be empty to do this.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ cd ~/sequences/
$ ls
proteins
$ rmdir proteins
$ ls
$
</code></pre></div></div>

<p>To copy a file, use the <strong>cp</strong> (copy) command:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ cd ~/sequences
$ touch testfile1
$ ls
testfile1
$ cp testfile1 testfile2
$ ls
testfile1 testfile2
</code></pre></div></div>

<p>To rename a file, or to move it to another directory, use the <strong>mv</strong> (move) command:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ cd 
$ touch testfile3
$ mv testfile3 junk
$ mkdir testdir
$ mv junk testdir
$ ls testdir
junk
</code></pre></div></div>

<p>To download a file, use the <strong>wget</strong> command:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ cd ~/Downloads
$ wget http://data.bits.vib.be/pub/trainingen/Linux/sample.sam
$ ls sample.sam
$
</code></pre></div></div>

<p>The commands covered so far represent a small but useful subset of the many commands available on a typical Linux system.</p>

<h3 id="make-a-project-folder-structure">Make a project folder structure</h3>
<p>We assume that start from your home folder.</p>
<blockquote>
  <p>Create the following directory structure</p>
  <figure id="figure-1"><img src="https://wiki.bits.vib.be/images/c/c4/Dirstructureex1.png" alt="tree-structure" /><figcaption><span class="figcaption-prefix">Figure 1:</span> Tree</figcaption></figure>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$mkdir -p docs/{pdf/{man,readme},html/{man,readme}}
</code></pre></div>  </div>
</blockquote>

<p>The ‘{‘ and ‘}’ can group arguments but you can also create the structure step by step.</p>
<blockquote>
  <p>The little tree figure above is created with the ‘tree’ command. Display such a tree.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>tree /home/bits/docs/
</code></pre></div>  </div>
</blockquote>

<h2 id="downloading-and-storing-bioinformatics-data">Downloading and storing bioinformatics data</h2>

<h3 id="create-a-project-folder">Create a project folder</h3>
<p>The first thing to do when you start a bioinformatics project, is to create a structure of folders to put your data in an organised fashion.</p>

<h3 id="downloading">Downloading</h3>
<p>As an example, we will download the rice genome from the Rice Annotation Project database. But first create the folder structure.</p>
<blockquote>
  <p>Create following folder structure.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code> $ mkdir "Rice Example"
 $ cd Rice\ Example
 $ mkdir Genome\ data
 $ cd Genome\ data
 $ mkdir Sequence
 $ mkdir Annotation
 $ cd
</code></pre></div>  </div>
</blockquote>

<p>** Be aware of white spaces on the command line!**
On the command line, programs, options and arguments are separated by <strong>white spaces</strong>. If you choose to use a folder name containing a white space, it will interpret every word as an option or argument. So you have to tell Bash to <strong>ignore the white space</strong>. This can be done by: putting <strong>strings between quotes</strong> like ‘ or “ <strong>escape</strong> a white space with . See the examples above.
Hence, you might save yourself some trouble (and typing!) by putting _ instead of white spaces in names. Also make sure to use tab expansion, wherever possible!</p>

<h3 id="download-the-genome-data-directly-on-the-command-line">Download the genome data directly on the command line</h3>
<p>You can fetch the rice genome <a href="http://rapdb.dna.affrc.go.jp/download/archive/build5/IRGSPb5.fa.masked.gz">from this link</a>.</p>
<blockquote>
  <p>Download the genome data to the “Rice example”/”Genome data”/Sequence folder. Use <strong>wget</strong> to download from the link.
Right-click on the download link, and copy the download link. The download link is: http://rapdb.dna.affrc.go.jp/download/archive/build5/IRGSPb5.fa.masked.gz
Go the directory and execute wget</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ cd      ## to go back to the home directory
$ cd Ric&lt;tab&gt;
$ cd Gen&lt;tab&gt;/Seq&lt;tab&gt;
$ wget http://rapdb.dna.affrc.go.jp/download/archive/build5/IRGSPb5.fa.masked.gz
--2013-10-15 09:36:01--  http://rapdb.dna.affrc.go.jp/download/archive/build5/IRGSPb5.fa.masked.gz
Resolving rapdb.dna.affrc.go.jp (rapdb.dna.affrc.go.jp)... 150.26.230.179
Connecting to rapdb.dna.affrc.go.jp (rapdb.dna.affrc.go.jp)|150.26.230.179|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 122168025 (117M) [application/x-gzip]
Saving to: `IRGSPb5.fa.masked.gz'
100%[======================================&gt;] 122,168,025  973K/s   in 2m 40s  
2013-10-15 09:38:42 (747 KB/s) - `IRGSP-1.0_genome.fasta.gz' saved [122168025/122168025]
$ ls
IRGSPb5.fa.masked.gz
</code></pre></div>  </div>
</blockquote>

<p>Allright. We have fetched our first genome sequence!</p>

<h3 id="did-your-data-get-through-correctly">Did your data get through correctly?</h3>
<p>Large downloads or slow downloads like this can take a long time. Plenty of opportunity for the transfer to go wrong. Therefore, large downloads should always have a <strong>checksum</strong> mentioned. You can find the md5 checksum on the downloads page. The md5 checksum is an unique string identifying (and calculated from) this data. Once downloaded, you should calculate this string yourself with <strong>md5sum</strong>.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ md5sum IRGSPb5.fa.masked.gz
7af391c32450de873f80806bbfaedf05  IRGSPb5.fa.masked.gz
</code></pre></div></div>

<p>You should go to the rice genome download page, and compare this string with the MD5 checksum mentioned over there. You can do this manually. Now that you know the concept of checksums, there is an easier way to verify the data using <strong>md5sum</strong>. Can you find the easier way?</p>
<blockquote>
  <p>Search how to use md5sum to check the downloaded files with the .md5 file from the website. Check the man page</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ man md5sum
</code></pre></div>  </div>
</blockquote>

<p>It does not say much: in the end it refers to</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ info coreutils 'md5sum invocation'
</code></pre></div></div>

<p>Reading the options, there is one option sounding promising:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>`-c'
`--check'
     Read file names and checksum information (not data) from each FILE
     (or from stdin if no FILE was specified) and report whether the
     checksums match the contents of the named files. 
</code></pre></div></div>

<p>This way we can check the download:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ wget http://rapdb.dna.affrc.go.jp/download/archive/build5/IRGSPb5.fa.masked.gz.md5
--2013-10-15 09:47:02--  http://rapdb.dna.affrc.go.jp/download/archive/build5/IRGSPb5.fa.masked.gz.md5
Resolving rapdb.dna.affrc.go.jp (rapdb.dna.affrc.go.jp)... 150.26.230.179
Connecting to rapdb.dna.affrc.go.jp (rapdb.dna.affrc.go.jp)|150.26.230.179|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 55 [application/x-gzip]
Saving to: `IRGSPb5.fa.masked.gz.md5'
100%[======================================&gt;] 55          --.-K/s   in 0s      
2013-10-15 09:47:03 (757 KB/s) - `IRGSPb5.fa.masked.gz.md5' saved [55/55]
$ ls
IRGSPb5.fa.masked.gz  IRGSPb5.fa.masked.gz.md5
$ md5sum -c IRGSPb5.fa.masked.gz.md5 
IRGSPb5.fa.masked.gz: OK
</code></pre></div></div>

<h2 id="ensuring-integrity-of-downloads">Ensuring integrity of downloads</h2>
<p>A handy tool to use is the <a href="https://addons.mozilla.org/nl/firefox/addon/downthemall/">DownThemAll</a> addon for Firefox, in which you have to provide the checksum at the time of download. It will automatically check whether the download is finished.
The Short Read Archive (SRA), storing NGS data sets, makes use of <a href="http://asperasoft.com/technology/transport/fasp/">Aspera</a> to download data a great speeds, ensuring integrity. To download from SRA using aspera in linux, follow the <a href="http://www.ebi.ac.uk/ena/about/read_download">this guide from EBI</a>.</p>

<h3 id="extracting-the-data">Extracting the data</h3>
<blockquote>
  <p>What type of file have you downloaded?</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ file IRGSPb5.fa.masked.gz
IRGSPb5.fa.masked.gz: gzip compressed data, was "IRGSPb5.fa.masked", from Unix, last modified: Wed Aug 18 03:45:47 2010
</code></pre></div>  </div>
</blockquote>

<p>It is a compressed file. Files are compressed to save storage space. Before using these files, you have to decompress them. What can you do with this type of file? Check the command apropos.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ apropos gzip
gzip (1)             - compress or expand files
lz (1)               - gunzips and shows a listing of a gzip'd tar'd archive
tgz (1)              - makes a gzip'd tar archive
uz (1)               - gunzips and extracts a gzip'd tar'd archive
zforce (1)           - force a '.gz' extension on all gzip files
</code></pre></div></div>

<p><strong>apropos</strong> is a command that helps you discover new commands. In case you have a type of file that you don’t know about, use apropos to search for corresponding programs.</p>
<blockquote>
  <p>Decompress the file. Check the man page of gzip. From the man page:&lt;pre&gt;gunzip [ -acfhlLnNrtvV ] [-S suffix] [ name …  ]&lt;/pre&gt;</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ gunzip IRGSPb5.fa.masked.gz 
$ ls
IRGSPb5.fa.masked  IRGSPb5.fa.masked.gz.md5
</code></pre></div>  </div>
</blockquote>




                    
                    <blockquote class="key_points">
                        <h3><i class="fa fa-key" aria-hidden="true"></i><span class="visually-hidden">keypoints</span> Key points</h3>

                        <ul>
                            
                            <li>Where can I find command line information</li>
                            
                            <li>What are useful commands</li>
                            
                        </ul>
                    </blockquote>
                    

                    

                    

                </div>
                <!-- sidebar, which will be hidden on a small screen -->
                <nav class="col-md-2 sticky-toc bs-docs-sidebar hidden-print hidden-xs hidden-sm" role="complementary" id="toc" data-toggle="toc"></nav>
            </div>
        </div>

        <h3><i class="fa fa-thumbs-up" aria-hidden="true"></i><span class="visually-hidden">congratulations</span> Congratulations on successfully completing this tutorial!</h3>

        

        

</div>
</div>
</div>
<footer>
   <div class="well row">
       <div class="col-md-2 text-center" style="padding: 25px 20px 25px 20px;">				
         <a href="https://www.bits.vib.be" target="_blank"><img src="/assets/images/bioinformatics_core_rgb_neg.png" width="200"></a>
       </div>
       <div class="col-md-8 text-center" style="padding-top: 20px;">	
        <a href="https://github.com/vibbits/training-material" title="View on GitHub" class="margin20" target="_blank">
                            <i class="fa fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> View on GitHub
                        </a>
       </div>
       <div class="col-md-2 text-center" style="padding-top: 30px;">
         <span class="slogan"><br>SCIENCE MEETS LIFE</span>
       </div>
   </div>
</footer>


</body>
<script type="text/javascript" src="/assets/js/jquery.slim.min.js"></script>
<script type="text/javascript" src="/assets/js/popper.min.js"></script>
<script type="text/javascript" src="/assets/js/bootstrap.min.js?v=3"></script>
<script type="text/javascript" src="/assets/js/details-element-polyfill.js"></script>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script type="text/javascript" src="/assets/js/bootstrap-toc.min.js"></script>
<script type="text/javascript" src="/assets/js/main.js"></script>
<script type="text/javascript" src="/assets/js/clipboard.min.js"></script>
<script type="text/javascript">
    var snippets = document.querySelectorAll('div.highlight');
    [].forEach.call(snippets, function(snippet) {
        snippet.firstChild.insertAdjacentHTML('beforebegin', '<button class="btn btn-light" data-clipboard-snippet><i class="fa fa-copy"></i>&nbsp;Copy</button>');
    });

    var clipboardSnippets = new ClipboardJS('[data-clipboard-snippet]', {
        target: function(trigger) {
            return trigger.nextElementSibling;
        }
    });
</script>

</html>
