--- ---
<!doctype html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="google-site-verification" content="dw_1zBYRNxF7-vYNDXh3Ght_JuqgJt3BD1N_Pynqmv0" />
    <title>03 Compare Structures</title>
    <script async="" src="https://www.google-analytics.com/analytics.js"></script>
    <script>
        (function(i, s, o, g, r, a, m) {
            i['GoogleAnalyticsObject'] = r;
            i[r] = i[r] || function() {
                (i[r].q = i[r].q || []).push(arguments)
            }, i[r].l = 1 * new Date();
            a = s.createElement(o),
                m = s.getElementsByTagName(o)[0];
            a.async = 1;
            a.src = g;
            m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');
        ga('create', 'UA-61194136-18', 'auto');
        ga('send', 'pageview');
    </script>
    <link rel="stylesheet" href="/assets/css/bootstrap.min.css?v=3">
    <link rel="stylesheet" href="/assets/css/bootstrap-toc.min.css">
    <link rel="stylesheet" href="/assets/css/font-awesome.css">
    <link rel="stylesheet" href="/assets/css/main.scss">
    <link rel="stylesheet" href="/assets/css/VIB.css">
    <link rel="stylesheet" href="/assets/css/academicons.css">
    <link rel="stylesheet" href="/assets/css/syntax_highlighting.css">
    <link rel="shortcut icon" href="/favicon.ico " type="image/x-icon" />     
    <meta name="description" content="Analyzing the protein structure of your protein-of-intere..." />
    <meta property="og:title" content="VIB Bioinformatics Training: 03 Compare Structures" />
    <meta property="og:description" content="Analyzing the protein structure of your protein-of-intere..." />
    <meta property="og:image" content="/assets/images/bioinformatics_core_rgb_neg.png" />
</head>

<body data-spy="scroll" data-target="#toc">
    <div class="top-bar-vib">
 <a href="http://vib.be/en" target="_blank" style="font-size:14px">VIB Home </a>
</div>

<div class="container main-content">

<div class="container page-top__container">
          <div class="region region-top">
  <div id="logo" class="col-md-3 col-lg-3 text-left" style="padding-left:0px;margin-top:20px">
       <a href="/">
           <img src="/assets/images/logo.svg" height="60" alt="VIB logo" class="margin20">
       </a>
    </div>
    <div class="col-md-9 col-lg-9 text-right" style="margin-top:50px;">
    <nav role="navigation" aria-labelledby="block-mainnavigation-menu" id="block-mainnavigation" class="block block-menu navigation menu--main">
  


        <a href="/" class="topmenu">Home</a>


        <a href="/about.html" class="topmenu" >About</a>


        <a href="/hall-of-fame.html" class="topmenu">Contribute</a>


        <a href="/faq.html" class="topmenu">Faq</a>


        <a href="/search/index.html" class="topmenu">Search</a>


  </nav>

    </div>
  </div>
</div>














    <script type="application/ld+json">
        


{
  "@context": "http://schema.org",
  "@type": "Course",
  "accessMode": [
    "textual",
    "visual"
  ],
  "accessModeSufficient": [
    "textual",
    "visual"
  ],
  "accessibilityControl": [
    "fullKeyboardControl",
    "fullMouseControl"
  ],
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "accessibilitySummary": "Short descriptions are present but long descriptions will be needed for non-visual users",
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "copyrightHolder": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "headline": "03 Compare Structures",
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "license": "/blob//LICENSE.md",
  "producer": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "provider": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "sourceOrganization": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "courseCode": "protein-structure-analysis / compare-structures / hands-on",
  "learningResourceType": "hands-on tutorial",
  "name": "Hands-on for '03 Compare Structures' tutorial",
  "isPartOf": {
    "@type": "Course",
    "name": "03 Compare Structures",
    "description": "03 Compare Structures",
    "learningResourceType": "tutorial",
    "interactivityType": "expositive",
    "provider": {
      "@type": "Organization",
      "email": "bits@vib.be",
      "name": "VIB Bioinformatics Core",
      "url": "https://www.bits.vib.be"
    }
  },
  "url": "https://material.bits.vib.be//topics/protein-structure-analysis/tutorials/compare-structures/tutorial.html",
  "timeRequired": "PT15M",
  "description": "03 Compare Structures\\nThe questions this  addresses are:\n - how to structurally align protein structures\n\n\\nThe objectives are:\n - Use rigid body structural alignment in YASARA\n\n",
  "hasPart": [

  ],
  "mentions": [
    {
      "@type": "Thing",
      "url": "",
      "name": "Training data for 03 Compare Structures tutorial"
    }
  ],
  "author": [
    {
      "@type": "Person",
      "name": "Alexander Botzki"
    },
    {
      "@type": "Person",
      "name": "Joost Van Durme"
    }
  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "Alexander Botzki"
    },
    {
      "@type": "Person",
      "name": "Joost Van Durme"
    }
  ],
  "about": [
    {
      "@type": "CreativeWork",
      "name": "Protein Structure Analysis",
      "description": "Analyzing the protein structure of your protein-of-interest can be advantageous in multiple ways. It can help you discover regions which are good candidates to interact with other proteins. It can help you discover new domains. It can help with identifying differences with homologuous proteins and a lot more.",
      "url": "https://material.bits.vib.be//topics/protein-structure-analysis/"
    }
  ]
}
    </script>

    <div class="row tutorial">
      <div class="col-md-12" role="main">
        <h1 data-toc-skip>03 Compare Structures</h1>
        <div class="contributors-line">By: 

<a href="/hall-of-fame#abotzki" class="contributor-badge">Alexander Botzki</a>
, <a href="/hall-of-fame#jvdurme" class="contributor-badge">Joost Van Durme</a>


</div>
        <blockquote class="overview">

            <strong><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</strong>
            <ul>
            
            <li>how to structurally align protein structures</li>
            
            </ul>

            <strong><i class="fa fa-bullseye" aria-hidden="true"></i><span class="visually-hidden">objectives</span> Objectives</strong>
            <ul>
            
            <li>Use rigid body structural alignment in YASARA</li>
            
            </ul>

            

            
            <p><strong><i class="fa fa-hourglass-end" aria-hidden="true"></i><span class="visually-hidden">time</span> Time estimation:</strong> 15 minutes</p>
            

            


            

        </blockquote>

        <div class="container" data-field="body">
            <div class="row">
                <div class="col-md-10" role="main">
                    <h2 id="structural-comparison-and-rmsd">Structural comparison and RMSD</h2>
<p>We compare structures by structurally aligning them on top of each other. That is, we
align structurally equivalent atoms. For now, we will only use CA atoms as a representation of the backbones. 
But Yasara also can align on any type of atom you want. You always need to specify:</p>

<ul>
  <li>source object(s): the structure(s) that needs to be rotated and translated to superpose on anoth
er structure</li>
  <li>target object: the structure to superpose on</li>
</ul>

<p>An optimal alignment is found when the root-mean-square deviation (RMSD) is at a minimum. 
The RMSD is given as:</p>
<figure id="figure-1"><img src="../../images/RMSD.gif" alt="RMSD" /><figcaption><span class="figcaption-prefix">Figure 1:</span> calculation of RMSD</figcaption></figure>
<p>where R is the distance between two structurally equivalent atom pairs (CA in our case) and n is the total number of atom pairs.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-data-download"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Data download</h3>

  <ol>
    <li>
      <p>Download the following adapted PDB files from <a href="https://zenodo.org/record/3550492#.XdeNL1dKiUk">Zenodo</a></p>

      <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code> 1DKX_1.pdb 1DKY_1.pdb 1DKZ_1.pdb 3DPO_1.pdb 3DPP_1.pdb 
</code></pre></div>      </div>
    </li>
  </ol>

</blockquote>

<h2 id="aligning-multiple-structures-using-yasara">Aligning multiple structures using YASARA</h2>

<p>Now load all of them in YASARA:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>File &gt; Load &gt; PDB File
</code></pre></div></div>

<p>and select the CA (C-alpha) view (F4) and superpose with the MUSTANG algorithm:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Analyze &gt; Align &gt; Objects with MUSTANG
</code></pre></div></div>

<p>In the first window you have to select the source objects that will be repositioned. Select Objects 2 till 5. In the second window you select the target Object to superpose on. That would then be the first object.</p>

<p>Notice that YASARA prints the RMSD of every structural alignment in the lower Console. Open the Console by pressing the spacebar once or twice to extend it.</p>

<p>Color the atoms by their B-factor:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>View &gt; Color &gt; Atom &gt; Belongs to or has &gt; All
Then choose BFactor in the next window and press 'Apply unique color'.
</code></pre></div></div>

<p>High BFactors are yellow, low BFactors are blue.</p>

<h3 id="question-questions"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</h3>

<p>Do you see a correlation between the BFactors and the variability in the structure?</p>

<blockquote>
  <details>
    <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>

    <p><img src="../../images/aligned-structures.png" alt="Structural alignemnt" /></p>

  </details>
</blockquote>

<h1 class="no_toc" id="conclusion">Conclusion</h1>

<p>Structural alignment of related structures is a very efficient approach to spot similarities and differences of structutally related proteins.</p>



                    
                    <blockquote class="key_points">
                        <h3><i class="fa fa-key" aria-hidden="true"></i><span class="visually-hidden">keypoints</span> Key points</h3>

                        <ul>
                            
                            <li>Superimposition of related structures is a very efficient approach to spot similarities and differences of structutally related proteins.</li>
                            
                        </ul>
                    </blockquote>
                    

                    
                    <h1 data-toc-skip>Useful literature</h1>
                    <p>Further information, including links to documentation and original publications, regarding the tools, analysis techniques and the interpretation of results described in this tutorial can be found <a href="/topics/protein-structure-analysis#references">here</a>.</p>
                    

                    

                </div>
                <!-- sidebar, which will be hidden on a small screen -->
                <nav class="col-md-2 sticky-toc bs-docs-sidebar hidden-print hidden-xs hidden-sm" role="complementary" id="toc" data-toggle="toc"></nav>
            </div>
        </div>

        <h3><i class="fa fa-thumbs-up" aria-hidden="true"></i><span class="visually-hidden">congratulations</span> Congratulations on successfully completing this tutorial!</h3>

        

        

</div>
</div>
</div>
<footer>
   <div class="well row">
       <div class="col-md-2 text-center" style="padding: 25px 20px 25px 20px;">				
         <a href="https://www.bits.vib.be" target="_blank"><img src="/assets/images/bioinformatics_core_rgb_neg.png" width="200"></a>
       </div>
       <div class="col-md-8 text-center" style="padding-top: 20px;">	
        <a href="https://github.com/vibbits/training-material" title="View on GitHub" class="margin20" target="_blank">
                            <i class="fa fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> View on GitHub
                        </a>
       </div>
       <div class="col-md-2 text-center" style="padding-top: 30px;">
         <span class="slogan"><br>SCIENCE MEETS LIFE</span>
       </div>
   </div>
</footer>


</body>
<script type="text/javascript" src="/assets/js/jquery.slim.min.js"></script>
<script type="text/javascript" src="/assets/js/popper.min.js"></script>
<script type="text/javascript" src="/assets/js/bootstrap.min.js?v=3"></script>
<script type="text/javascript" src="/assets/js/details-element-polyfill.js"></script>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script type="text/javascript" src="/assets/js/bootstrap-toc.min.js"></script>
<script type="text/javascript" src="/assets/js/main.js"></script>
<script type="text/javascript" src="/assets/js/clipboard.min.js"></script>
<script type="text/javascript">
    var snippets = document.querySelectorAll('div.highlight');
    [].forEach.call(snippets, function(snippet) {
        snippet.firstChild.insertAdjacentHTML('beforebegin', '<button class="btn btn-light" data-clipboard-snippet><i class="fa fa-copy"></i>&nbsp;Copy</button>');
    });

    var clipboardSnippets = new ClipboardJS('[data-clipboard-snippet]', {
        target: function(trigger) {
            return trigger.nextElementSibling;
        }
    });
</script>

</html>