--- ---
<!doctype html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="google-site-verification" content="dw_1zBYRNxF7-vYNDXh3Ght_JuqgJt3BD1N_Pynqmv0" />
    <title>01 Exploring the Protein Data Bank exercises</title>
    <script async="" src="https://www.google-analytics.com/analytics.js"></script>
    <script>
        (function(i, s, o, g, r, a, m) {
            i['GoogleAnalyticsObject'] = r;
            i[r] = i[r] || function() {
                (i[r].q = i[r].q || []).push(arguments)
            }, i[r].l = 1 * new Date();
            a = s.createElement(o),
                m = s.getElementsByTagName(o)[0];
            a.async = 1;
            a.src = g;
            m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');
        ga('create', 'UA-61194136-18', 'auto');
        ga('send', 'pageview');
    </script>
    <link rel="stylesheet" href=" /assets/css/bootstrap.min.css?v=3 ">
    <link rel="stylesheet" href=" /assets/css/bootstrap-toc.min.css ">
    <link rel="stylesheet" href=" /assets/css/font-awesome.css ">
    <link rel="stylesheet" href=" /assets/css/main.css ">
    <link rel="stylesheet" href=" /assets/css/VIB.css ">
    <link rel="stylesheet" href=" /assets/css/academicons.css ">
    <link rel="stylesheet" href=" /assets/css/syntax_highlighting.css ">
    <link rel="shortcut icon" href=" /favicon.ico " type="image/x-icon" />     
    <meta name="description" content="Analyzing the protein structure of your protein-of-intere..." />
    <meta property="og:title" content="VIB Bioinformatics Training: 01 Exploring the Protein Data Bank exercises" />
    <meta property="og:description" content="Analyzing the protein structure of your protein-of-intere..." />
    <meta property="og:image" content=" /assets/images/bioinformatics_core_rgb_neg.png " />
</head>

<body data-spy="scroll" data-target="#toc">
    <div class="top-bar-vib">
 <a href="http://vib.be/en" target="_blank" style="font-size:14px">VIB Home </a>
</div>

<div class="container main-content">

<div class="container page-top__container">
          <div class="region region-top">
  <div id="logo" class="col-md-3 col-lg-3 text-left" style="padding-left:0px;margin-top:20px">
       <a href="/">
           <img src="/assets/images/logo.svg" height="60" alt="VIB logo" class="margin20">
       </a>
    </div>
    <div class="col-md-9 col-lg-9 text-right" style="margin-top:50px;">
    <nav role="navigation" aria-labelledby="block-mainnavigation-menu" id="block-mainnavigation" class="block block-menu navigation menu--main">
  


        <a href="/" class="topmenu">Home</a>


        <a href="/about.html" class="topmenu" >About</a>


        <a href="/hall-of-fame.html" class="topmenu">Contribute</a>


        <a href="/faq.html" class="topmenu">Faq</a>


        <a href="/search/index.html" class="topmenu">Search</a>


  </nav>

    </div>
  </div>
</div>














    <script type="application/ld+json">
        


{
  "@context": "http://schema.org",
  "@type": "Course",
  "accessMode": [
    "textual",
    "visual"
  ],
  "accessModeSufficient": [
    "textual",
    "visual"
  ],
  "accessibilityControl": [
    "fullKeyboardControl",
    "fullMouseControl"
  ],
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "accessibilitySummary": "Short descriptions are present but long descriptions will be needed for non-visual users",
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "copyrightHolder": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "headline": "01 Exploring the Protein Data Bank exercises",
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "license": "/blob//LICENSE.md",
  "producer": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "provider": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "sourceOrganization": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "courseCode": "protein-structure-analysis / explore-pdb / hands-on",
  "learningResourceType": "hands-on tutorial",
  "name": "Hands-on for '01 Exploring the Protein Data Bank exercises' tutorial",
  "isPartOf": {
    "@type": "Course",
    "name": "01 Exploring the Protein Data Bank exercises",
    "description": "01 Exploring the Protein Data Bank exercises",
    "learningResourceType": "tutorial",
    "interactivityType": "expositive",
    "provider": {
      "@type": "Organization",
      "email": "bits@vib.be",
      "name": "VIB Bioinformatics Core",
      "url": "https://www.bits.vib.be"
    }
  },
  "url": "https://material.bits.vib.be//topics/protein-structure-analysis/tutorials/explore-pdb/tutorial.html",
  "timeRequired": "PT30M",
  "description": "01 Exploring the Protein Data Bank exercises\\nThe questions this  addresses are:\n - How to find structures in the PDB database?\n\n\\nThe objectives are:\n - Basic knowledge of working with the PDB database\n\n",
  "hasPart": [

  ],
  "mentions": [
    {
      "@type": "Thing",
      "url": "https://data.bits.vib.be/pub/trainingen/PSA/",
      "name": "Training data for 01 Exploring the Protein Data Bank exercises tutorial"
    }
  ],
  "author": [
    {
      "@type": "Person",
      "name": "Alexander Botzki"
    },
    {
      "@type": "Person",
      "name": "Joost Van Durme"
    }
  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "Alexander Botzki"
    },
    {
      "@type": "Person",
      "name": "Joost Van Durme"
    }
  ],
  "about": [
    {
      "@type": "CreativeWork",
      "name": "Protein Structure Analysis",
      "description": "Analyzing the protein structure of your protein-of-interest can be advantageous in multiple ways. It can help you discover regions which are good candidates to interact with other proteins. It can help you discover new domains. It can help with identifying differences with homologuous proteins and a lot more.",
      "url": "https://material.bits.vib.be//topics/protein-structure-analysis/"
    }
  ]
}
    </script>

    <div class="row tutorial">
      <div class="col-md-12" role="main">
        <h1 data-toc-skip>01 Exploring the Protein Data Bank exercises</h1>
        <div class="contributors-line">By: 

<a href="/hall-of-fame#abotzki" class="contributor-badge">Alexander Botzki</a>
, <a href="/hall-of-fame#jvdurme" class="contributor-badge">Joost Van Durme</a>


</div>
        <blockquote class="overview">

            <strong><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</strong>
            <ul>
            
            <li>How to find structures in the PDB database?</li>
            
            </ul>

            <strong><i class="fa fa-bullseye" aria-hidden="true"></i><span class="visually-hidden">objectives</span> Objectives</strong>
            <ul>
            
            <li>Basic knowledge of working with the PDB database</li>
            
            </ul>

            

            
            <p><strong><i class="fa fa-hourglass-end" aria-hidden="true"></i><span class="visually-hidden">time</span> Time estimation:</strong> 30 minutes</p>
            

            


            

        </blockquote>

        <div class="container" data-field="body">
            <div class="row">
                <div class="col-md-10" role="main">
                    <blockquote class="agenda">
  <h3 id="agenda">Agenda</h3>

  <p>In this tutorial, we will deal with:</p>

<ol id="markdown-toc">
  <li><a href="#search-for-a-structure" id="markdown-toc-search-for-a-structure">Search for a structure</a></li>
  <li><a href="#the-pdb-file" id="markdown-toc-the-pdb-file">The PDB file</a>    <ol>
      <li><a href="#introduction" id="markdown-toc-introduction">Introduction</a></li>
      <li><a href="#downloading-the-structure" id="markdown-toc-downloading-the-structure">Downloading the structure</a></li>
    </ol>
  </li>
</ol>

</blockquote>

<h1 id="search-for-a-structure">Search for a structure</h1>
<h4 id="via-uniprot">Via <a href="http://www.uniprot.org/">UniProt</a></h4>
<p>The way of searching for a specific protein structure depends on the data you already have. You might already have the PDB ID (a unique identifier), that’s an easy one. But mostly you have the protein name or you just have a sequence. In the last cases I recommend to start from the UniProt website at <a href="http://www.uniprot.org">http://www.uniprot.org</a>, which is the best annotated protein database in the world. Our first model protein will be the molecular chaperone DnaK from <em>E. coli</em>. Below is an image of the UniProt search box where you can start your search for proteins.</p>

<figure id="figure-1"><img src="../../images/uniprotsearchbox.png" alt="uniprotsearchbox.png" /><figcaption><span class="figcaption-prefix">Figure 1:</span> Search box</figcaption></figure>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-explore-a-pdb-structure-on-the-uniprot-web-site"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Explore a PDB structure on the Uniprot web site</h3>

  <ol>
    <li>Go to the UniProt website and search for the DnaK protein
      <ul>
        <li>The UniProt search engine returns a list of DnaK protein sequences from a variety of organisms. An entry with accession code <strong>P0A6Y8</strong> and entry name <strong>DNAK_ECOLI</strong> should be near the top of this list.</li>
      </ul>
    </li>
    <li>Click on the <em>accession code</em> (column Entry) to view the protein page of this DnaK from the model organism <em>Escherichia coli</em>.</li>
    <li>Click on <em>Structure</em> in the left-side menu and then look at the <em>3D structure databases</em> table.</li>
  </ol>
</blockquote>

<blockquote class="question">
  <h3 id="question-guidelines-which-pdb-structures-to-select"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Guidelines which PDB structures to select</h3>

  <p>Which structures (give the 4-character PDB ID) of the C-terminal domain of DnaK should preferentially be use for analysis and why?</p>
  <blockquote class="solution">
    <details>
      <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
      <p>Usually, the recommended selection criteria are using an X-ray structure with low resolution and low Rfree factor. Furthermore, the PDB database has pre-calculated a validation report for all of the structures.
As an example, have a look at https://www.ebi.ac.uk/pdbe/entry/pdb/4EZX under the section ‘Experiments and Validation’. For many PDB structures, there is also a re-done structure available with a vast amount of informaton on the quality of the X-ray structure and suggested ‘better’ models e.g. (https://pdb-redo.eu/db/4ezx). In our case, we could opt for the structures 1DKX and 4EZX.</p>

      <p>This is a difficult example since there are so many high resolution structures available. So, it is recommended to study the articles and compare the available structures to find your favorite structure for further analysis.</p>
    </details>
  </blockquote>
</blockquote>

<h4 id="via-the-protein-data-bank-by-pdb-id">Via the Protein Data Bank by PDB ID</h4>

<iframe src="https://h5p.org/h5p/embed/577970" width="699" height="418" frameborder="0" allowfullscreen="allowfullscreen"></iframe>
<script src="https://h5p.org/sites/all/modules/h5p/library/js/h5p-resizer.js" charset="UTF-8"></script>

<iframe src=" https://docs.google.com/presentation/d/1lgEeigU8M45xF2BjQgPumgxalzfiHZeD/preview" width="640" height="480"></iframe>

<p>You can find structural information directly at the PDB database. The web site of the PDB consortium is located at <a href="http://www.wwpdb.org">http://www.wwpdb.org</a>. This web site provides links to all members of the PDB (left side). It is a question of taste which resource you start off with. For X-ray structures, it is currently PDBe, RCSB PDB, PDBj. For NMR structres, you find the BMRB. In today’s course, we focus on the PDB resources only.</p>

<p>Below is an image of the RCSB search box <a href="http://www.rcsb.org/pdb/home/home.do">http://www.rcsb.org/pdb/home/home.do</a> where you can start your search for structures.</p>

<p><img src="../../images/Pdbsearchbox_RCSB.png" alt="Pdbsearchbox_RCSB.png" /></p>

<p>The PDB file with ID <strong>1DKX</strong> contains the atomic coordinates of the molecular chaperone (DnaK) from <em>E. coli</em>.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-search-a-structure-on-the-rcsb-web-site"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Search a structure on the RCSB web site</h3>

  <ol>
    <li>Go to the PDB website and type 1DKX in the search box</li>
  </ol>
</blockquote>

<p>This will lead you to the same page we got earlier through UniProt.</p>

<h4 id="via-the-protein-data-bank-by-sequence">Via the Protein Data Bank by sequence</h4>

<p>In lots of cases we only have a sequence of which we would like to find out if there is structural information. The PDB can be searched using a sequence as input. Here is the sequence of the C-terminal substrate binding domain of DnaK:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    DVKDVLLLDVTPLSLGIETMGGVMTTLIAKNTTIPTKHSQVFSTAEDNQSAVTIHVLQGE
    RKRAADNKSLGQFNLDGINPAPRGMPQIEVTFDIDADGILHVSAKDKNSGKEQKITIKAS
    SGLNEDEIQKMVRDAEANAEADRKFEELVQTRNQGDHLLHSTRKQVEEAGDKLPADDKTA
    IESALTALETALKGEDKAAIEAKMQELAQVSQKLMEIAQQQHAQQQTAGADASANNAKDD
    DVVDAEFEEVKDKK
</code></pre></div></div>
<p>The PDB allows sequence searches through the same search box we used before.</p>

<p><img src="../../images/Pdbsearchbox_RCSB.png" alt="Pdbsearchbox_RCSB.png" /></p>

<p>There is also an Advanced Search section, with a Blast/Fasta option in the Sequence Features section.</p>

<p><img src="../../images/Blastpdb.png" alt="Blastpdb.png" /></p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-blast-search-for-pdb-structure"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: BLAST search for PDB structure</h3>

  <ol>
    <li>Go to the Advanced Search section</li>
    <li>Please select ‘Sequence BLAST/PSI-BLAST’ in the Query type drop down.
This method allows you to change some parameters for the search.</li>
    <li>Copy and paste the sequence in the ‘‘Sequence’’ field</li>
    <li>Press ‘‘Submit query’’.</li>
    <li>You should see the same structures popping up as you saw in the UniProt page of DnaK.</li>
  </ol>
</blockquote>

<h1 id="the-pdb-file">The PDB file</h1>

<h2 id="introduction">Introduction</h2>

<p>A PDB (Protein Data Bank) file is a plain text file that contains the
atom coordinates of a solved 3D structure of a protein or even DNA. Such
coordinate files can be obtained at the Protein Data Bank at
<a href="http://www.rcsb.org/pdb">http://www.rcsb.org/pdb</a>. Each PDB file has a unique identifier (ID)
consisting of 4 characters, the first one is always a number. Note: It
has been announced that the 4 character code will change in the future
<a href="https://www.wwpdb.org/news/news?year=2017\#5910c8d8d3b1d333029d4ea8">https://www.wwpdb.org/news/news?year=2017\#5910c8d8d3b1d333029d4ea8</a>.</p>

<p>The PDB file with ID <strong>1DKX</strong> contains the atomic coordinates of the
molecular chaperone (DnaK) from <em>E coli</em>.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-blast-search-for-pdb-structure-1"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: BLAST search for PDB structure</h3>

  <ol>
    <li>Go to the PDB website at <a href="http://www.rcsb.org/pdb">http://www.rcsb.org/pdb</a></li>
    <li>Type 1DKX in the search and try to answer the following questions.</li>
  </ol>
</blockquote>

<blockquote class="question">
  <h3 id="question-questions"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</h3>

  <ol>
    <li>How many molecules were solved in this PDB file? What kind of molecules are these (proteins, peptides, DNA, …)?</li>
    <li>Does the structure represent the full protein? If not, how many residues are missing? Hint: Click on the UniProt KB link in the Sequence tab to see the full sequence.</li>
    <li>Was this structure solved by X-Ray or NMR?</li>
    <li>What is the atomic resolution and R-factor?</li>
  </ol>

  <blockquote>
    <details>
      <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
      <ol>
        <li>Two, called polymers or chains: they are polypeptides <img src="../../images/Mol_desc_1DKX.png" alt="Type" /></li>
        <li>To answer this question you can go to the sequence tab at the top:
          <ul>
            <li><img src="../../images/Pdb_firstresiduesmissing_1dkx.png" alt="Uniprot view" /></li>
            <li>Summary: a large chunk of the N-terminus is missing from the structure, the C-terminus is virtually complete.</li>
          </ul>
        </li>
        <li>X-RAY diffraction, as shown by Experimental Details</li>
        <li>Atomic resolution: 2.00 Ångstrom and R-factor of 0.206</li>
      </ol>
    </details>
  </blockquote>

</blockquote>

<h2 id="downloading-the-structure">Downloading the structure</h2>

<p>The file that holds the 3D coordinates can be downloaded by clicking on
<em>Download files</em> in the top right corner and then choosing <em>PDB file (text)</em>.
For convenience, save this file on your desktop. The filename is the
4-character unique PDB ID.</p>

<p><img src="../../images/Pdbdownloadfile1.png" alt="Pdbdownloadfile1.png" /></p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-open-downloaded-pdb-file-in-an-editor"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Open downloaded PDB file in an editor</h3>
  <ol>
    <li>Open this file with a text editor, e.g. WordPad is an excellent tool for that.</li>
    <li>Do you see the different sections in the PDB file? Analyse some ATOM lines and try to explain what kind of data is in each column.</li>
  </ol>
</blockquote>

<p>Additional exercises on searching PDB can be found on <a href="http://wiki.bits.vib.be/index.php/Exercises_on_Protein_Structure">the basic bioinformatics exercises page</a>.</p>



                    
                    <blockquote class="key_points">
                        <h3><i class="fa fa-key" aria-hidden="true"></i><span class="visually-hidden">keypoints</span> Key points</h3>

                        <ul>
                            
                            <li>Protein ID</li>
                            
                            <li>Uniprot</li>
                            
                            <li>PDB file format</li>
                            
                        </ul>
                    </blockquote>
                    

                    
                    <h1 data-toc-skip>Useful literature</h1>
                    <p>Further information, including links to documentation and original publications, regarding the tools, analysis techniques and the interpretation of results described in this tutorial can be found <a href="/topics/protein-structure-analysis#references">here</a>.</p>
                    

                    

                </div>
                <!-- sidebar, which will be hidden on a small screen -->
                <nav class="col-md-2 sticky-toc bs-docs-sidebar hidden-print hidden-xs hidden-sm" role="complementary" id="toc" data-toggle="toc"></nav>
            </div>
        </div>

        <h3><i class="fa fa-thumbs-up" aria-hidden="true"></i><span class="visually-hidden">congratulations</span> Congratulations on successfully completing this tutorial!</h3>

        

        

</div>
</div>
</div>
<footer>
   <div class="well row">
       <div class="col-md-2 text-center" style="padding: 25px 20px 25px 20px;">				
         <a href="https://www.bits.vib.be" target="_blank"><img src="/assets/images/bioinformatics_core_rgb_neg.png" width="200"></a>
       </div>
       <div class="col-md-8 text-center" style="padding-top: 20px;">	
        <a href="https://github.com/vibbits/training-material" title="View on GitHub" class="margin20" target="_blank">
                            <i class="fa fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> View on GitHub
                        </a>
       </div>
       <div class="col-md-2 text-center" style="padding-top: 30px;">
         <span class="slogan"><br>SCIENCE MEETS LIFE</span>
       </div>
   </div>
</footer>


</body>
<script type="text/javascript" src=" /assets/js/jquery.slim.min.js "></script>
<script type="text/javascript" src=" /assets/js/popper.min.js "></script>
<script type="text/javascript" src=" /assets/js/bootstrap.min.js?v=3 "></script>
<script type="text/javascript" src=" /assets/js/details-element-polyfill.js "></script>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script type="text/javascript" src=" /assets/js/bootstrap-toc.min.js "></script>
<script type="text/javascript" src=" /assets/js/main.js "></script>
<script type="text/javascript" src=" /assets/js/clipboard.min.js "></script>
<script type="text/javascript" src="https://app.cloudsh.com/js/cloudsh.js"></script>
<script type="text/javascript">
    var snippets = document.querySelectorAll('div.highlight');
    [].forEach.call(snippets, function(snippet) {
        snippet.firstChild.insertAdjacentHTML('beforebegin', '<button class="btn btn-light" data-clipboard-snippet><i class="fa fa-copy"></i>&nbsp;Copy</button>');
    });

    var clipboardSnippets = new ClipboardJS('[data-clipboard-snippet]', {
        target: function(trigger) {
            return trigger.nextElementSibling;
        }
    });
</script>

</html>