<!doctype html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="google-site-verification" content="dw_1zBYRNxF7-vYNDXh3Ght_JuqgJt3BD1N_Pynqmv0" />
    <title>03 Linux file system</title>
    <script async="" src="https://www.google-analytics.com/analytics.js"></script>
    <script>
        (function(i, s, o, g, r, a, m) {
            i['GoogleAnalyticsObject'] = r;
            i[r] = i[r] || function() {
                (i[r].q = i[r].q || []).push(arguments)
            }, i[r].l = 1 * new Date();
            a = s.createElement(o),
                m = s.getElementsByTagName(o)[0];
            a.async = 1;
            a.src = g;
            m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');
        ga('create', 'UA-61194136-18', 'auto');
        ga('send', 'pageview');
    </script>
    <link rel="stylesheet" href="/assets/css/bootstrap.min.css?v=3">
    <link rel="stylesheet" href="/assets/css/bootstrap-toc.min.css">
    <link rel="stylesheet" href="/assets/css/font-awesome.css">
    <link rel="stylesheet" href="/assets/css/main.scss">
    <link rel="stylesheet" href="/assets/css/VIB.css">
    <link rel="stylesheet" href="/assets/css/academicons.css">
    <link rel="stylesheet" href="/assets/css/syntax_highlighting.css">
    <link rel="shortcut icon" href="/favicon.ico " type="image/x-icon" />     
    <meta name="description" content="This training is a prerequisite for all Linux novices who..." />
    <meta property="og:title" content="VIB Bioinformatics Training: 03 Linux file system" />
    <meta property="og:description" content="This training is a prerequisite for all Linux novices who..." />
    <meta property="og:image" content="/assets/images/bioinformatics_core_rgb_neg.png" />
</head>

<body data-spy="scroll" data-target="#toc">
    <div class="top-bar-vib">
 <a href="http://vib.be/en" target="_blank" style="font-size:14px">VIB Home </a>
</div>

<div class="container main-content">

<div class="container page-top__container">
          <div class="region region-top">
  <div id="logo" class="col-md-3 col-lg-3 text-left" style="padding-left:0px;margin-top:20px">
       <a href="/">
           <img src="/assets/images/logo.svg" height="60" alt="VIB logo" class="margin20">
       </a>
    </div>
    <div class="col-md-9 col-lg-9 text-right" style="margin-top:50px;">
    <nav role="navigation" aria-labelledby="block-mainnavigation-menu" id="block-mainnavigation" class="block block-menu navigation menu--main">
  


        <a href="/" class="topmenu">Home</a>


        <a href="/about.html" class="topmenu" >About</a>


        <a href="/hall-of-fame.html" class="topmenu">Contribute</a>


        <a href="/faq.html" class="topmenu">Faq</a>


        <a href="/search/index.html" class="topmenu">Search</a>


  </nav>

    </div>
  </div>
</div>














    <script type="application/ld+json">
        


{
  "@context": "http://schema.org",
  "@type": "Course",
  "accessMode": [
    "textual",
    "visual"
  ],
  "accessModeSufficient": [
    "textual",
    "visual"
  ],
  "accessibilityControl": [
    "fullKeyboardControl",
    "fullMouseControl"
  ],
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "accessibilitySummary": "Short descriptions are present but long descriptions will be needed for non-visual users",
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "copyrightHolder": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "headline": "03 Linux file system",
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "license": "/blob//LICENSE.md",
  "producer": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "provider": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "sourceOrganization": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "courseCode": "linux / file-system / hands-on",
  "learningResourceType": "hands-on tutorial",
  "name": "Hands-on for '03 Linux file system' tutorial",
  "isPartOf": {
    "@type": "Course",
    "name": "03 Linux file system",
    "description": "03 Linux file system",
    "learningResourceType": "tutorial",
    "interactivityType": "expositive",
    "provider": {
      "@type": "Organization",
      "email": "bits@vib.be",
      "name": "VIB Bioinformatics Core",
      "url": "https://www.bits.vib.be"
    }
  },
  "url": "https://material.bits.vib.be//topics/linux/tutorials/file-system/tutorial.html",
  "timeRequired": "PT30M",
  "description": "03 Linux file system\\nThe questions this  addresses are:\n - How to get started with the file system in the terminal?\n\n\\nThe objectives are:\n - get familiar with the most essential commands\n - learn what commands to use when navigating the file system\n\n",
  "hasPart": [

  ],
  "mentions": [

  ],
  "author": [
    {
      "@type": "Person",
      "name": "Christof De Bo"
    }
  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "Christof De Bo"
    }
  ],
  "about": [
    {
      "@type": "CreativeWork",
      "name": "Initiation to Linux Command Line",
      "description": "This training is a prerequisite for all Linux novices who want to follow a training that requires the use of Linux command line (e.g. metagenomics or GATK training).",
      "url": "https://material.bits.vib.be//topics/linux/"
    }
  ]
}
    </script>

    <div class="row tutorial">
      <div class="col-md-12" role="main">
        <h1 data-toc-skip>03 Linux file system</h1>
        <div class="contributors-line">By: 

<a href="/hall-of-fame#chdeb" class="contributor-badge">Christof De Bo</a>


</div>
        <blockquote class="overview">

            <strong><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</strong>
            <ul>
            
            <li>How to get started with the file system in the terminal?</li>
            
            </ul>

            <strong><i class="fa fa-bullseye" aria-hidden="true"></i><span class="visually-hidden">objectives</span> Objectives</strong>
            <ul>
            
            <li>get familiar with the most essential commands</li>
            
            <li>learn what commands to use when navigating the file system</li>
            
            </ul>

            

            
            <p><strong><i class="fa fa-hourglass-end" aria-hidden="true"></i><span class="visually-hidden">time</span> Time estimation:</strong> 30 minutes</p>
            

            


            

        </blockquote>

        <div class="container" data-field="body">
            <div class="row">
                <div class="col-md-10" role="main">
                    <h2 id="tutorial-on-the-linux-file-system">Tutorial on the linux file system</h2>

<h3 id="which-protocol-achieves-highest-compression-ratio">Which protocol achieves highest compression ratio?</h3>
<p>Let’s do a little test. Download <a href="http://data.bits.vib.be/pub/trainingmaterial/introduction_to_linux_for_bioinformatics/data_linux_training.tar.gz">this compressed file</a>.</p>

<blockquote>
  <p>Create a folder named ‘Compression_exercise’ in your home. Copy the downloaded tar.gz to it.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ cd
$ mkdir Compression_exercise
$ cp Downloads/data_linux_training.tar.gz Compression_exercise/
</code></pre></div>  </div>
</blockquote>

<blockquote>
  <p>Unpack the data_linux_training.tar.gz  file.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ tar -xvzf data_linux_training.tar.gz 
</code></pre></div>  </div>
</blockquote>

<p>Alternative: you can specify the options without the ‘-‘ sign.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ tar xvfz data_linux_training.tar.gz 
</code></pre></div></div>

<blockquote>
  <p>Decompress the file DRR000542_1.fastq.subset.gz</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ gunzip DRR000542_1.fastq.subset.gz
</code></pre></div>  </div>
</blockquote>

<blockquote>
  <p>Copy the DRR000542_1.fastq.subset file to a new file called ‘bzip2_test.fastq’. Compress this file with bzip2.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ bzip2 bzip2_test.fastq
</code></pre></div>  </div>
</blockquote>

<p><strong>Tip!</strong> If you would like to know how long the command took to finish, use “time”</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ time bzip2 bzip2_test.fastq
real	0m5.878s
user	0m5.728s
sys	0m0.112s
</code></pre></div></div>

<p>Three different times are given. What matters to you is the line ‘real’, also called the wall-clock time.</p>
<blockquote>
  <p>Copy DRR000542_1.fastq.subset file to a new file called gzip_test.fastq and compress with gzip.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ time gzip gzip_test.fastq
real	0m5.878s
user	0m5.728s
sys	0m0.112s
</code></pre></div>  </div>
</blockquote>

<p>A relatively unknown package is lrzip, ‘long range zip’, which achieves very good results on big files. Let’s try that one also!</p>
<blockquote>
  <p>Copy DRR000542_1.fastq.subset file to a new file called lrzip_test.fastq and compress with lrzip.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ lrzip lrzip_test.fastq
The program 'lrzip' is currently not installed.  You can install it by typing:
sudo apt-get install lrzip
</code></pre></div>  </div>
</blockquote>

<p><strong>apt-get</strong> is the command line tool to install software on Debian distro’s. Equivalent to the software center.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ sudo apt-get install lrzip
[sudo] password for joachim: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages were automatically installed and are no longer required:
libnet-ip-perl diffstat libnet-dns-perl libparse-debianchangelog-perl
gir1.2-unique-3.0 kde-l10n-engb python-webpy libnet-domain-tld-perl
libemail-valid-perl libapt-pkg-perl python-flup kde-l10n-zhcn
Use 'apt-get autoremove' to remove them.
The following NEW packages will be installed:
lrzip
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 159 kB of archives.
After this operation, 313 kB of additional disk space will be used.
Get:1 http://be.archive.ubuntu.com/ubuntu/ precise/universe lrzip amd64 0.608-1 [159 kB]
Fetched 159 kB in 0s (780 kB/s) 
Selecting previously unselected package lrzip.
(Reading database ... 662617 files and directories currently installed.)
Unpacking lrzip (from .../lrzip_0.608-1_amd64.deb) ...
Processing triggers for man-db ...
Setting up lrzip (0.608-1) ...
</code></pre></div></div>

<p>Now we can compress:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Output filename is: lrzip_test.fastq.lrz
lrzip_test.fastq - Compression Ratio: 6.724. Average Compression Speed:  0.563MB/s.
Total time: 00:03:02.97
real	3m3.026s
user	3m1.947s
sys	0m0.804s
</code></pre></div></div>

<blockquote>
  <p>Compare the sizes of the different resulting compressed files.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ ls -lh *zip*
-rw------- 1 bits bits 17M Oct 22 14:06 bzip2_test.fastq.bz2
-rw------- 1 bits bits 21M Oct 22 14:06 gzip_test.fastq.gz
-rw------- 1 bits bits 104M Oct 22 14:06 lrzip_test.fastq
-rw------- 1 bits bits 16M Oct 22 14:10 lrzip_test.fastq.lrz
</code></pre></div>  </div>
</blockquote>

<p>Decide for yourself whether the extra time needed for higher compression is worth the gain in compression.</p>
<blockquote>
  <p>Put the three files in a newly created folder ‘results’, and make an archive of it.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ mkdir results
$ mv  *{bz2,q.gz,lrz} results/
$ ls results/
bzip2_test.fastq.bz2  gzip_test.fastq.gz  lrzip_test.fastq.lrz
$ tar cvf results.tar results/
$ rm -rf results/
$ ls -lh
total 281M
-rw------- 1 bits bits 104M May  4  2011 ERX000016.test.fastq
-rw-r--r-- 1 bits bits 21M Oct 22 14:02 ERX000016.test.fastq.tar.gz
-rw------- 1 bits bits 104M Oct 22 14:06 lrzip_test.fastq
-rw-r--r-- 1 bits bits 53M Oct 22 14:28 results.tar
</code></pre></div>  </div>
</blockquote>

<h3 id="symbolic-links">Symbolic links</h3>
<p>Symbolic links (symlinks) point to a file, making the file accessible in another directory than where the file is. So you can avoid copying! When the original file is deleted, the symlink is dead. When you remove the symlink, the original file is still present. 
The syntax for symbolic links is:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ ln -s /home/bits /data/large.fastq /home/bits /Projects/ProjectA/
</code></pre></div></div>

<p>Tip: when using <strong>ln</strong>, preferably provide absolute paths. If you want to use relative paths, make sure first going to the directory you want the link to be in, and create the link using a relative path (using ‘.’ and ‘..’ to make the path).
Removing symbolic links as such:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ unlink /home/bits /Projects/ProjectA
</code></pre></div></div>

<p>In contrast, there is also something as a “hard link” (ln without the -s option). When you delete a hard link, the file to which it referred is gone. So ‘ln -s’ is mostly used.</p>

<h3 id="linking-data-instead-of-copying">Linking data instead of copying</h3>
<p>In the Rice Example directory (should be available under your home): download <a href="http://rice.plantbiology.msu.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir/all.gff3">this annotation file</a> into the ‘Genome data’/’Annotation’ directory. Make a symbolic link to this file in the ‘Genome data’/’Sequence’ directory. Read the first 10 lines from the symbolic link file.</p>
<blockquote>
  <p>When you have tried yourself, see the solution.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ cd Rice\ Example/
~/Rice Example $ ls
bin  Genome data
~/Rice Example $ cd Genome\ data/Annotation/
~/Rice Example/Genome data/Annotation $ ls
~/Rice Example/Genome data/Annotation $ wget http://rice.plantbiology.msu.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir/all.gff3
--2013-10-28 11:45:26--  http://rice.plantbiology.msu.edu/pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir/all.gff3
           =&gt; `all.gff3'
Resolving http://rice.plantbiology.msu.edu (http://rice.plantbiology.msu.edu)... 35.8.196.190
Connecting to http://rice.plantbiology.msu.edu (http://rice.plantbiology.msu.edu)|35.8.196.190|:21... connected.
Logging in as anonymous ... Logged in!
==&gt; SYST ... done.    ==&gt; PWD ... done.
==&gt; TYPE I ... done.  ==&gt; CWD (1) /pub/data/Eukaryotic_Projects/o_sativa/annotation_dbs/pseudomolecules/version_7.0/all.dir ... done.
==&gt; SIZE all.gff3 ... 81498659
==&gt; PASV ... done.    ==&gt; RETR all.gff3 ... done.
Length: 81498659 (78M) (unauthoritative)
100%[======================================&gt;] 81,498,659  1.34M/s   in 65s     
2013-10-28 11:46:33 (1.20 MB/s) - `all.gff3' saved [81498659]
~/Rice Example/Genome data/Annotation $ ls ..
Annotation  Sequence
~/Rice Example/Genome data/Annotation $ cd ../Sequence/
~/Rice Example/Genome data/Sequence $ ln -s ../Annotation/all.gff3 .
~/Rice Example/Genome data/Sequence $ ls -l
total 381300
lrwxrwxrwx 1 bits bits 22 Oct 28 11:49 all.gff3 -&gt; ../Annotation/all.gff3
-rw-r--r-- 1 bits bits 390444160 Mar  8  2013 IRGSPb5.fa.masked
-rw-r--r-- 1 bits bits 55 Mar  8  2013 IRGSPb5.fa.masked.gz.md5
~/Rice Example/Genome data/Sequence $ head all.gff3 
##gff-version 3
Chr1	MSU_osa1r7	gene	2903	10817	.	+	.	ID=LOC_Os01g01010;Name=LOC_Os01g01010;Note=TBC%20domain%20containing%20protein%2C%20expressed
Chr1	MSU_osa1r7	mRNA	2903	10817	.	+	.	ID=LOC_Os01g01010.1;Name=LOC_Os01g01010.1;Parent=LOC_Os01g01010
Chr1	MSU_osa1r7	exon	2903	3268	.	+	.	ID=LOC_Os01g01010.1:exon_1;Parent=LOC_Os01g01010.1
Chr1	MSU_osa1r7	exon	3354	3616	.	+	.	ID=LOC_Os01g01010.1:exon_2;Parent=LOC_Os01g01010.1
Chr1	MSU_osa1r7	exon	4357	4455	.	+	.	ID=LOC_Os01g01010.1:exon_3;Parent=LOC_Os01g01010.1
Chr1	MSU_osa1r7	exon	5457	5560	.	+	.	ID=LOC_Os01g01010.1:exon_4;Parent=LOC_Os01g01010.1
Chr1	MSU_osa1r7	exon	7136	7944	.	+	.	ID=LOC_Os01g01010.1:exon_5;Parent=LOC_Os01g01010.1
Chr1	MSU_osa1r7	exon	8028	8150	.	+	.	ID=LOC_Os01g01010.1:exon_6;Parent=LOC_Os01g01010.1
Chr1	MSU_osa1r7	exon	8232	8320	.	+	.	ID=LOC_Os01g01010.1:exon_7;Parent=LOC_Os01g01010.1
</code></pre></div>  </div>
</blockquote>

<h3 id="introduction-symbolic-links-to-easily-install-manually-applications">Introduction: symbolic links to easily install manually applications</h3>
<p>If a package is not available via a package manager, manual installation might be an option. I put manually applications in ‘’‘/opt’’’. Next, I link them to a correct location on our system, usually ‘’‘/usr/local/bin’’’. Below you have some examples of this, which you can try out yourself.</p>

<p>If you want to manually install apps, ‘’‘/opt’’’ is the advised directory. However, only the administrator (‘root’) can access /opt. You can check that the /opt directory belongs to root with &lt;pre&gt;ls -l /opt&lt;/pre&gt;
To be able to copy and write stuff into /opt, we need root permissions. To do so, precede your commands with ‘'’sudo’’’, as exemplified in the next exercise below. When we do that, our password will first be asked. Next, the command is executed with root permissions. In this way, we can edit contents in root-owned directories! You are a sudoer!</p>

<h3 id="transpose-a-tool-to-transpose">Transpose, a tool to transpose</h3>
<p>Transpose is an extremely convenient text tool to transpose tabular data. We will use it later. The code is hosted on <a href="http://sourceforge.net/projects/transpose/">SourceForge</a>.</p>
<blockquote>
  <p>Download transpose installation file (zip) via the browser. Copy them to /opt using <strong>sudo cp</strong>.
Go to the Sourceforce website with the browser, and click on the Download button.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Downloads $ sudo cp transpose-2.0.zip /opt
[sudo] password for joachim: 
 Downloads $
</code></pre></div>  </div>
</blockquote>

<p>We need to precede the ‘‘cp’’ command with the ‘‘sudo’’ command, since only the root user can copy into ‘‘/opt’’.</p>
<blockquote>
  <p>Unpack the installation in /opt, compile the binary and test it with ‘tranpose –help’. Use sudo to do so.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ pwd
/opt
$ ls trans*
transpose-2.0.zip
$ sudo unzip transpose-2.0.zip 
Archive:  transpose-2.0.zip
   creating: transpose-2.0/
   creating: transpose-2.0/win32-bin/
  inflating: transpose-2.0/win32-bin/transpose.exe  
   creating: transpose-2.0/src/
  inflating: transpose-2.0/src/transpose.c  
  inflating: transpose-2.0/README 
</code></pre></div>  </div>
</blockquote>

<p>The zip file is now unpacked. Let us now compile the code. ALWAYS have a look at the README file for this.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ cd transpose-2.0
$ head README
To Compile:
	gcc transpose.c -o transpose
To Install - Just copy into your path. e.g.:
	cp transpose /usr/local/bin/
$ cd src/
$ sudo gcc transpose.c -o transpose
</code></pre></div></div>

<p>The program <strong>gcc</strong> compiles the human readable code in the file <strong>transpose.c</strong> and produces a binary file out of it, called <strong>transpose</strong>.</p>
<blockquote>
  <p>We can now run the binary file from within the directory.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ ./transpose --help
       Description:     
	This software is released under the GPL license
	Reshapes delimited text data - amongst other things, it can transpose a matrix of plain text data.
</code></pre></div>  </div>
</blockquote>

<blockquote>
  <p>Create a symbolic link to the newly created binary to /usr/local/bin. This directory collects binaries/commands to be used on the command line.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ sudo ln -s /opt/transpose-2.0/src/transpose /usr/local/bin
$ which transpose 
/usr/local/bin/transpose
</code></pre></div>  </div>
</blockquote>




                    
                    <blockquote class="key_points">
                        <h3><i class="fa fa-key" aria-hidden="true"></i><span class="visually-hidden">keypoints</span> Key points</h3>

                        <ul>
                            
                            <li>Where can I find my files</li>
                            
                            <li>What are useful commands for file system operations</li>
                            
                        </ul>
                    </blockquote>
                    

                    

                    

                </div>
                <!-- sidebar, which will be hidden on a small screen -->
                <nav class="col-md-2 sticky-toc bs-docs-sidebar hidden-print hidden-xs hidden-sm" role="complementary" id="toc" data-toggle="toc"></nav>
            </div>
        </div>

        <h3><i class="fa fa-thumbs-up" aria-hidden="true"></i><span class="visually-hidden">congratulations</span> Congratulations on successfully completing this tutorial!</h3>

        

        

</div>
</div>
</div>
<footer>
   <div class="well row">
       <div class="col-md-2 text-center" style="padding: 25px 20px 25px 20px;">				
         <a href="https://www.bits.vib.be" target="_blank"><img src="/assets/images/bioinformatics_core_rgb_neg.png" width="200"></a>
       </div>
       <div class="col-md-8 text-center" style="padding-top: 20px;">	
        <a href="https://github.com/vibbits/training-material" title="View on GitHub" class="margin20" target="_blank">
                            <i class="fa fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> View on GitHub
                        </a>
       </div>
       <div class="col-md-2 text-center" style="padding-top: 30px;">
         <span class="slogan"><br>SCIENCE MEETS LIFE</span>
       </div>
   </div>
</footer>


</body>
<script type="text/javascript" src="/assets/js/jquery.slim.min.js"></script>
<script type="text/javascript" src="/assets/js/popper.min.js"></script>
<script type="text/javascript" src="/assets/js/bootstrap.min.js?v=3"></script>
<script type="text/javascript" src="/assets/js/details-element-polyfill.js"></script>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script type="text/javascript" src="/assets/js/bootstrap-toc.min.js"></script>
<script type="text/javascript" src="/assets/js/main.js"></script>
<script type="text/javascript" src="/assets/js/clipboard.min.js"></script>
<script type="text/javascript">
    var snippets = document.querySelectorAll('div.highlight');
    [].forEach.call(snippets, function(snippet) {
        snippet.firstChild.insertAdjacentHTML('beforebegin', '<button class="btn btn-light" data-clipboard-snippet><i class="fa fa-copy"></i>&nbsp;Copy</button>');
    });

    var clipboardSnippets = new ClipboardJS('[data-clipboard-snippet]', {
        target: function(trigger) {
            return trigger.nextElementSibling;
        }
    });
</script>

</html>
