<!doctype html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="x-ua-compatible" content="ie=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
        <meta name="google-site-verification" content="dw_1zBYRNxF7-vYNDXh3Ght_JuqgJt3BD1N_Pynqmv0" />
        <title>03 Ecology Analysis using vegan</title>
        <script async="" src="https://www.google-analytics.com/analytics.js"></script> 
        <script>
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
            ga('create', 'UA-61194136-18' , 'auto');
            ga('send', 'pageview');
        </script>
        <link rel="stylesheet" href="/assets/css/bootstrap.min.css?v=3">
        <link rel="stylesheet" href="/assets/css/bootstrap-toc.min.css">
        <link rel="stylesheet" href="/assets/css/font-awesome.css">
        <link rel="stylesheet" href="/assets/css/VIB.css">
        <link rel="stylesheet" href="/assets/css/academicons.css">
        <link rel="stylesheet" href="/assets/css/syntax_highlighting.css">
        <link rel="shortcut icon" href="/favicon.ico" type="image/x-icon" />

        
        
        
        
        
        <meta name="description" content="Metagenomics" />
        <meta property="og:title" content="VIB Bioinformatics Training: 03 Ecology Analysis using vegan" />
        <meta property="og:description" content="Metagenomics" />
        <meta property="og:image" content="/assets/images/bioinformatics_core_rgb_neg.png" />
    </head>
    <body data-spy="scroll" data-target="#toc">
        <div class="top-bar-vib">
 <a href="http://vib.be/en" target="_blank" style="font-size:14px">VIB Home </a>
</div>

<div class="container main-content">

<div class="container page-top__container">
          <div class="region region-top">
  <div id="logo" class="col-md-3 col-lg-3 text-left" style="padding-left:0px;margin-top:20px">
       <a href="/">
           <img src="/assets/images/logo.svg" height="60" alt="VIB logo" class="margin20">
       </a>
    </div>
    <div class="col-md-9 col-lg-9 text-right" style="margin-top:50px;">
    <nav role="navigation" aria-labelledby="block-mainnavigation-menu" id="block-mainnavigation" class="block block-menu navigation menu--main">
  


        <a href="/" class="topmenu">Home</a>


        <a href="/about.html" class="topmenu" >About</a>


        <a href="/hall-of-fame.html" class="topmenu">Contribute</a>


        <a href="/faq.html" class="topmenu">Faq</a>


        <a href="/search/index.html" class="topmenu">Search</a>


  </nav>

    </div>
  </div>
</div>














    <script type="application/ld+json">
        


{
  "@context": "http://schema.org",
  "@type": "Course",
  "accessMode": [
    "textual",
    "visual"
  ],
  "accessModeSufficient": [
    "textual",
    "visual"
  ],
  "accessibilityControl": [
    "fullKeyboardControl",
    "fullMouseControl"
  ],
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "accessibilitySummary": "Short descriptions are present but long descriptions will be needed for non-visual users",
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "copyrightHolder": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "headline": "03 Ecology Analysis using vegan",
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "license": "/blob//LICENSE.md",
  "producer": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "provider": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "sourceOrganization": {
    "@type": "Organization",
    "email": "bits@vib.be",
    "name": "VIB Bioinformatics Core",
    "url": "https://www.bits.vib.be"
  },
  "courseCode": "metagenomics / Ecology_Analysis_using_vegan / hands-on",
  "learningResourceType": "hands-on tutorial",
  "name": "Hands-on for '03 Ecology Analysis using vegan' tutorial",
  "isPartOf": {
    "@type": "Course",
    "name": "03 Ecology Analysis using vegan",
    "description": "03 Ecology Analysis using vegan",
    "learningResourceType": "tutorial",
    "interactivityType": "expositive",
    "provider": {
      "@type": "Organization",
      "email": "bits@vib.be",
      "name": "VIB Bioinformatics Core",
      "url": "https://www.bits.vib.be"
    }
  },
  "url": "https://material.bits.vib.be//topics/metagenomics/tutorials/Ecology_Analysis_using_vegan/tutorial.html",
  "timeRequired": "PT75M",
  "description": "03 Ecology Analysis using vegan\\nThe questions this  addresses are:\n - How to do an ecology analysis with the R package vegan?\n\n\\nThe objectives are:\n - recapitulate the basics of working with R\n - learn how to visualize genus abundance tables\n - learn how to normalize abundances\n - learn how to use the vegan package including clustering and statistical testing\n\n",
  "hasPart": [

  ],
  "mentions": [

  ],
  "author": [
    {
      "@type": "Person",
      "name": "Falk Hildebrand"
    },
    {
      "@type": "Person",
      "name": "Janick Mathys"
    }
  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "Falk Hildebrand"
    },
    {
      "@type": "Person",
      "name": "Janick Mathys"
    }
  ],
  "about": [
    {
      "@type": "CreativeWork",
      "name": "Metagenomics",
      "description": "Metagenomics",
      "url": "https://material.bits.vib.be//topics/metagenomics/"
    }
  ]
}
    </script>

    <div class="row tutorial">
      <div class="col-md-12" role="main">
        <h1 data-toc-skip>03 Ecology Analysis using vegan</h1>
        <div class="contributors-line">By: 

<a href="/hall-of-fame#hildebra" class="contributor-badge">Falk Hildebrand</a>
, <a href="/hall-of-fame#janick-bits" class="contributor-badge">Janick Mathys</a>


</div>
        <blockquote class="overview">

            <strong><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</strong>
            <ul>
            
            <li>How to do an ecology analysis with the R package vegan?</li>
            
            </ul>

            <strong><i class="fa fa-bullseye" aria-hidden="true"></i><span class="visually-hidden">objectives</span> Objectives</strong>
            <ul>
            
            <li>recapitulate the basics of working with R</li>
            
            <li>learn how to visualize genus abundance tables</li>
            
            <li>learn how to normalize abundances</li>
            
            <li>learn how to use the vegan package including clustering and statistical testing</li>
            
            </ul>

            

            
            <p><strong><i class="fa fa-hourglass-end" aria-hidden="true"></i><span class="visually-hidden">time</span> Time estimation:</strong> 75 minutes</p>
            

            


            

        </blockquote>

        <div class="container" data-field="body">
            <div class="row">
                <div class="col-md-10" role="main">
                    <h1 class="no_toc" id="ecology-analysis-using-vegan">Ecology Analysis using vegan</h1>

<p>In this exercise we will look at a data matrix of 16S rRNA counts in 74 samples.</p>

<p>This dataset is the microbiota composition of 74 mice from 5 different mice strains. The original research aim was to define the effect that the mouse genome has on the microbiota and what the effect of living in the same cage would be. However, we found much stronger trends in the data, and these we will look at in this exercise.</p>

<p>The 454 data was already compiled into a matrix with genus abundance per sample in a previous step. This matrix is called a feature abundance matrix, or abundance matrix for short. We will do an ecology-oriented analysis of the data, in later steps also taking metadata (experimental, environmental or clinical data that was collected for each sample, independent of the DNA) into account. The aim of this tutorial is to get an idea of the very basic steps of ecological data analysis using the programming language R.</p>

<p>The gene abundance table (Genus.txt) can be found in the folder /home/VIBTrainingX/NGS/metagenomics/higherLvl folder on the server. Those who are working on their own laptop can download it <a href="http://psbweb05.psb.ugent.be/lotus/data/LotuStutorial.tar.gz">from the lotus website</a>.</p>

<ol>
  <li>Set the folder with the provided files as your working directory in R using <code class="highlighter-rouge">setw</code>. This way required files can be easily loaded. To find out how to use this command, you can type ?setwd() to open the help. If there are other R-commands that you want to know more about, you can open the R-help for that command by entering in the R-prompt <code class="highlighter-rouge">?command</code>. This will be very useful when working with R, make sure to use this a lot as you can only learn more :o).</li>
</ol>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-1"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 1</h3>

  <blockquote class="question">
    <h3 id="question-question"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to set the working directory in R</p>
    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">setwd</span><span class="p">(</span><span class="s2">"dir_to_data"</span><span class="p">)</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<ol>
  <li>Load the provided data into the matrix M (Genus.txt, actual genus abundance data), using the read.delim command, saving the loaded table as <code class="highlighter-rouge">M</code>. Make sure, the row names are correctly read in. As R reads the matrix as an object of class data.frame, we convert M from a data.frame to a matrix <code class="highlighter-rouge">M=as.matrix(M)</code>. This is important for some of the following calculations, where we need a <code class="highlighter-rouge">matrix</code> class object.</li>
</ol>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-1-1"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 1</h3>

  <blockquote class="question">
    <h3 id="question-question-1"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to read in data as matrix ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># read in data as matrix</span><span class="w">
</span><span class="n">M</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">read.delim</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="s2">"Genus.txt"</span><span class="p">,</span><span class="n">row.names</span><span class="o">=</span><span class="m">1</span><span class="p">)</span><span class="w">
</span><span class="n">M</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.matrix</span><span class="p">(</span><span class="n">M</span><span class="p">)</span><span class="w"> 
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>The matrix you loaded represents the number of 16S sequences assignable to each genus, which we could find in the samples. Also note that not all genera are real genera, but partly assigned unknown sequences. With these groups we do not know if this is a single genus or in fact several genera or in extreme cases even several classes, that just all fall under the same phylum tag. What are the advantages and disadvantages of keeping such undefined groups in the data?
Use the function <code class="highlighter-rouge">edit(M)</code> to better view the abundance matrix.</p>

<ol>
  <li>Let’s look at some  basic features of the abundance matrix. The <code class="highlighter-rouge">summary(M)</code> command is a good start, but also look at total row and column counts (<code class="highlighter-rouge">colSums</code>, <code class="highlighter-rouge">rowSums</code> command). To see how the genera are distributed within each sample, we will plot a sample-wise density plot.We will be using a combination of the <code class="highlighter-rouge">density</code>, <code class="highlighter-rouge">lines</code> and <code class="highlighter-rouge">lapply</code> functions, to draw the densities of values found in each sample. Let’s start with looking at the density of the first sample. In R you can access specific columns by writing the matrix coordinates in square brackets. For example <code class="highlighter-rouge">M[1,]</code> shows the first row of a matrix, <code class="highlighter-rouge">M[,7]</code> shows the 7th column etc:</li>
</ol>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-3"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 3</h3>

  <blockquote class="question">
    <h3 id="question-question-2"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to estimate density of first sample ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># estimate density of first sample</span><span class="w">
</span><span class="n">densityOfSample1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">density</span><span class="p">(</span><span class="n">M</span><span class="p">[,</span><span class="m">1</span><span class="p">])</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>Look at the object densityOfSample1 by simply entering the object name into the command prompt. Next try to visualize it with <code class="highlighter-rouge">plot(densityOfSample1)</code>. In this plot you see that most genera are at 0 abundance, some genera have an abundance &lt;10 and some rare genera actually occur with a higher frequency, one genus even having ~1100 16S reads assigned to it. Which genus is this?</p>

<p>Alternatively you can also use the function <code class="highlighter-rouge">hist</code>, to plot a histogram of the abundances. Try to do this now.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-4"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 4</h3>

  <blockquote class="question">
    <h3 id="question-question-3"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to plot histogram of abundances ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># plot histogram of abundances</span><span class="w">
</span><span class="n">hist</span><span class="p">(</span><span class="n">M</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="n">nclass</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">50</span><span class="p">)</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>We can use the <code class="highlighter-rouge">apply</code> command, to apply the density command to every column of M, which will return a list of density objects. The second argument to the <code class="highlighter-rouge">apply</code> function is the <code class="highlighter-rouge">margin</code> and is set to 2, which tells the <code class="highlighter-rouge">apply</code> function that we want to work on columns (margin = 2) and not on rows (margin = 1). Save this into object  <code class="highlighter-rouge">S_densities</code>.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-5"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 5</h3>

  <blockquote class="question">
    <h3 id="question-question-4"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to estimate densities of all samples ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># estimate densities of all samples</span><span class="w">
</span><span class="n">S_densities</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">M</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="n">density</span><span class="p">)</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>To plot this start with:</p>
<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># open a new plot window and set range of x and y axis</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="n">type</span><span class="o">=</span><span class="s2">"n"</span><span class="p">,</span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">3</span><span class="p">),</span><span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">5000</span><span class="p">))</span><span class="w"> 
</span></code></pre></div></div>

<p>This will open a new plotting window, already set to the range of x and y coordinates (xlim, ylim) we will need in this example. In this case we just want to plot a blank space, this is done with the <code class="highlighter-rouge">type=n</code> argument. Try to replace the argument by <code class="highlighter-rouge">type=p</code>, to actually see that point! S_densities is a list, so we use <code class="highlighter-rouge">lapply</code> (list apply), in combination with the <code class="highlighter-rouge">lines</code> function, try this now to plot all the density lines into the open plot window.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-6"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 6</h3>

  <blockquote class="question">
    <h3 id="question-question-5"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to plot density distributions of all samples ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># plot density distributions of all samples</span><span class="w">
</span><span class="n">lapply</span><span class="p">(</span><span class="n">S_densities</span><span class="p">,</span><span class="n">lines</span><span class="p">)</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>What you should see now in the plot window is the density distribution of all samples. The lines function is adding new lines, while a plot function makes a completely new plot. Try to replace the <code class="highlighter-rouge">lines</code> with <code class="highlighter-rouge">plot</code> to see this (it’s very fast, so keep a close eye on your plot). How are these lines already telling us something about the differences between the communities of each sample?</p>

<ol>
  <li>Maybe you noticed that the <code class="highlighter-rouge">colSums</code> command showed that the totals are not equal. What does this mean? In this state the data is actually not comparable among each other. One way to <code class="highlighter-rouge">correct</code> the data for this shortcoming is to normalize the matrix. In this step we will normalize the abundance matrix into variable M1:</li>
</ol>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># normalize matrix: divide each column by the total of that column</span><span class="w">
</span><span class="n">M1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sweep</span><span class="p">(</span><span class="n">M</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="n">colSums</span><span class="p">(</span><span class="n">M</span><span class="p">),</span><span class="s2">"/"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>The <code class="highlighter-rouge">sweep</code> command is extremely useful, as it will apply a simple arithmetic operation (like divide) in a matrix column- or row-wise with a vector of your choice. So it is very similar to <code class="highlighter-rouge">apply</code>, but takes more basic functions. In this case we will divide each column by the sum of the column, this is called normalization.</p>

<p>Now we will compare these matrices using the <code class="highlighter-rouge">barplot</code> function. For this we need to open another graphical window, using the <code class="highlighter-rouge">X11</code> function:</p>
<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># create barplot of original and normalized data</span><span class="w">
</span><span class="n">barplot</span><span class="p">(</span><span class="n">M</span><span class="p">)</span><span class="w">
</span><span class="n">X11</span><span class="p">()</span><span class="w">
</span><span class="n">barplot</span><span class="p">(</span><span class="n">M1</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>What do you notice about the sample composition? What does the graph mean? Discuss where you would want to normalize the data (and where not).</p>

<p>Close all open plots.</p>

<p>Now replot the sample-wise density plot (as you did in step 3), but start the plot with these adapted x and y ranges. Additionally we will this time label the x- and y-axis:</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># open a new plot and define ranges and titles of x and y axis</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="n">type</span><span class="o">=</span><span class="s2">"n"</span><span class="p">,</span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">80</span><span class="p">),</span><span class="n">xlim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">),</span><span class="n">xlab</span><span class="o">=</span><span class="s2">"relative genus abundance"</span><span class="p">,</span><span class="w"> </span><span class="n">ylab</span><span class="o">=</span><span class="s2">"Frequency of genera"</span><span class="p">)</span><span class="w"> 
</span></code></pre></div></div>

<p>You will notice that the graph looks different from you previous plot. What changed due to the normalization? Are the samples more similar to each other using M or M1?</p>

<p>If you spot a difference in species abundance between two samples using matrix M, is this difference real, does it have scientific value?</p>

<p>For the next step the R-library vegan is required. It is a set of functions specifically designed for ecological data analysis. The package has been installed on the bits laptops. If you were to install the package, you could do so using the command: <code class="highlighter-rouge">install.packages(“vegan”)</code>. More details on the <a href="http://cc.oulu.fi/~jarioksa/softhelp/vegan.html">vegan web site </a>. Load vegan, using the <code class="highlighter-rouge">library</code> command.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-7"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 7</h3>

  <blockquote class="question">
    <h3 id="question-question-6"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to load the vegan package ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># load vegan package</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">vegan</span><span class="p">)</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>Let’s try to put the differences we observed in sample density into numbers. To do this, ecologists rely on the concept of diversity. Diversity describes the evenness of species distributions as well as the richness of species that are observed in a given ecological system. We will first calculate the Shannon diversity, using vegan’s <code class="highlighter-rouge">diversity</code> command. Try to do this per sample, using the <code class="highlighter-rouge">apply</code> function again. Save the result in object <code class="highlighter-rouge">div</code>.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-8"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 8</h3>

  <blockquote class="question">
    <h3 id="question-question-7"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to calculate Shannon diversity index for each sample using the normalized data ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># calculate Shannon diversity index for each sample using the normalized data </span><span class="w">
</span><span class="n">div</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">M1</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="n">diversity</span><span class="p">,</span><span class="n">index</span><span class="o">=</span><span class="s2">"shannon"</span><span class="p">)</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>Now we can see in action what these indices are actually doing for us. Plot the density of the sample with the lowest and highest diversity in red and blue on your previous density plot of M1, this you do by first finding out which diversity indexes are the maximum and minimum values using the <code class="highlighter-rouge">which.max</code> and <code class="highlighter-rouge">which.min</code> functions on the object <code class="highlighter-rouge">div</code>. Don’t forget to have the last density plot still open (or replot it from step 4 on M1), than add the lowest samples as a blue line and the highest sample as a red line, using the <code class="highlighter-rouge">lines</code> command.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-9"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 9</h3>

  <blockquote class="question">
    <h3 id="question-question-8"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>Find samples with lowest and highest Shannon diversity index and add them to the density plot ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># find samples with lowest and highest Shannon diversity index and add them to the density plot</span><span class="w">
</span><span class="n">which.min</span><span class="p">(</span><span class="n">div</span><span class="p">)</span><span class="w"> </span><span class="c1">#should be bl16</span><span class="w">
</span><span class="n">which.max</span><span class="p">(</span><span class="n">div</span><span class="p">)</span><span class="w"> </span><span class="c1">#should be bl48</span><span class="w">
</span><span class="n">lines</span><span class="p">(</span><span class="n">density</span><span class="p">(</span><span class="n">M1</span><span class="p">[,</span><span class="s2">"bl16"</span><span class="p">],</span><span class="n">adjust</span><span class="w"> </span><span class="o">=</span><span class="m">0.5</span><span class="p">),</span><span class="n">col</span><span class="o">=</span><span class="s2">"blue"</span><span class="p">)</span><span class="w">
</span><span class="n">lines</span><span class="p">(</span><span class="n">density</span><span class="p">(</span><span class="n">M1</span><span class="p">[,</span><span class="s2">"bl48"</span><span class="p">],</span><span class="n">adjust</span><span class="w"> </span><span class="o">=</span><span class="m">0.5</span><span class="p">),</span><span class="n">col</span><span class="o">=</span><span class="s2">"red"</span><span class="p">)</span><span class="o">&amp;</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>You can now readjust the window by changing the <code class="highlighter-rouge">ylim</code> and <code class="highlighter-rouge">xlim</code> attribute in the plot function, if necessary (tip, try to rerun using <code class="highlighter-rouge">ylim=c(0,180)</code>). Try to explain why the colored samples have the highest &amp; lowest diversity. What does this tell about an ecosystem (remember that these are genus abundances).
Raise your hand if you reached this step.</p>

<p>A different way to normalize the data is to sample exactly equal amounts of 16S rDNA for each sample in this experiment. Of course in practice this is impossible to do, but we can simulate this, by randomly selecting a subset of 16S rDNA. This is called rarefaction. Rarefy your original abundance matrix (M) into M2, using 1000 reads per sample, using the <code class="highlighter-rouge">rrarefy</code> function of vegan. Note that you need to transpose (command <code class="highlighter-rouge">t()</code>) the matrix, before giving it to <code class="highlighter-rouge">rrarefy</code>. Transform the matrix back and save it as <code class="highlighter-rouge">M2</code>.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-10"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 10</h3>

  <blockquote class="question">
    <h3 id="question-question-9"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to normalize via rarefaction ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Alternative way of normalization</span><span class="w">
</span><span class="n">M2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">t</span><span class="p">(</span><span class="n">rrarefy</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">M</span><span class="p">),</span><span class="n">sample</span><span class="o">=</span><span class="m">2000</span><span class="p">))</span><span class="w">  </span><span class="c1">#vegan needs transformed matrix, and we need it back-transformed</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>Use <code class="highlighter-rouge">colSums(M2)</code> to check if the rarefaction worked. The main use of rarefaction is in calculating diversity and richness correctly, for this we will look in the following step at observed richness.</p>

<p>The concept of observed richness within a sample is pretty simple (but useful): richness describes the number of different species that occur at least once in a sample. This can be calculated in two steps:</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Species present in sample: TRUE or 1 if species is present, FALSE or 0 if species is absent</span><span class="w">
</span><span class="n">OnceOrMoreOftenPresent</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">M1</span><span class="o">&gt;</span><span class="m">0</span><span class="w">
</span></code></pre></div></div>

<p>The sum of each column in this matrix will tell us how many species were detected in total within the respective sample, use the <code class="highlighter-rouge">apply</code> and <code class="highlighter-rouge">sum</code> functions , saving the result in <code class="highlighter-rouge">rich1</code>.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-11"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 11</h3>

  <blockquote class="question">
    <h3 id="question-question-10"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to calculate the sum of each column ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Calculate sum of each column</span><span class="w">
</span><span class="n">rich1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">OnceOrMoreOftenPresent</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="n">sum</span><span class="p">)</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>Compare the richness values in <code class="highlighter-rouge">rich1</code> to the richness obtained on the rarefied matrix <code class="highlighter-rouge">M2</code>, calculated with a shortened command:</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Calculate number of present species in each sample using the rarefied data</span><span class="w">
</span><span class="n">rich2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">M2</span><span class="o">&gt;</span><span class="m">0</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="n">sum</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Compare rich1 and rich2 in a matrix value by value. We use the <code class="highlighter-rouge">cbind</code> command to bind two vectors column wise together, so we get a matrix with 2 columns. Order this matrix by the richness values in rich1, using the <code class="highlighter-rouge">order</code> command and accessing the vector representation with <code class="highlighter-rouge">[]</code> square brackets.</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Create new matrix with two columns: rich1 and rich2 and order rows according to rich1 values</span><span class="w">
</span><span class="n">cbind</span><span class="p">(</span><span class="n">rich1</span><span class="p">,</span><span class="n">rich2</span><span class="p">)[</span><span class="n">order</span><span class="p">(</span><span class="n">rich1</span><span class="p">),]</span><span class="w">
</span></code></pre></div></div>

<p>What does the second part of the formula do? What happens if you change that to order(rich2)?</p>

<p>Discuss which richness values have the highest value to the researcher and why the order is very different between these two richness estimates. Is one way clearly wrong?</p>

<p>Why did we choose 1000 as cutoff for the sequences per sample? What is the maximum value we could choose?</p>

<p>First samples are clustered to see underlying data structures. For this tutorial we will choose a hierarchical clustering, based on a bray-curtis distance between samples, using the function <code class="highlighter-rouge">vegdist</code>. Make  sure the distances are calculated between Samples and not Genera.</p>

<p>Next, use the function <code class="highlighter-rouge">hclust</code> on the distance matrix, saving the output in variable <code class="highlighter-rouge">cluster</code>, and subsequently plot the clustering of the samples (using <code class="highlighter-rouge">plot</code>).
Take a guess of how many groups there might be in this clustering?</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-11-1"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 11</h3>

  <blockquote class="question">
    <h3 id="question-question-11"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to cluster samples and plot results ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># cluster samples and plot results</span><span class="w">
</span><span class="n">BCD</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">vegdist</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">M1</span><span class="p">),</span><span class="w"> </span><span class="n">dist</span><span class="o">=</span><span class="s2">"bray"</span><span class="p">)</span><span class="w">
</span><span class="n">cluster</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">hclust</span><span class="p">(</span><span class="n">BCD</span><span class="p">)</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>To visualize the samples and their relatedness to each other in a two-dimensional space, we can use an ordination to visualize the data in a low dimensional space. The dimensionality of the original matrix (73 genera=73 dimensions) is reduced to two dimensions. If you know what a PCA (Principal component analysis) is, this step will use a conceptually similar, but methodologically quite different technique to perform an ordination of the data, NMDS (non-metric multidimensional scaling).</p>

<p>Start by calculating a 2-dimensional NMDS of the data using M1, using the Bray-Curtis distance in the function <code class="highlighter-rouge">metaMDS</code>, saving the result to <code class="highlighter-rouge">nmds</code>. Again, make sure that samples are being ordinated and not Genera.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-11-2"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 11</h3>

  <blockquote class="question">
    <h3 id="question-question-12"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to calculate the NMDS ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># calculate NMDS</span><span class="w">
</span><span class="n">nmds</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">metaMDS</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">M1</span><span class="p">),</span><span class="n">distance</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bray"</span><span class="p">)</span><span class="w"> </span><span class="c1">#actual NMDS command, matrix needs to be transformed to conform with vegan’s standards</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>Take a look at the <code class="highlighter-rouge">nmds</code> object and explore some of its features (e.g. type <code class="highlighter-rouge">str(nmds)</code> to see what variables are stored within the NMDS object). Try to find out what the <code class="highlighter-rouge">stress</code> of your ordination is. What does stress stand for (tip: go to the R help on metaMDS)? Next we can visualize the NMDS, similar to what you get out of PCA’s, displaying samples only:</p>
<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># plot NMDS</span><span class="w">
</span><span class="n">plot</span><span class="p">(</span><span class="n">nmds</span><span class="p">,</span><span class="n">display</span><span class="w"> </span><span class="o">=</span><span class="s2">"sites"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>The important difference of NMDS compared to PCA is, that NMDS works with any kind of distance metric, while PCA can only use Euclidean distances between samples. A second important feature of NMDS is, that this method finds non-parametric, monotonic relationships between objects; in short: it doesn’t assume a specific data distribution. Why might these two features be important for ecologists?</p>

<p>You might have noticed that you see two clusters, similar to the hierarchical clustering of the data. We can get for each sample the identity within the two clusters using the <code class="highlighter-rouge">cutree</code> commands, specifying k=2 (2 clusters). This can be plotted into the NMDS with the following command:</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># identify clusters</span><span class="w">
</span><span class="n">memb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cutree</span><span class="p">(</span><span class="n">cluster</span><span class="p">,</span><span class="w"> </span><span class="n">k</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span><span class="n">ordispider</span><span class="p">(</span><span class="n">nmds</span><span class="p">,</span><span class="n">memb</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Congratulations, you have just visualized the mouse enterotypes. Next we are going to look closer at these. If you want to know the exact methods to detect enterotypes in your data visit [http://enterotype.embl.de/enterotypes.html http://enterotype.embl.de/enterotypes.html]</p>

<p>In the last step, we will test for all the genera in the matrix whether they show significant differences between two clusters. The scientific question we are posing here is: what are the significant differences in the gut microbiota of between enterotypes? We will use a non-parametric test (kruskal-wallis) to do the tests, as ecological data is in most cases not normally distributed. This test is very similar to the student t-test, and the interpretation works just the same way. Use the function <code class="highlighter-rouge">kruskal.test</code> to test the first genera (M[1,]) for significant differences between the two cluster groups (in object <code class="highlighter-rouge">memb</code>). Save the output of this command in variable <code class="highlighter-rouge">Kt</code>.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-12"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 12</h3>

  <blockquote class="question">
    <h3 id="question-question-13"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to test if there is a difference between the two clusters for the first genus ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Test if there is a difference between the two clusters for the first genus</span><span class="w">
</span><span class="n">Kt</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">kruskal.test</span><span class="p">(</span><span class="n">M1</span><span class="p">[</span><span class="m">1</span><span class="p">,],</span><span class="n">memb</span><span class="p">)</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>Look at the output of this function. This will show you a human readable summary of the test and the result. You can access elements of a list (<code class="highlighter-rouge">Kt</code> is a list in this case) using the <code class="highlighter-rouge">$</code> operator. Try to extract the p-value from the <code class="highlighter-rouge">Kt</code> object.</p>

<p>Once you know how, we can start to calculate the significance for every genus in the M1 matrix,. These p-values we will store in a newly created vector <code class="highlighter-rouge">pvals</code>. Let’s add the first 2 p-values to the vector:</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Test if there is a difference between the two clusters for the first and second genera. Store p-values in a vector.</span><span class="w">
</span><span class="n">pvals</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">()</span><span class="w">
</span><span class="n">pvals</span><span class="p">[</span><span class="m">1</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">kruskal.test</span><span class="p">(</span><span class="n">M1</span><span class="p">[</span><span class="m">1</span><span class="p">,],</span><span class="w"> </span><span class="n">memb</span><span class="p">)</span><span class="o">$</span><span class="n">p.value</span><span class="w">
</span><span class="n">pvals</span><span class="p">[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">kruskal.test</span><span class="p">(</span><span class="n">M1</span><span class="p">[</span><span class="m">2</span><span class="p">,],</span><span class="w"> </span><span class="n">memb</span><span class="p">)</span><span class="o">$</span><span class="n">p.value</span><span class="w">
</span></code></pre></div></div>

<p>Since doing this 73 times takes a long time, we will be using a for-loop to <code class="highlighter-rouge">loop</code> over the matrix and do this for us. We could as well use the apply function, but the syntax would get a little more complicated, since we are only interested in a subpart of the result, the $p.value part. Try to write a for-loop, to calculate the p-value 73 times.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-13"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 13</h3>

  <blockquote class="question">
    <h3 id="question-question-14"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to test if there is a difference between the two clusters for all genera ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Test if there is a difference between the two clusters for all genera</span><span class="w">
</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="nf">dim</span><span class="p">(</span><span class="n">M1</span><span class="p">)[</span><span class="m">1</span><span class="p">])</span><span class="w">
</span><span class="p">{</span><span class="w">
        </span><span class="n">pvals</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">kruskal.test</span><span class="p">(</span><span class="n">M1</span><span class="p">[</span><span class="n">i</span><span class="p">,],</span><span class="w"> </span><span class="n">memb</span><span class="p">)</span><span class="o">$</span><span class="n">p.value</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>As an additional help, you can add the name of the taxa to the pvals vector using the names command (that will name a vector):</p>

<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Add names to the vector</span><span class="w">
</span><span class="nf">names</span><span class="p">(</span><span class="n">pvals</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">dimnames</span><span class="p">(</span><span class="n">M1</span><span class="p">)[[</span><span class="m">1</span><span class="p">]]</span><span class="w"> 
</span></code></pre></div></div>

<p>Which taxa are significantly different?</p>

<p>In this case we will use the normalized M1 matrix, can you explain why we do not use the M or M2 matrix? Would either be wrong to use?</p>

<p>In total we were testing in 73 genera, if their p-value was below a threshold of 0.05. What is the chance of observing data with a p-value &gt;0.05 by random chance? How many genera do you expect to be below this threshold by random chance?</p>

<p>To avoid statistical errors of this kind, we will use a Benjamini-Hochberg multiple testing correction, implemented in the R function <code class="highlighter-rouge">p.adjust</code>. Save the result as <code class="highlighter-rouge">qvals</code>.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-exercise-14"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Exercise 14</h3>

  <blockquote class="question">
    <h3 id="question-question-15"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

    <p>How to perform multiple testing correction of p-values using Benjamini-Hochberg method ?</p>

    <blockquote>
      <details>
        <summary><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution
</summary>
        <div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Multiple testing correction of p-values using Benjamini-Hochberg method</span><span class="w">
</span><span class="n">qvals</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">p.adjust</span><span class="p">(</span><span class="n">pvals</span><span class="p">,</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"hochberg"</span><span class="p">)</span><span class="w">
</span></code></pre></div>        </div>
      </details>
    </blockquote>

  </blockquote>
</blockquote>

<p>What do you see in this test? What would you report on this dataset, based on these values?</p>

<p>Try sorting the q-values to see the most significant differences first:</p>
<div class="language-R highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c1"># Sorting q-values</span><span class="w">
</span><span class="n">sort</span><span class="p">(</span><span class="n">qvals</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Now that you have finished the tutorials, you should be able to analyze any new dataset of amplicon data, using the LotuS pipeline and performing a basic analysis with R, including</p>
<ul>
  <li>Data normalization</li>
  <li>Clustering analysis</li>
  <li>Ordination</li>
  <li>Univariate statistics
You can always expand upon these concepts, using this tutorial as starting point. Just remember that R is a very flexible language, and all these commands can be expanded for new purposes and visualizations.</li>
</ul>

<h3 id="data-sources">Data sources</h3>
<p>All the material provided in this tutorial are from metagenomic study on mice knockouts. Further analysis of the data can be found in the reference below.</p>

<h3 id="reference">Reference</h3>

<p>Hildebrand, F., Nguyen, A. T. L., Brinkman, B., Yunta, R. G., Cauwe, B., Vandenabeele, P., … Raes, J. (2013). Inflammation-associated enterotypes, host genotype, cage and inter-individual effects drive gut microbiota variation in common laboratory mice. Genome Biology, 14(1), R4. doi:10.1186/gb-2013-14-1-r4</p>




                    
                    <blockquote class="key_points">
                        <h3><i class="fa fa-key" aria-hidden="true"></i><span class="visually-hidden">keypoints</span> Key points</h3>

                        <ul>
                            
                            <li>We showed how to install and load packages</li>
                            
                            <li>We showed how to access the help system</li>
                            
                            <li>We showed how to set the folder to work in</li>
                            
                            <li>We showed how to define variables</li>
                            
                            <li>We showed how to write a script</li>
                            
                        </ul>
                    </blockquote>
                    

                    
                    <h1 data-toc-skip>Useful literature</h1>
                    <p>Further information, including links to documentation and original publications, regarding the tools, analysis techniques and the interpretation of results described in this tutorial can be found <a href="/topics/metagenomics#references">here</a>.</p>
                    

                    

                </div>
                <!-- sidebar, which will be hidden on a small screen -->
                <nav class="col-md-2 sticky-toc bs-docs-sidebar hidden-print hidden-xs hidden-sm" role="complementary" id="toc" data-toggle="toc"></nav>
            </div>
        </div>

        <h3><i class="fa fa-thumbs-up" aria-hidden="true"></i><span class="visually-hidden">congratulations</span> Congratulations on successfully completing this tutorial!</h3>

        

        

</div>
</div>
</div>
<footer>
   <div class="well row">
       <div class="col-md-2 text-center" style="padding: 25px 20px 25px 20px;">				
         <a href="https://www.bits.vib.be" target="_blank"><img src="/assets/images/bioinformatics_core_rgb_neg.png" width="200"></a>
       </div>
       <div class="col-md-8 text-center" style="padding-top: 20px;">	
        <a href="https://github.com/vibbits/training-material" title="View on GitHub" class="margin20" target="_blank">
                            <i class="fa fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> View on GitHub
                        </a>
       </div>
       <div class="col-md-2 text-center" style="padding-top: 30px;">
         <span class="slogan"><br>SCIENCE MEETS LIFE</span>
       </div>
   </div>
</footer>


    </body>
    <script type="text/javascript" src="/assets/js/jquery.slim.min.js"></script>
    <script type="text/javascript" src="/assets/js/popper.min.js"></script>
    <script type="text/javascript" src="/assets/js/bootstrap.min.js?v=3"></script>
    <script type="text/javascript" src="/assets/js/details-element-polyfill.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="/assets/js/bootstrap-toc.min.js"></script>
    <script type="text/javascript" src="/assets/js/main.js"></script>
    <script type="text/javascript" src="/assets/js/clipboard.min.js"></script>
    <script type="text/javascript" src="https://app.cloudsh.com/js/cloudsh.js"></script>
    <script type="text/javascript">
    var snippets=document.querySelectorAll('div.highlight');
    [].forEach.call(snippets,function(snippet){
        snippet.firstChild.insertAdjacentHTML('beforebegin','<button class="btn btn-light" data-clipboard-snippet><i class="fa fa-copy"></i>&nbsp;Copy</button>');
    });

    var clipboardSnippets=new ClipboardJS('[data-clipboard-snippet]',{
        target:function(trigger){return trigger.nextElementSibling;
    }});
    </script>
</html>
