<!doctype html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="x-ua-compatible" content="ie=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="google-site-verification" content="dw_1zBYRNxF7-vYNDXh3Ght_JuqgJt3BD1N_Pynqmv0" />
    <title>05 Introduction to RNA-Seq analysis</title>
    <script async="" src="https://www.google-analytics.com/analytics.js"></script>
    <script>
        (function(i, s, o, g, r, a, m) {
            i['GoogleAnalyticsObject'] = r;
            i[r] = i[r] || function() {
                (i[r].q = i[r].q || []).push(arguments)
            }, i[r].l = 1 * new Date();
            a = s.createElement(o),
                m = s.getElementsByTagName(o)[0];
            a.async = 1;
            a.src = g;
            m.parentNode.insertBefore(a, m)
        })(window, document, 'script', 'https://www.google-analytics.com/analytics.js', 'ga');
        ga('create', 'UA-61194136-18', 'auto');
        ga('send', 'pageview');
    </script>
    <link rel="stylesheet" href="/assets/css/bootstrap.min.css?v=3">
    <link rel="stylesheet" href="/assets/css/bootstrap-toc.min.css">
    <link rel="stylesheet" href="/assets/css/font-awesome.css">
    <link rel="stylesheet" href="/assets/css/main.scss">
    <link rel="stylesheet" href="/assets/css/VIB.css">
    <link rel="stylesheet" href="/assets/css/academicons.css">
    <link rel="stylesheet" href="/assets/css/syntax_highlighting.css">
    <link rel="shortcut icon" href="/favicon.ico " type="image/x-icon" />     
    <meta name="description" content="Collection of tutorials developed and maintained by the V..." />
    <meta property="og:title" content="VIB Bioinformatics Training: 05 Introduction to RNA-Seq analysis" />
    <meta property="og:description" content="Collection of tutorials developed and maintained by the V..." />
    <meta property="og:image" content="/assets/images/bioinformatics_core_rgb_neg.png" />
</head>

<body data-spy="scroll" data-target="#toc">
    <div class="top-bar-vib">
 <a href="http://vib.be/en" target="_blank" style="font-size:14px">VIB Home </a>
</div>

<div class="container main-content">

<div class="container page-top__container">
          <div class="region region-top">
  <div id="logo" class="col-md-3 col-lg-3 text-left" style="padding-left:0px;margin-top:20px">
       <a href="/">
           <img src="/assets/images/logo.svg" height="60" alt="VIB logo" class="margin20">
       </a>
    </div>
    <div class="col-md-9 col-lg-9 text-right" style="margin-top:50px;">
    <nav role="navigation" aria-labelledby="block-mainnavigation-menu" id="block-mainnavigation" class="block block-menu navigation menu--main">
  


        <a href="/" class="topmenu">Home</a>


        <a href="/about.html" class="topmenu" >About</a>


        <a href="/hall-of-fame.html" class="topmenu">Contribute</a>


        <a href="/faq.html" class="topmenu">Faq</a>


        <a href="/search/index.html" class="topmenu">Search</a>


  </nav>

    </div>
  </div>
</div>














    <script type="application/ld+json">
        


{}
    </script>

    <div class="row tutorial">
      <div class="col-md-12" role="main">
        <h1 data-toc-skip>05 Introduction to RNA-Seq analysis</h1>
        <div class="contributors-line">By: 

<a href="/hall-of-fame#janick-bits" class="contributor-badge">Janick Mathys</a>


</div>
        <blockquote class="overview">

            <strong><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</strong>
            <ul>
            
            <li>Download the data from GEO or ENA</li>
            
            </ul>

            <strong><i class="fa fa-bullseye" aria-hidden="true"></i><span class="visually-hidden">objectives</span> Objectives</strong>
            <ul>
            
            <li>Basic knowledge of downloading data from GEO or ENA</li>
            
            </ul>

            

            
            <p><strong><i class="fa fa-hourglass-end" aria-hidden="true"></i><span class="visually-hidden">time</span> Time estimation:</strong> 15 minutes</p>
            

            


            

        </blockquote>

        <div class="container" data-field="body">
            <div class="row">
                <div class="col-md-10" role="main">
                    <p><a href="http://data.bits.vib.be/pub/trainingen/NGSIntro/NGSBroad.pdf">Download the slides for this training session</a>.</p>

<p>The dataset comes from a 2014 publication on <em>Human Airway Smooth Muscle Transcriptome Changes in Response to Asthma Medications</em>.
The goal of the analysis is to find DE genes (differentially expressed: genes with different expression levels in one group of samples compared to other groups of samples). Typically the groups of samples represent different treatments: one consisting of biological replicates that have received a control treatment, others consisting of replicates that received a specific biological treatment.</p>

<p>In this experiment the data consists of four groups (<strong>treatment</strong>):</p>
<ul>
  <li>The <strong>dex</strong> group: samples from 4 cell lines after treatment with the glucocorticoid dexamethasone (dex), used as astma medication</li>
  <li>The <strong>alb</strong> group: samples from the same cell lines after treatment with albuterol (alb), another astma medication</li>
  <li>The <strong>alb_dex</strong> group: samples from the same cell lines after treatment with both astma medications</li>
  <li>The <strong>untreated</strong> group: samples from the same untreated cell lines cultured in parallel.</li>
</ul>

<p>So all samples come from the same 4 cell lines (<strong>cells</strong>).</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>#   run_accession  read_count  samples            cells    treatment
1   SRR1039508     22935521    CL1_untreated   CL1   untreated
2   SRR1039509     21155707    CL1_Dex         CL1   Dex
3   SRR1039510     22852619    CL1_Alb         CL1   Alb
4   SRR1039511     21938637    CL1_Alb_Dex     CL1   Alb_Dex
5   SRR1039512     28136282    CL2_untreated   CL2  untreated
...
</code></pre></div></div>
<p>The data comes from a paired-end sequencing experiment so we have two files for each sample. 
For simplicity we will do the analysis on a single sample, SRR1039509, obtained from dexamethasone treated cell line 1.</p>

<h3 id="quality-checks">Quality checks</h3>

<p>Before you analyze the data, it is crucial to check the quality of the data.
We use the standard tool for checking the quality of NGS data generated on the Illumina platform: <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FASTQC</a></p>

<p>Correct interpretation of the FASTQC report is very important.
If the quality of your data is good, you can proceed with the analysis.
<strong>!! If the quality of your data is very bad, don’t immediately throw the data in the recycle bin but contact an expert and ask for his/her opinion. !!</strong></p>

<p>Double click the FASTQC icon on the Desktop and open the fastq file (it’s in the summer folder of your home folder). FASTQC consists of multiple modules each checking a specific aspect of the quality of the data. On the first page you can select the module you wish to view.
The names of the modules are preceded by an icon that reflects the quality of the data. The icon indicates whether the results of the module seem normal (green tick), slightly abnormal (orange triangle) or very unusual (red cross).</p>

<p>However, these evaluations must be interpreted in the context of what you expect from your library. A ‘normal’ sample as far as FastQC is concerned is random and diverse. Some experiments may be expected to produce libraries which are biased. You should treat the icons as pointers to where you should concentrate your attention on and understand why your library may not look normal.</p>

<h4 id="general-information-on-the-reads">General information on the reads</h4>

<blockquote>
  <p>How long are the reads in this file ?
 63 nucleotides
 <img src="../../images/FASTQCRNASeqB1.png" alt="" /></p>
</blockquote>

<h4 id="checking-the-quality-scores-of-the-reads">Checking the quality scores of the reads</h4>

<p>Phred scores represent base call quality. The higher the score the more reliable the base call. Often the quality of reads degrades over the length of the read. Therefore, it is common practice to determine the average quality of the first, second, third,…nth base by plotting the distribution of the Phred scores on each position of the reads using box plots.</p>

<blockquote>
  <p>Evaluate the quality scores per position
 Go to the <strong>Per base sequence quality</strong> module: 
 <img src="../../images/FASTQCRNASeqB2.png" alt="" />
The y-axis on the graph shows the Phred quality scores, the x-axis shows the position in the read. So again you see that the reads are 63 bases long.</p>
</blockquote>

<p>The average Phred score is depicted by the blue line, the median Phred score by the red line. The yellow boxes contain 50% of all Phred scores on a certain position. As expected the quality is steadily declining.</p>

<p>The background of the graph divides the y-axis into very good quality calls (green), calls of reasonable quality (orange), and calls of poor quality (red; Phred score &lt; 20). As you can see the Phred scores of this data set are very high.</p>

<p>Remark: In new Illumina kits the sequence quality goes up a bit first before it steadily declines.</p>

<p>Instead of showing the quality of each position separately, you can calculate the average Phred score of each read and show a cumulative plot of the average qualities of all the reads.</p>

<blockquote>
  <p>Evaluate the overall quality
Go to the <strong>Per sequence quality scores</strong> module: 
 <img src="../../images/FASTQCRNASeqB3.png" alt="" />
The y-axis on the graph shows the number of reads, the x-axis shows the Phred score.</p>
</blockquote>

<p>Most reads have an average Phred score of 40. This is a very high score (Phred scores of Illumina calls range from -5 to 41).</p>

<p>Illumina flow cells are divided into tiles. To see if there is a loss in quality associated with specific parts of the flow cell, FASTQC calculates average quality scores for each tile across all positions in the reads.</p>

<blockquote>
  <p>Evaluate the quality per tile
Go to the <strong>Per tile sequence quality</strong> module: 
 <img src="../../images/FASTQCRNASeqB4.png" alt="" />
The y-axis on the graph shows the tile number, the x-axis shows the position in the reads.</p>
</blockquote>

<p>The plot shows the deviation from the average tile quality. The colours are on a cold to hot scale, with blue being the average tile quality and other colours representing tiles where the quality was different from the average. In the example you see that a few tiles show poor quality over a few positions. A good plot should be blue all over. Although the plot isn’t entirely blue the results of this module are still acceptable.</p>

<p>Reasons for seeing warnings or failures on this plot could be transient problems such as bubbles going through the flow cell, or they could be more permanent problems such as smudges or debris on/in the flow cell or a very high density of clusters in a tile. The most common cause of warnings in this module is the flow cell being overloaded.</p>

<p>It is recommended to ignore warnings/failures which mildly affect a small number of tiles for only a few cycles, and to only pursue larger effects which show high deviation in scores, or which persist for a high number of cycles.</p>

<h4 id="checking-duplicates">Checking duplicates</h4>

<p>In a diverse library generated by shearing genomic DNA, most fragments will occur only once. A low level of duplication may indicate a very high level of coverage of some target sequences, but a high level of duplication indicates a bias (eg PCR overamplification, contamination of the library with adapter dimers…).</p>

<p>The <strong>Sequence duplication levels</strong> module counts the degree of duplication for every read and creates a plot showing the relative number of reads with different degrees of duplication.</p>

<blockquote>
  <p>Evaluate the sequence duplication levels
Go to the <strong>Sequence duplication levels</strong> module: 
 <img src="../../images/FASTQCRNASeq5.png" alt="" />
The y-axis on the graph shows the percentage of occurrence, the x-axis shows the duplication level.</p>
</blockquote>

<p>The blue line represents the counts of all duplicated sequences. The percentage is computed relative to the total number of reads.</p>

<p>The red line represents the number of <strong>distinct</strong> sequences that are duplicated. The percentage is computed relative to the total number of <strong>distinct</strong> sequences in the data (see slides).</p>

<p>Seeing duplication in RNA-Seq data is normal. To sequence lowly expressed genes you must oversequence the genes with high expression levels. However, RNA-Seq libraries may be contaminated with adapters.</p>

<p>The presence of contaminating adapters will produce spikes on the far right of the plot. These peaks will appear in the blue trace as they make up a high proportion of the original library, but usually disappear in the red trace as they make up a small proportion of the deduplicated set (you only use 2 adapters to create a library).</p>

<p>Additionally, as in every RNA-Seq file you also see a substantial number of oversequenced reads with lower duplication levels.</p>

<p>Since the reads are random fragments from the genome sequence, the contribution of A, C, G and T should be identical on each position.</p>

<blockquote>
  <p>Evaluate the per base sequence content
Go to the <strong>Per base sequence content</strong> module:</p>
</blockquote>

<p><img src="../../images/FASTQCRNASeqB6.png" alt="" /></p>

<p>The y-axis on the graph shows the percentage of occurrence, the x-axis shows the position in the read.</p>

<p>On this plot you should see straight lines for the four nucleotides. In reality you often see that this is not the case for the first positions. Libraries produced by priming using random hexamers (nearly all RNA-Seq libraries) and those which were fragmented using transposases inherit an intrinsic bias in the first positions of the reads. This bias does not come from a single sequence, but because of enrichment of a number of different K-mers at the 5’ end of the reads. So it isn’t something you can correct by trimming (you do not have one specific sequence that you can remove from the reads). In most cases it doesn’t adversely affect the downstream analysis but it will produce a warning or failure in this module.</p>

<p>Duplicates often arise because libraries are contaminated with adapter sequences. You can check for contaminating sequences using the <strong>Overrepresented sequences</strong> module: it lists all sequences which make up more than 0.1% of the total. For each sequence in the list the module will look for matches in a database of common contaminants and will report the best hit it finds.</p>

<blockquote>
  <p>Which contaminants are found in this library ?
Go to the <strong>Overrepresented sequences</strong> module:</p>
</blockquote>

<p><img src="../../images/FASTQCRNASeqB7.png" alt="" /></p>

<p>As you can see, a single illumina adapter was found contaminating the library to a small extent: 0,5% of the library consists of adapter sequences.</p>

<p>The <strong>Overrepresented sequences</strong> module shows contamination with full adapter sequences (= reads that completely correspond to adapters), but often the library also contains reads that have remnants of adapter sequences at their 3’ ends. These reads are not detected by the <strong>Overrepresented sequences</strong> module.</p>

<p>This was the quality check of one file from one of the 16 samples. We do not have the time to do all quality checks. But in the real world, you would have to do this for each of the 32 files of this experiment.</p>

<h3 id="improving-the-quality-of-the-data">Improving the quality of the data</h3>
<p>There are many possible steps to improve the quality of the data. Due to time constraints, we are going to focus on</p>

<ul>
  <li>removing adapter sequences, both filtering full adapter sequences and trimming remnants of adapters from the 3’ ends of the reads</li>
  <li>filter reads of low quality</li>
</ul>

<p>There are many tools to remove adapters from reads, but we chose cutadapt because it works on paired-end reads and it can do the two steps at the same time (removing adapters and filtering reads of poor quality).</p>

<p>To make it feasible to go through the complete RNA-Seq workflow during the training, we have limited the data set to reads that map to chromosome 22. The data come from a paired-end experiment so we have two files with reads. You can download these limited data sets: [http://data.bits.vib.be/pub/trainingen/NGSIntro/chr22_SRR1039509_1.fastq.gz data_set_1] (first reads of a pair that map to chromosome 22) and [http://data.bits.vib.be/pub/trainingen/NGSIntro/chr22_SRR1039509_2.fastq.gz data_set_2] (second reads of a pair that map to chromosome 22). On the bits laptops, the files are already present in the /home/bits/NGS/RNASeq/ folder.</p>

<p>Remember that the <strong>Overrepresented sequence</strong> module of the FASTQC report showed contamination with the following TruSeq adapter in the first file of sample SRR1039509:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ACACGTCTGAACTCCAGTCACTGACCAATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAAAAA
</code></pre></div></div>
<p>We will remove this adapter from the file containing the reads that map to chromosome 22.</p>

<p>Open the terminal.</p>

<blockquote>
  <p>Make a variable called adapter to hold the sequence of the adapter
|-</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>adapter=ACACGTCTGAACTCCAGTCACTGACCAATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAAAAA
</code></pre></div>  </div>
  <p>Remember to use Shift + Insert to paste in the terminal !</p>
</blockquote>

<blockquote>
  <p>Check the cutadapt option for defining the number of mismatches you allow (= error rate)
Cutadapt is not a regular bash command but a Python script so it doesn’t have a manual. So to open the cutadapt help type:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cutadapt -h
</code></pre></div>  </div>
</blockquote>

<p>Scrolling down the help file shows that the <strong>-e</strong> option defines the maximum allowed error rate: the default is 0.1 meaning that it allows one mismatch every 10 nucleotides. Adapters are identified by aligning each read to the adapter sequence: if the frequency of mismatches in the alignment is below the allowed error rate then the adapter is trimmed from the read.</p>

<blockquote>
  <p>Check the option for defining the adapter sequence
In the help file you see that you have multiple options:</p>
</blockquote>

<ul>
  <li><strong>-a</strong> to trim adapter sequences at the 3’ end of the reads. In most cases this is the end that is causing the problems: when small RNA fragments are sequenced, the resulting reads can be longer than the RNA fragments. As a results they will contain (parts of) the adapter at their 3’end. In long reads the adapter might even lie within the read:
    <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>MYSEQUEN                         (no adapter contaimination)
MYSEQUENCEADAP                   (part of adapter at 3' end)
MYSEQUENCEADAPTER                (adapter at 3' end)
MYSEQUENCEADAPTERSOMETHINGELSE   (adapter within the read)
</code></pre></div>    </div>
    <p>Cutadapt will cut the adapter (part) and all sequence following it resulting in:</p>
    <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>MYSEQUEN
MYSEQUENCE
MYSEQUENCE
MYSEQUENCE
</code></pre></div>    </div>
  </li>
  <li><strong>-g</strong> to trim adapter sequences at the 5’ end of the reads. These adapters are expected to appear at the start of a read (where they can be just partially there) or somewhere within the read:
    <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ADAPTERMYSEQUENCE              (5' end)
DAPTERMYSEQUENCE               (partial)
TERMYSEQUENCE                  (partial)
SOMETHINGADAPTERMYSEQUENCE     (within)
</code></pre></div>    </div>
    <p>In all cases, the adapter itself and the sequence preceding it will be removed, leaving in all examples above:</p>
    <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>MYSEQUENCE
</code></pre></div>    </div>
  </li>
  <li><strong>-b</strong> to trim adapters at the 3’ or 5’ end of the read. If there is at least one base before the adapter, then the adapter is trimmed as a 3’ adapter and the adapter itself and everything following it is removed. Otherwise, the adapter is trimmed as a 5’ adapter and it is removed from the read, but the sequence after it remains:
    <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Before trimming 	        After trimming 	
MYSEQUENCEADAPTERSOMETHING 	MYSEQUENCE 
MYSEQUENCEADAPTER 	        MYSEQUENCE
MYSEQUENCEADAP 	                MYSEQUENCE
MADAPTER 	                M
ADAPTERMYSEQUENCE 	        MYSEQUENCE
PTERMYSEQUENCE 	                MYSEQUENCE
TERMYSEQUENCE 	                MYSEQUENCE
</code></pre></div>    </div>
  </li>
</ul>

<p>Since we have contaminating adapter at the 3’end we’ll take the -a option</p>

<p>We will use a few other options:</p>

<ul>
  <li>Discard trimmed reads that are shorter than 20 bases after trimming using the <strong>-m</strong> option</li>
  <li>Trim low-quality bases at the 3’ends from reads before adapter removal if their Phred score is less than 10 using the <strong>-q</strong> option</li>
</ul>

<blockquote>
  <p>Check the -q option in the help file
Scroll down to the <strong>Additional modifications to the reads:</strong> section to check the usage of the -q option:</p>
</blockquote>

<p><img src="../../images/CLcutadapt1.png" alt="" /></p>

<blockquote>
  <p>Check the -m option in the help file
Scroll up to the <strong>Options for filtering of processed reads:</strong> section to check the usage of the -m option:</p>
</blockquote>

<p><img src="../../images/CLcutadapt2.png" alt="" /></p>

<p>Remember that we are working with paired-end reads !</p>

<blockquote>
  <p>Check the usage of cutadapt for paired-end reads in the help file
Scroll up to the start of the help file to check the usage of cutadapt for paired-end reads:</p>
</blockquote>

<p><img src="../../images/CLcutadapt3.png" alt="" /></p>

<p>Since we have to specify the location in the file system of two input and two output files, we are going to create a variable called folder for holding the path.</p>

<blockquote>
  <p>Create the variable path
|-</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>folder=/home/bits/NGS/RNASeq/
</code></pre></div>  </div>
</blockquote>

<p>Remember to use tab autocompletion.</p>

<blockquote>
  <p>Clean up the files using the knowledge you have obtained
|-</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cutadapt -a ${adapter} -q 10 -m 20 -o ${path}chr22_SRR1039509_1t.fastq -p ${path}chr22_SRR1039509_2t.fastq ${path}chr22_SRR1039509_1.fastq.gz ${path}chr22_SRR1039509_2.fastq.gz
</code></pre></div>  </div>
</blockquote>

<ul>
  <li><strong>-a</strong> to specify the sequence of the adapter and to specify that we want to cut adapter sequences from the 3’ends of the reads</li>
  <li><strong>-q</strong> to specify the minimal quality score is 10. Bases at the 3’end of the reads with a quality score below 10 will be removed</li>
  <li><strong>-m</strong> to specify the minimal length of the read after trimming. Reads smaller than 20 bases will be removed</li>
  <li><strong>-o</strong> to specify the location in the file system where you want to write the output files</li>
  <li><strong>-p</strong> to specify the location in the file system where you want to write the results for the reads from the other end of the fragments. As such you specify that these are paired-end reads.</li>
</ul>

<p>In the cutadapt stats you see we only trimmed one file (containing sequences from one end of the fragments).</p>

<p><img src="../../images/CLcutadapt4.png" alt="" /></p>

<p>That is because the sequences from the other end of the fragments contain another adapter:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>GTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATTAAAAAAAAAAAAAAAAAA
</code></pre></div></div>

<blockquote>
  <p>Redefine the adapter variable
|-</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>adapter=GTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATTAAAAAAAAAAAAAAAAAA
</code></pre></div>  </div>
  <p>Remember to use Shift + Insert to paste in the terminal.</p>
</blockquote>

<blockquote>
  <p>Run cutadapt again to remove the other adapter
Remember to switch the two files now:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cutadapt -a ${adapter} -q 10 -m 20 -o ${path}chr22_SRR1039509_2trim.fastq -p ${path}chr22_SRR1039509_1trim.fastq ${path}chr22_SRR1039509_2t.fastq ${path}chr22_SRR1039509_1t.fastq
</code></pre></div>  </div>
  <p>Remember to use the up arrow to go back in the history.</p>
</blockquote>

<p>Now you see in the cutadapt stats that you have trimmed adapters from (both files) both ends of the fragments.</p>

<p><img src="../../images/CLcutadapt5.png" alt="" /></p>

<h3 id="check-the-quality-of-the-cleaned-reads">Check the quality of the cleaned reads</h3>

<p>I have done this also for the complete files and rechecked the trimmed reads in FASTQC. You can download <a href="http://data.bits.vib.be/hidden/jhslbjcgnchjdgksqngcvgqdlsjcnv/ngsrnade2015/ex1/fastqc_SRR1039509_1/trimmed_reads/SRR1039509_1.fastq.pdf">the report for the complete trimmed reads from sample SRR1039509</a>.</p>

<blockquote>
  <p>Are all the reads still 63 nt long after trimming ?
In the <strong>Basic statistics</strong> tab you see that the length of the reads varies between 20 (in the cutadapt command we set 20 as the minimal length for a read to be retained) and 63 (reads that were not trimmed):
 <img src="../../images/FASTQCRNASeqB9.png" alt="" /></p>
</blockquote>

<blockquote>
  <p>Have the quality scores of the reads significantly changed after trimming ?
The <strong>Per base sequence quality</strong> is similar to that of the untrimmed file, as is the <strong>Per sequence quality</strong>.</p>
</blockquote>

<p>Quality scores have of course changed a bit since we trimmed low quality bases, but the initial quality of the reads was so good that you don’t really see the effect of the trimming.</p>

<blockquote>
  <p>Has the per base sequence content improved as a result of the trimming ?
The <strong>Per base sequence content</strong> - the tool to detect adapter contamination - plot has improved, it’s even more stable now.</p>
</blockquote>

<p><img src="../../images/FASTQCRNASeq10.png" alt="" /></p>

<blockquote>
  <p>What is the little bump you see in the Sequence length distribution plot ?</p>

  <p>Apparently many reads contain 3 bases that belong to the adapter. These 3 bases have been cut leaving reads of 60 nt long: this is the small peak you see on the plot at length 60. All intermediate lengths of adapter contamination have been detected but in such a small fraction of reads that you cannot see the influence of the trimming on the plot.</p>
</blockquote>

<blockquote>
  <p>Are there any overrepresented sequences left ?
No.</p>
</blockquote>

<p><img src="../../images/FASTQCRNASeq11.png" alt="" /></p>

<blockquote>
  <p>Are there any overrepresented heptamers ?
FASTQC still detects overrepresented heptamers although at much lower counts than before.</p>
</blockquote>

<p><img src="../../images/fastqcTrim5.png" alt="" /></p>

<p>FastQC confirmed the removal of the two adapters by cutadapt.</p>

<h3 id="mapping">Mapping</h3>

<h4 id="obtaining-the-reference-genome">Obtaining the reference genome</h4>

<p>Before we can do any mapping we need a reference sequence first. We will map the reads against the hg19 human genome build. Mapping requires a specially formatted file (hash database). This hash database can be derived from the reference genome using the bowtie2 tools. However, for some organisms like human the hash table can be obtained ‘ready-to-use’ from the bowtie2 website. If you also need a fasta copy of the hg19 genome, you can obtain it from the hash table using bowtie2. 
We can download the hash table from the <a href="ftp://ftp.ccb.jhu.edu/pub/data/bowtie2_indexes/hg19.zip">bowtie2 website</a> using the <strong>wget</strong> command. It takes about 90 minutes to download so we are not going to download it during the training, it is already present in the /home/bits/NGS/RNASeq/reference/ folder.</p>

<p>Go to this folder and look at its contents. As you can see the file is a compressed .zip file</p>

<blockquote>
  <p>Unzip the reference genome file
To decompress a .zip file you need the <strong>unzip</strong> command:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>unzip hg19.zip
</code></pre></div>  </div>
</blockquote>

<p>It will take a few minutes and it will generate 7 files:</p>

<p><img src="../../images/MapRNASeq1.png" alt="" /></p>

<p>To transform the hash table into a fasta sequence we use bowtie2. From the <a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#the-bowtie2-inspect-index-inspector">bowtie2 documentation</a> we learn that we should use bowtie2-inspect without options to generate the fasta file.</p>

<blockquote>
  <p>Generate the human reference genome sequence in fasta format.
The required command simply is:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bowtie2-inspect hg19 &gt; hg19.fa
</code></pre></div>  </div>
</blockquote>

<p>The <strong>&gt;</strong> sign indicates that the output of the command (so the fasta file with the reference sequence) needs to be written to a file called <strong>hg19.fa</strong>.
It will take a few minutes to generate the file.</p>

<p><img src="../../images/MapRNASeq2.png" alt="" /></p>

<h4 id="installing-tophat">Installing TopHat</h4>

<p>Mapping RNA-Seq reads is done using the TopHat tool. So we need to install the <a href="http://ccb.jhu.edu/software/tophat/tutorial.shtml">TopHat tool</a>.</p>

<blockquote>
  <p>Go to the TopHat website and fetch the download link.</p>

  <ul>
    <li>Go to the <a href="http://ccb.jhu.edu/software/tophat/tutorial.shtml">TopHat website</a></li>
    <li>Right click the Linux download link</li>
    <li>Select <strong>Copy Link Location</strong></li>
  </ul>
</blockquote>

<blockquote>
  <p>Download the file.</p>

  <ul>
    <li>Go to the terimnal</li>
    <li>Navigate to the /usr/bin/NGS/ folder</li>
    <li>Type **wget **</li>
    <li>Press the Shift and Insert keys simultaneously to paste the url</li>
  </ul>
</blockquote>

<p>TopHat is downloaded as a .tar.gz file</p>

<blockquote>
  <p>Decompress the file
For decompressing a .tar.gz file you need the following command:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>tar -xzvf tophat-2.1.0.Linux_x86_64.tar.gz
</code></pre></div>  </div>
  <p>Remember to use tab autocompletion !</p>
</blockquote>

<p>This creates a new folder:</p>

<p><img src="../../images/MapRNASeq3.png" alt="" /></p>

<p>Go into the tophat folder and type:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>./tophat
</code></pre></div></div>

<p>If this opens the help of tophat, it means the software has been installed correctly. It does not mean that you can use the software now. Well you can but you will always have to type the commands from inside the tophat folder or provide the full path to the tophat folder. To avoid this we can create a symbolic link for tophat2.</p>

<blockquote>
  <p>Create a symbolic link for tophat2
For creating the link you need the following command:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ln -s /usr/bin/NGS/tophat-2.1.0.Linux_x86_64/tophat2 /usr/local/bin/tophat2
</code></pre></div>  </div>
  <p>Remember to use tab autocompletion !</p>
</blockquote>

<p>Now go to a different folder and type <strong>tophat2</strong>. If you see the help file, the link works.</p>

<h4 id="installing-samtools">Installing samtools</h4>

<p>When you navigate to the <strong>tophat</strong> folder in /usr/bin/NGS/ you see that samtools is automatically installed when TopHat was installed:
 <img src="../../images/MapRNASeq5.png" alt="" /></p>

<p>If you see the samtools help page when you type</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>./samtools_0.1.18
</code></pre></div></div>
<p>it means that samtools is indeed installed
 <img src="../../images/MapRNASeq6.png" alt="" /></p>

<p>If you want to use samtools from anywhere in the file system you can create a soft link.</p>

<blockquote>
  <p>Create a soft link for samtools.
Create a link using the <strong>ln -s</strong> command:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ln -s /usr/bin/NGS/tophat-2.1.0_Linux_x86_64/samtools-0.1.18/samtools /usr/local/bin/samtools
</code></pre></div>  </div>
  <p>Go up one directory in the file system and check if the command works. If you type</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>samtools view
</code></pre></div>  </div>
  <p>(one of the possible samtools commands) you should see the manual of the command.</p>
</blockquote>

<h4 id="mapping-the-reads">Mapping the reads</h4>

<p>We are not going to do the actual mapping since it takes almost 25 minutes even with the chromosome22-limited datasets. If we were to map the reads we would use the following command:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>folder=/home/bits/NGS/RNASeq/
tophat2 --no-coverage-search ${folder}reference/hg19 ${folder}chr22_SRR1039509_1.fastq.gz ${folder}chr22_SRR1039509_2.fastq.gz 
</code></pre></div></div>

<ul>
  <li><strong>–no-coverage-search</strong>: is related to how TopHat finds splice junctions. I’m not going to go into detail here but in the TopHat manual the developers of TopHat say: “We only suggest users use the <strong>–coverage-search</strong> option for short reads (&lt; 45bp) and with a small number of reads (&lt;= 10 million).” Since we have the double amount of longer reads (63bp) we have to go for the <strong>–no-coverage-search</strong> option.</li>
  <li>the first argument is the location of the hash table of the reference genome</li>
  <li>the second argument is the (cleaned) fastq file containing the reads from one end of the fragments. As you can see TopHat can work directly on the compressed file.</li>
  <li>the third argument is the (cleaned) fastq file containing the reads from the other end of the fragments</li>
</ul>

<p>Other useful options for Tophat:</p>

<ul>
  <li><strong>-p</strong>: the number of processors (cpu) that TopHat can use for the mapping. The default is 1. This is ok for a laptop since laptops do not contain manu cpu but of course the more cpu you give TopHat the faster the mapping. So it’s better to do the mapping on a strong computer with many cpu</li>
  <li><strong>-o</strong>: if you want to store the results of the mapping in another folder</li>
</ul>

<p>The mapping generates a new folder <strong>tophat_out</strong> containing 3 .bed files and 2 .bam files containing the resulting alignments:</p>

<ul>
  <li><strong>accepted_hits.bam</strong>: a list of read alignments.</li>
  <li><strong>unmapped.bam</strong>: a list of reads that could not be mapped. As you can see the size of this file is quite small compared to the accepted_hits.</li>
  <li><strong>junctions.bed</strong>: a list of splice junctions between exons in <a href="http://genome.ucsc.edu/FAQ/FAQformat.html#format1">UCSC BED format</a> (that can be opened as a track in the UCSC genome browser).
Tophat can find novel - not yet annotated - splice junctions based on the alignment of the reads to a reference genome. This is what Tophat is specifically good at, compared to mappers like bwa and bowtie which will only find annotated splice junctions. This is why we use Tophat for mapping RNA-Seq data.</li>
  <li><strong>insertions.bed</strong>: a list of insertions.</li>
  <li><strong>deletions.bed</strong>: a list of deletions.</li>
</ul>

<p><img src="../../images/MapRNASeq4.png" alt="" />
Since we haven’t actually done the mapping, we do not have this folder. However, you can find the bam file with the read alignments in the /home/bits/NGS/RNASeq folder.</p>

<h4 id="quality-control-of-the-mapping">Quality control of the mapping</h4>

<p>It is vital to check the quality of mapping before proceeding with the RNASeq workflow. The mapping to a reference genome has sorted the reads and it is now possible to identify</p>
<ul>
  <li>the regions of the genome the reads originate from</li>
  <li>duplicate reads</li>
  <li>RNA degradation…</li>
</ul>

<p>Several program exist to perform quality control of bam files; e.g. RSeQC pubmed: 22743226, <a href="http://qualimap.bioinfo.cipf.es/">QualiMap</a> pubmed: 22914218 , samtools, <a href="https://github.com/fidelram/deepTools/wiki/All-command-line-options">deeptools</a> pubmed: 24799436 , <a href="http://broadinstitute.github.io/picard/">Picard</a> which is part of the very popular GATK platform - pubmed: 20644199…</p>

<p>We are going to use <strong>samtools</strong> here.</p>

<blockquote>
  <p>Get the basic stats of the bam file.<br />
On the <a href="http://davetang.org/wiki/tiki-index.php?page=SAMTools">samtools wiki</a></p>
</blockquote>

<p><img src="../../images/samtools2b.png" alt="" /></p>

<p>You can see that you need the <strong>samtools flagstat</strong> command for this. The bam file is in the /home/bits/NGS/RNASeq/ folder so we are going to reuse the folder variable that we created for the mapping:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>folder=/home/bits/NGS/RNAseq/
samtools_0.1.18 flagstat ${folder}accepted_hits_chr22.bam
</code></pre></div></div>

<p>or if you have done the mapping yourself:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>samtools_0.1.18 flagstat /usr/bin/NGS/tophat-2.1.0_Linux_x86_64/tophat_out/accepted_hits_chr22.bam
</code></pre></div></div>

<p>The samtools flagstat command displays an overview of the alignment results on your screen. You just see that 100% of the reads were mapped. This is extremely high but it is of course because we reversed engineered our chromosome 22 limited data set. From the complete fastq files we took the reads that mapped to chromosome 22 so it’s normal that we get an almost perfect mapping.</p>

<p><img src="../../images/MapRNASeq7.png" alt="" /></p>

<p>This overview deserves some explanation:</p>

<ul>
  <li><strong>nan</strong> means <strong>Not A Number</strong> (e.g: divided by 0 )</li>
  <li><strong>paired in sequencing</strong> means reads that belong to a pair regardless of the fact that they are really mapped as a pair</li>
  <li><strong>read1</strong> means forward reads</li>
  <li><strong>read2</strong> means reverse reads</li>
  <li><strong>properly paired</strong> means that both mates of a read pair map to the same chromosome, oriented towards each other, and with a sensible insert size</li>
  <li><strong>with itself and mate mapped</strong> means that both reads of a pair map to the genome but they are not necessarily properly paired, they just map somewhere on the genome</li>
  <li><strong>singletons</strong> means that one of the reads of a pair is unmapped while its mate is mapped</li>
  <li><strong>with mate mapped to a different chr</strong> means reads with a mate mapped on a different chromosome</li>
  <li><strong>with mate mapped to a different chr (mapQ &gt;= 5)</strong> means reads with a mate mapped on a different chromosome having a mapping quality greater than 5</li>
</ul>

<blockquote>
  <p>Compare the number of forward and reverse reads in the paired-end experiment.<br />
the counts of forward and reverse reads are to be found on the lines ending with read1 and read2 respectively. As you see the number of forward reads exceeds the number of reverse reads by 55.</p>
</blockquote>

<blockquote>
  <p>How many reads were mapped as a pair in the paired-end experiment? <br />
814320 reads were properly mapped as a pair, that’s 99,46% of the total number of reads.</p>
</blockquote>

<p>Tools like Qualimap, RSeqQC and Picard will give much more detailed information on the quality of the mapping. Unfortunately we do not have time to use them.</p>

<h3 id="calculating-a-count-table">Calculating a count table</h3>

<p>In order to compute differential expression between groups of samples, we need to convert mapping results to read counts for each gene in each sample. The counting can also be done in R using various packages but will be slower as compared to command-line tools.
We will use the popular <a href="http://www-huber.embl.de/users/anders/HTSeq/doc/count.html">HTSeq-count tool</a> to compute gene counts.</p>

<h4 id="prepare-the-alignment-file">Prepare the alignment file</h4>

<p>We need to sort the bam file since we have paired-end reads. HTSeq assumes the file is sorted so that reads belonging to the same pair are in adjacent lines. If you don’t sort the bam file by read name, HTSeq will think there are lot of reads with missing mates. 
In the <a href="http://www.htslib.org/doc/samtools.html">samtools manual</a> we can look up which command we need to do the sorting.</p>

<blockquote>
  <p>Sort the reads in the .bam file by name
As you can see in the manual the <strong>samtools sort</strong> command sorts .bam files:</p>
</blockquote>

<p><img src="../../images/CountTable7.png" alt="" /></p>

<p>The input and output file are located in the /home/bits/NGS/RNASeq/ folder (or the /usr/bin/NGS/tophat-2.1.0.Linux_x86_64/tophat_out/ folder if you have done the mapping yourself). We are going to create a folder variable:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>folder=/home/bits/NGS/RNASeq/
samtools_0.1.18 sort -n ${folder}accepted_hits.bam ${folder}accepted_hits_sort
</code></pre></div></div>

<p>Go to the folder where input and output file are stored and check if the sorted .bam file was generated:
According to <a href="http://www-huber.embl.de/users/anders/HTSeq/doc/count.html">the HTSeq manual</a> the input file for HTSeq contains the aligned reads in <strong>SAM</strong> format. In our case the mapping generated a .bam file. Fortunately samtools contains scripts to convert BAM format to SAM.</p>

<p>In the <a href="http://www.htslib.org/doc/samtools.html">samtools manual</a> we can look up which command we need to do the transformation.</p>

<blockquote>
  <p>Transform the .bam into a .sam file<br />
As you can see in the manual the <strong>samtools view</strong> command can transform any alignment format into standard SAM format:</p>
</blockquote>

<p><img src="../../images/CountTable8.png" alt="" /></p>

<p>We are going to reuse the folder variable:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>samtools_0.1.18 view ${folder}accepted_hits_sort.bam &gt; ${folder}accepted_hits.sam
</code></pre></div></div>

<p>Go to the folder where input and output file are stored and check if the .sam file was generated.</p>

<h4 id="obtaining-a-reference-annotation-file">Obtaining a reference annotation file</h4>

<p>To calculate read counts we need a gtf file containing the annotation of all exons. You can obtain such files from genome annotation databases such as NCBI, Ensembl, and UCSC. The problem is that there are small differences between the formats of annotation files coming from different databases. These differences have implications for counting the reads.
For instance, we used pre-built index files from the Bowtie website for the mapping. These files have UCSC format. So it seems obvious to use UCSC annotation files for the counting. However, HTSeq prefers Ensembl gtf files. As stated in the <a href="http://www-huber.embl.de/users/anders/HTSeq/doc/count.html">HTSeq documentation</a> using gtf file generated by UCSC will result in very low counts. In the UCSC files, the gene_id incorrectly contains the same value as the transcript_id. Hence, if a read maps to an exon shared by several transcripts of the same gene, this will appear to htseq-count as an overlap between different genes since the different transcripts have different gene_ids. The read will be considered ambiguous and not counted. Therefore, the counts will be incorrect.
As a solution, HTSeq recommends to use a gtf file from Ensembl. You can find Ensembl gtf files on the <a href="ftp://ftp.ensembl.org/pub/">Ensembl ftp server</a>. The version that we need is called grch37 (this corresponds to UCSC genome build hg19). So you can download the gtf file from <a href="ftp://ftp.ensembl.org/pub/grch37/">this web site</a>.</p>

<p>Navigate to the /home/bits/NGS/RNASeq/reference/ folder:</p>

<blockquote>
  <p>Decompress the gtf file.<br />
.gz files are decompressed by the <strong>gunzip</strong> command:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>gunzip Homo_sapiens.GRCh37.82.gtf.gz 
</code></pre></div>  </div>
  <p>Use Tab autocompletion for the name of the file.</p>
</blockquote>

<blockquote>
  <p>Look at the first 10 lines of the gtf file.<br />
Use the <strong>head</strong> command to preview the file:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>head Homo_sapiens.GRCh37.82.gtf
</code></pre></div>  </div>
  <p>Use Tab autocompletion for the name of the file.</p>
</blockquote>

<p><img src="../../images/CountTable3B.png" alt="" /></p>

<p>As you can see the first column of the file contains chromosome numbers. Ensembl uses 1, 2, 3… as chromosome IDs.</p>

<blockquote>
  <p>Look at the first 10 lines of the sam file.<br />
Use the <strong>head</strong> command to preview the file:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>head accepted_hits.sam
</code></pre></div>  </div>
  <p>Use Tab autocompletion for the name of the file.</p>
</blockquote>

<p><img src="../../images/CountTable3C.png" alt="" /></p>

<p>As you can see the third column of the file contains chromosome IDs but they have UCSC format: <strong>chr</strong>22 (remember that all reads come from chromosome 22 in our example). So we need to:</p>

<ul>
  <li>Filter the annotation for chromosome 22 from the gtf file to limit processing time.</li>
  <li>Transform Ensembl chromosome IDs into UCSC format.</li>
</ul>

<p>First of all we’ll give the gtf file a simple name to simplify processing.</p>

<blockquote>
  <p>Use the move command to rename the file to hg19_EnsGene.gtf<br />
Use the <strong>mv</strong> command to rename the file:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mv Homo_sapiens.GRCh37.82.gtf ./hg19_EnsGene.gtf
</code></pre></div>  </div>
  <p>The <strong>./</strong> defines to move the file to the current folder (the folder that you are in when you type the command). So you will move the file to the same folder but under another name, which corresponds to just renaming it.</p>
</blockquote>

<blockquote>
  <p>Filter chromsome 22 annotations from the gtf file. Name the resulting file chr22_Ens.gtf
Chromosome 22 annotations are lines starting with <strong>22</strong>. Use the <strong>grep</strong> command to filter the file:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>grep "^22" hg19_EnsGene.gtf &gt; chr22_Ens.gtf
</code></pre></div>  </div>
  <p>The <strong>^</strong> defines the start of a line. So <strong>^22</strong> means: search for lines that start with <strong>22</strong>.</p>
</blockquote>

<blockquote>
  <p>Look at the first 10 lines of the filtered gtf file.<br />
Use the <strong>head</strong> command to preview the file:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>head chr22_Ens.gtf
</code></pre></div>  </div>
  <p>Use Tab autocompletion for the name of the file.</p>
</blockquote>

<p><img src="../../images/CountTable3D.png" alt="" /></p>

<p>Now, we still need to transform the Ensembl chromosome IDs into UCSC format, meaning that we simply need to add the prefix <strong>chr</strong> to each line in the filtered gtf file. You’ll need the <strong>sed</strong> command for this. Look at <a href="http://www.grymoire.com/Unix/Sed.html">the sed documentation</a> the sed documentation] before you try to do the substitution. To add the word <strong>chr</strong> to the start of each line, you essentially need to replace the start of a line by <strong>chr</strong>.</p>

<blockquote>
  <p>Add the prefix to each line of the gtf file. Name the resulting file chr22_Ens_corr.gtf
To do a replacement or substitution you need to use the <strong>s</strong> command, followed by what you want to replace and what to replace it with, each separated by a <strong>/</strong>. Remember from the filtering exercise that the start of a line is represented by <strong>^</strong>. So use the following command to make the substitution:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sed 's/^/chr/' chr22_Ens.gtf &gt; chr22_Ens_corr.gtf
</code></pre></div>  </div>
  <p>Use Tab autocompletion for the name of the file.</p>
</blockquote>

<blockquote>
  <p>Look at the first 10 lines of the substituted gtf file.<br />
Use the <strong>head</strong> command to preview the file:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>head chr22_Ens_corr.gtf
</code></pre></div>  </div>
  <p>Use Tab autocompletion for the name of the file.</p>
</blockquote>

<p><img src="../../images/CountTable3E.png" alt="" /></p>

<h4 id="installing-htseq">Installing HTSeq</h4>

<p>HTSeq is a Python script. Python scripts can be installed using the <strong>pip install</strong> command. Remember you need administrator privileges for installing tools.</p>

<blockquote>
  <p>Try to install HTSeq. What happens ?<br />
Use the following command to install the tool:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pip install HTSeq
</code></pre></div>  </div>
</blockquote>

<p>As you can see this generates an error and the tool is not installed.</p>

<p><img src="../../images/CountTable4.png" alt="" /></p>

<p>Looking up the error in Google leads to <a href="http://www.cyberciti.biz/faq/debian-ubuntu-linux-python-h-file-not-found-error-solution/">this web page</a>, where you can find the solution to the problem: some dependencies are missing.</p>

<blockquote>
  <p>Install the missing dependencies and try again. What happens ?<br />
Use the following command to install the dependencies:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>apt-get install python-dev
</code></pre></div>  </div>
</blockquote>

<p>Then try to install HTseq again:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pip install HTSeq
</code></pre></div></div>

<p>As you can see this generates a new error and the tool is not installed.</p>

<p><img src="../../images/CountTable5.png" alt="" /></p>

<p>Looking up the error in Google leads to <a href="http://ubuntuforums.org/archive/index.php/t-345201.html">this web page</a>, where you can find the solution to the problem: the C compiler is missing.</p>

<blockquote>
  <p>Install the missing compiler and try again. What happens ?<br />
Use the following command to install the dependencies:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>apt-get install g++
</code></pre></div>  </div>
</blockquote>

<p>Then try to install HTseq again:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pip install HTSeq
</code></pre></div></div>

<p>As you can see this does not generate an error. To check if the tool works type:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>htseq-count
</code></pre></div></div>

<p>If this displays the help file, you know that the tool was correctly installed</p>

<p><img src="../../images/CountTable6.png" alt="" /></p>

<h4 id="calculating-the-count-table">Calculating the count table</h4>

<p><strong>HTSeq</strong> counts reads in different <strong>modes</strong>: 
<a href="http://www-huber.embl.de/users/anders/HTSeq/doc/_images/count_modes.png">HTSeq</a></p>

<p>In the <a href="http://www-huber.embl.de/users/anders/HTSeq/doc/count.html">HTSeq manual</a> we get an overview of the options we can use.</p>

<blockquote>
  <p>How to define that we want to use the union mode ?
The <strong>-m</strong> option allows to define the mode:
 <img src="../../images/CountTable10.png" alt="" /></p>
</blockquote>

<blockquote>
  <p>How to define that this was not a strand-specific experiment ?
The <strong>-s</strong> option allows to define if a strand-specific library was used:
 <img src="../../images/CountTable11.png" alt="" /></p>
</blockquote>

<blockquote>
  <p>How to define the minimum alignment quality score for a read to be counted ?
The <strong>-a</strong> option allows to define the minimum alignment score:
 <img src="../../images/CountTable12.png" alt="" /></p>
</blockquote>

<p>We’ll go for the default mininimum alignment score of 10 (90% confidence).</p>

<blockquote>
  <p>How to define we want to count based on exon annotation ?
The <strong>-t</strong> option allows to define the feature to base the count on:
 <img src="../../images/CountTable13.png" alt="" /></p>
</blockquote>

<p>For a .gtf file <strong>exon</strong> is the default. It means HTSeq will count the number of reads that align to each exon and then combine the counts for all exons of a transcript variant.</p>

<blockquote>
  <p>How to define the feature we want to use as an ID after the counting ?
The <strong>-i</strong> option allows to define the feature to use as ID.</p>
</blockquote>

<p>For a .gtf file gene_id is the default: it means that the output of HTSeq will be a list of gene_ids and for each gene_id you’ll see the number of reads that align to all its exons.</p>

<blockquote>
  <p>Calculate the count table
HTSeq was installed by <strong>pip install</strong> which automatically creates a link. So the HTSeq commands will work from anywhere in the file system. We will go to the folder that contains the input sam file: <strong>/home/bits/NGS/RNAseq/</strong> and run the command from there or create a variable folder containing the path.
In the <a href="http://www-huber.embl.de/users/anders/HTSeq/doc/count.html">HTSeq manual</a> we get an overview of the options we can use. Default options should not be defined.</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>folder=/home/bits/NGS/RNAseq/
htseq-count -m union -s no ${folder}accepted_hits.sam ${folder}reference/chr22_Ens_corr.gtf &gt; ${folder}chr22_counts.txt
</code></pre></div>  </div>
</blockquote>

<blockquote>
  <p>View the first 10 lines of the resulting count table</p>

  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>head chr22_counts.txt
</code></pre></div>  </div>
</blockquote>

<p><img src="../../images/CountTable14.png" alt="" />
You nicely see the read counts for each gene…</p>



                    
                    <blockquote class="key_points">
                        <h3><i class="fa fa-key" aria-hidden="true"></i><span class="visually-hidden">keypoints</span> Key points</h3>

                        <ul>
                            
                            <li>SRA ID</li>
                            
                            <li>GEO</li>
                            
                            <li>ENA</li>
                            
                        </ul>
                    </blockquote>
                    

                    

                    

                </div>
                <!-- sidebar, which will be hidden on a small screen -->
                <nav class="col-md-2 sticky-toc bs-docs-sidebar hidden-print hidden-xs hidden-sm" role="complementary" id="toc" data-toggle="toc"></nav>
            </div>
        </div>

        <h3><i class="fa fa-thumbs-up" aria-hidden="true"></i><span class="visually-hidden">congratulations</span> Congratulations on successfully completing this tutorial!</h3>

        

        

</div>
</div>
</div>
<footer>
   <div class="well row">
       <div class="col-md-2 text-center" style="padding: 25px 20px 25px 20px;">				
         <a href="https://www.bits.vib.be" target="_blank"><img src="/assets/images/bioinformatics_core_rgb_neg.png" width="200"></a>
       </div>
       <div class="col-md-8 text-center" style="padding-top: 20px;">	
        <a href="https://github.com/vibbits/training-material" title="View on GitHub" class="margin20" target="_blank">
                            <i class="fa fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> View on GitHub
                        </a>
       </div>
       <div class="col-md-2 text-center" style="padding-top: 30px;">
         <span class="slogan"><br>SCIENCE MEETS LIFE</span>
       </div>
   </div>
</footer>


</body>
<script type="text/javascript" src="/assets/js/jquery.slim.min.js"></script>
<script type="text/javascript" src="/assets/js/popper.min.js"></script>
<script type="text/javascript" src="/assets/js/bootstrap.min.js?v=3"></script>
<script type="text/javascript" src="/assets/js/details-element-polyfill.js"></script>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script type="text/javascript" src="/assets/js/bootstrap-toc.min.js"></script>
<script type="text/javascript" src="/assets/js/main.js"></script>
<script type="text/javascript" src="/assets/js/clipboard.min.js"></script>
<script type="text/javascript">
    var snippets = document.querySelectorAll('div.highlight');
    [].forEach.call(snippets, function(snippet) {
        snippet.firstChild.insertAdjacentHTML('beforebegin', '<button class="btn btn-light" data-clipboard-snippet><i class="fa fa-copy"></i>&nbsp;Copy</button>');
    });

    var clipboardSnippets = new ClipboardJS('[data-clipboard-snippet]', {
        target: function(trigger) {
            return trigger.nextElementSibling;
        }
    });
</script>

</html>
